// @webex/calling/dist/CallingClient/registration/webWorker.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs2/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/asyncToGenerator"));
var _uuid = require("uuid");
var _types = require("../../common/types");
/* eslint-env worker */

var keepaliveTimer;
var messageHandler = function messageHandler(event) {
  var type = event.data.type;
  var postKeepAlive = /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee(accessToken, deviceUrl, url) {
      var response;
      return _regenerator.default.wrap(function (_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            _context.next = 1;
            return fetch("".concat(url, "/status"), {
              method: _types.HTTP_METHODS.POST,
              headers: {
                'cisco-device-url': deviceUrl,
                'spark-user-agent': 'webex-calling/beta',
                Authorization: "".concat(accessToken),
                trackingId: "web_worker_".concat((0, _uuid.v4)())
              }
            });
          case 1:
            response = _context.sent;
            if (response.ok) {
              _context.next = 2;
              break;
            }
            throw response;
          case 2:
            return _context.abrupt("return", response);
          case 3:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    return function postKeepAlive(_x, _x2, _x3) {
      return _ref.apply(this, arguments);
    };
  }();
  if (type === _types.WorkerMessageType.START_KEEPALIVE) {
    var keepAliveRetryCount = 0;
    var _event$data = event.data,
      accessToken = _event$data.accessToken,
      deviceUrl = _event$data.deviceUrl,
      interval = _event$data.interval,
      retryCountThreshold = _event$data.retryCountThreshold,
      url = _event$data.url;
    if (keepaliveTimer) {
      clearInterval(keepaliveTimer);
      keepaliveTimer = undefined;
    }
    keepaliveTimer = setInterval(/*#__PURE__*/(0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee2() {
      var res, statusCode, _err$headers, _err$headers2, headers, error, _t;
      return _regenerator.default.wrap(function (_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            if (!(keepAliveRetryCount < retryCountThreshold)) {
              _context2.next = 4;
              break;
            }
            _context2.prev = 1;
            _context2.next = 2;
            return postKeepAlive(accessToken, deviceUrl, url);
          case 2:
            res = _context2.sent;
            statusCode = res.status;
            if (keepAliveRetryCount > 0) {
              postMessage({
                type: _types.WorkerMessageType.KEEPALIVE_SUCCESS,
                statusCode: statusCode
              });
            }
            keepAliveRetryCount = 0;
            _context2.next = 4;
            break;
          case 3:
            _context2.prev = 3;
            _t = _context2["catch"](1);
            headers = {};
            if ((_err$headers = _t.headers) !== null && _err$headers !== void 0 && _err$headers.has('Retry-After')) {
              headers['retry-after'] = _t.headers.get('Retry-After');
            }
            if ((_err$headers2 = _t.headers) !== null && _err$headers2 !== void 0 && _err$headers2.has('Trackingid')) {
              // eslint-disable-next-line dot-notation
              headers['trackingid'] = _t.headers.get('Trackingid');
            }
            error = {
              headers: headers,
              statusCode: _t.status,
              statusText: _t.statusText,
              type: _t.type
            };
            keepAliveRetryCount += 1;
            postMessage({
              type: _types.WorkerMessageType.KEEPALIVE_FAILURE,
              err: error,
              keepAliveRetryCount: keepAliveRetryCount
            });
          case 4:
          case "end":
            return _context2.stop();
        }
      }, _callee2, null, [[1, 3]]);
    })), interval * 1000);
  }
  if (type === _types.WorkerMessageType.CLEAR_KEEPALIVE) {
    if (keepaliveTimer) {
      clearInterval(keepaliveTimer);
      keepaliveTimer = undefined;
    }
  }
};

// eslint-disable-next-line no-restricted-globals
self.addEventListener('message', messageHandler);
var _default = exports.default = messageHandler;
