// @webex/ladon-ts/build/module/lib/pipeline.js
import { WorkerGenerator as w } from "./worker-generator.js";
import { logger as s } from "./logger.js";
import { BlurRenderer as F } from "./renderer/blur-renderer.js";
import { PassthroughRenderer as k } from "./renderer/passthrough-renderer.js";
import { ReplacementRenderer as C } from "./renderer/replacement-renderer.js";
import { resizeImage as y } from "./util.js";
class D {
  /**
   * Create a new pipeline instance.
   *
   * @param inputCanvas - Canvas element for input frame processing.
   * @param outputCanvas - Canvas element where the final rendered output will be displayed.
   * @param config - Pipeline configuration containing processing and rendering settings.
   * @param inputVideoElement - Video element serving as the input source.
   * @throws Error if the input canvas 2D context cannot be obtained.
   */
  constructor(e, t, r, a) {
    this.inputCanvas = e, this.outputCanvas = t, this.config = r, this.inputVideoElement = a, this.loaded = !1, this.isProcessingFrame = !1, this.pendingFramePromise = null;
    const n = this.inputCanvas.getContext("2d", {
      alpha: !1,
      willReadFrequently: !0
    });
    n || s.errorAndThrow("Error getting input context 2d"), this.inputCanvasContext = n;
  }
  /**
   * Process a single frame and return the inference result.
   * This method handles frame data, performs inference if needed, and renders the output.
   * Includes frame processing serialization to prevent overlapping operations.
   *
   * @param timestamp - The timestamp of the frame in milliseconds.
   * @param performInference - Whether to perform mask inference or reuse the last result.
   * @param frameData - Optional frame data to process instead of capturing from input.
   * @returns Promise resolving to the frame processing result with timing statistics.
   * @throws Error if the pipeline (generator or renderer) is not loaded.
   */
  async onFrame(e, t = !0, r) {
    this.isProcessingFrame && this.pendingFramePromise && await this.pendingFramePromise, this.isProcessingFrame = !0;
    try {
      return this.pendingFramePromise = this.processFrameInternal(
        e,
        t,
        r
      ), await this.pendingFramePromise;
    } finally {
      this.isProcessingFrame = !1, this.pendingFramePromise = null;
    }
  }
  /**
   * Internal frame processing implementation.
   * Handles the actual frame processing logic including timing measurements.
   *
   * @param timestamp - The timestamp of the frame in milliseconds.
   * @param performInference - Whether to perform mask inference or reuse the last result.
   * @param frameData - Optional frame data to process instead of capturing from input.
   * @returns Promise resolving to the frame processing result with timing statistics.
   * @throws Error if the pipeline (generator or renderer) is not loaded.
   */
  async processFrameInternal(e, t = !0, r) {
    const a = performance.now();
    (!this.generator || !this.renderer) && s.errorAndThrow("Pipeline is not loaded.");
    let n = 0, o = 0, i;
    if (t || !this.lastMaskResult) {
      const g = performance.now();
      r ?? (r = this.getInputFrameData()), i = this.createFrameDataCopy(r), n = performance.now() - g;
      const c = performance.now(), p = r, f = this.getMaskImageData(p);
      await this.generator.process(f, e), this.lastMaskResult = await this.generator.getLadonOutputForTimestamp(e), o = performance.now() - c;
    } else
      i = this.inputVideoElement;
    const l = performance.now();
    await this.renderer.render(i, this.lastMaskResult);
    const m = performance.now() - l, h = performance.now(), u = h - a, d = this.createInferenceResult(e, t);
    return d.stats = {
      totalFrameTime: u,
      modelInferenceTime: o,
      preprocessingTime: n,
      postprocessingTime: m,
      frameStartTime: a,
      frameEndTime: h,
      inferenceSkipped: !t
    }, d;
  }
  /**
   * Get the raw image data from the input canvas context.
   *
   * @returns The frame data as a Uint8ClampedArray.
   */
  getInputFrameData() {
    return this.inputCanvasContext.getImageData(
      0,
      0,
      this.config.input.width,
      this.config.input.height
    ).data;
  }
  /**
   * Create a copy of frame data to prevent modifications during async operations.
   */
  createFrameDataCopy(e) {
    var t;
    return e instanceof Uint8ClampedArray ? (((t = this.frameDataBuffer) == null ? void 0 : t.length) !== e.length && (this.frameDataBuffer = new Uint8ClampedArray(e.length)), this.frameDataBuffer.set(e), this.frameDataBuffer) : e;
  }
  /**
   * Resize the image data if the model expects a different input size.
   * Handles scaling between input dimensions and model requirements.
   *
   * @param rawImageData - The raw image data to potentially resize.
   * @returns The processed mask image data, resized if necessary.
   */
  getMaskImageData(e) {
    const t = this.config.mask.inputSize, r = this.config.input;
    return t.width === r.width && t.height === r.height ? e : y(
      e,
      r.width,
      r.height,
      t.width,
      t.height
    );
  }
  /**
   * Create the inference result object with appropriate data.
   * Handles both successful inference results and fallback cases.
   *
   * @param timestamp - The timestamp of the frame in milliseconds.
   * @param performedInference - Whether inference was actually performed for this frame.
   * @returns The inference result object containing mask data and metadata.
   */
  createInferenceResult(e, t) {
    return this.lastMaskResult ? {
      facesAndLandmarks: this.lastMaskResult.facesAndLandmarks,
      gesture: this.lastMaskResult.gesture,
      motion: this.lastMaskResult.motion,
      skipped: !t,
      timestamp: e
      // Stats will be added by the caller if needed
    } : {
      facesAndLandmarks: { data: new Float32Array(0), dims: [] },
      gesture: new Float32Array(0),
      motion: 0,
      skipped: !0,
      timestamp: e
      // Stats will be added by the caller if needed
    };
  }
  /**
   * Check if the pipeline is fully loaded and ready for processing.
   *
   * @returns True if the pipeline is loaded and ready, false otherwise.
   */
  isLoaded() {
    return this.loaded;
  }
  /**
   * Load and initialize the pipeline components.
   * Creates the appropriate renderer and generator based on configuration.
   *
   * @returns Promise that resolves when all components are loaded and ready.
   * @throws Error if an unsupported renderer type is specified in the configuration.
   */
  async load() {
    switch (s.log("Creating renderer"), this.config.render.type) {
      case "blur":
        this.renderer = new F(this.outputCanvas, this.config);
        break;
      case "passthrough":
        this.renderer = new k(this.outputCanvas, this.config);
        break;
      case "replacement":
        this.renderer = new C(this.outputCanvas, this.config);
        break;
      default:
        s.errorAndThrow("Renderer not implemented.");
    }
    s.log("Loading renderer"), await this.renderer.load(), s.log("Creating generator"), this.generator = new w(this.config), s.log("Loading generator"), await this.generator.load(), this.loaded = !0;
  }
  /**
   * Reset the pipeline state by clearing cached mask results.
   * Useful when switching between different processing modes or sources.
   */
  reset() {
    this.lastMaskResult = void 0;
  }
  /**
   * Destroy the pipeline and clean up all resources.
   * Properly disposes of the renderer and generator to prevent memory leaks.
   *
   * @returns Promise that resolves when cleanup is complete.
   */
  async destroy() {
    var e, t;
    this.reset(), (e = this.renderer) == null || e.destroy(), await ((t = this.generator) == null ? void 0 : t.destroy()), this.loaded = !1;
  }
}
export {
  D as Pipeline
};
