// @webex/internal-plugin-metrics/dist/behavioral-metrics.js
"use strict";

var _Reflect$construct = require("@babel/runtime-corejs2/core-js/reflect/construct");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));
var _genericMetrics = _interopRequireDefault(require("./generic-metrics"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
/**
 * @description Util class to handle Behavioral Metrics
 * @export
 * @class BehavioralMetrics
 */
var BehavioralMetrics = exports.default = /*#__PURE__*/function (_GenericMetrics) {
  function BehavioralMetrics() {
    (0, _classCallCheck2.default)(this, BehavioralMetrics);
    return _callSuper(this, BehavioralMetrics, arguments);
  }
  (0, _inherits2.default)(BehavioralMetrics, _GenericMetrics);
  return (0, _createClass2.default)(BehavioralMetrics, [{
    key: "submitBehavioralEvent",
    value:
    /**
     * Submit a behavioral metric to our metrics endpoint.
     * @param {MetricEventProduct} product the product from which the metric is being submitted, e.g. 'webex' web client, 'wxcc_desktop'
     * @param {MetricEventAgent} agent the source of the action for this metric
     * @param {string} target the 'thing' that this metric includes information about
     * @param {MetricEventVerb} verb the action that this metric includes information about
     * @param {EventPayload} payload information specific to this event. This should be flat, i.e. it should not include nested objects.
     * @returns {Promise<any>}
     */
    function submitBehavioralEvent(_ref) {
      var product = _ref.product,
        agent = _ref.agent,
        target = _ref.target,
        verb = _ref.verb,
        payload = _ref.payload;
      var name = "".concat(product, ".").concat(agent, ".").concat(target, ".").concat(verb);
      var event = this.createTaggedEventObject({
        type: ['behavioral'],
        name: name,
        payload: payload
      });
      this.submitEvent({
        kind: 'behavioral-events -> ',
        name: name,
        event: event
      });
    }
  }]);
}(_genericMetrics.default);
