// @webex/internal-plugin-metrics/dist/index.js
"use strict";

var _typeof = require("@babel/runtime-corejs2/helpers/typeof");
var _WeakMap = require("@babel/runtime-corejs2/core-js/weak-map");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
_Object$defineProperty(exports, "BehavioralMetrics", {
  enumerable: true,
  get: function get() {
    return _behavioralMetrics.default;
  }
});
_Object$defineProperty(exports, "BusinessMetrics", {
  enumerable: true,
  get: function get() {
    return _businessMetrics.default;
  }
});
exports.CALL_DIAGNOSTIC_CONFIG = void 0;
_Object$defineProperty(exports, "CallDiagnosticLatencies", {
  enumerable: true,
  get: function get() {
    return _callDiagnosticMetricsLatencies.default;
  }
});
_Object$defineProperty(exports, "CallDiagnosticMetrics", {
  enumerable: true,
  get: function get() {
    return _callDiagnosticMetrics2.default;
  }
});
exports.CallDiagnosticUtils = void 0;
_Object$defineProperty(exports, "NewMetrics", {
  enumerable: true,
  get: function get() {
    return _newMetrics.default;
  }
});
_Object$defineProperty(exports, "OperationalMetrics", {
  enumerable: true,
  get: function get() {
    return _operationalMetrics.default;
  }
});
_Object$defineProperty(exports, "PreLoginMetrics", {
  enumerable: true,
  get: function get() {
    return _preloginMetrics.default;
  }
});
_Object$defineProperty(exports, "RtcMetrics", {
  enumerable: true,
  get: function get() {
    return _rtcMetrics.default;
  }
});
exports.Utils = void 0;
_Object$defineProperty(exports, "config", {
  enumerable: true,
  get: function get() {
    return _config.default;
  }
});
_Object$defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _metrics.default;
  }
});
_Object$defineProperty(exports, "getOSNameInternal", {
  enumerable: true,
  get: function get() {
    return _metrics.getOSNameInternal;
  }
});
var _webexCore = require("@webex/webex-core");
var _metrics = _interopRequireWildcard(require("./metrics"));
var _config = _interopRequireDefault(require("./config"));
var _newMetrics = _interopRequireDefault(require("./new-metrics"));
var Utils = _interopRequireWildcard(require("./utils"));
exports.Utils = Utils;
var CALL_DIAGNOSTIC_CONFIG = _interopRequireWildcard(require("./call-diagnostic/config"));
exports.CALL_DIAGNOSTIC_CONFIG = CALL_DIAGNOSTIC_CONFIG;
var CallDiagnosticUtils = _interopRequireWildcard(require("./call-diagnostic/call-diagnostic-metrics.util"));
exports.CallDiagnosticUtils = CallDiagnosticUtils;
var _callDiagnosticMetrics2 = _interopRequireDefault(require("./call-diagnostic/call-diagnostic-metrics"));
var _callDiagnosticMetricsLatencies = _interopRequireDefault(require("./call-diagnostic/call-diagnostic-metrics-latencies"));
var _behavioralMetrics = _interopRequireDefault(require("./behavioral-metrics"));
var _operationalMetrics = _interopRequireDefault(require("./operational-metrics"));
var _businessMetrics = _interopRequireDefault(require("./business-metrics"));
var _rtcMetrics = _interopRequireDefault(require("./rtcMetrics"));
var _preloginMetrics = _interopRequireDefault(require("./prelogin-metrics"));
function _interopRequireWildcard(e, t) { if ("function" == typeof _WeakMap) var r = new _WeakMap(), n = new _WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = _Object$defineProperty) && _Object$getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
/*!
 * Copyright (c) 2015-2020 Cisco Systems, Inc. See LICENSE file.
 */

(0, _webexCore.registerInternalPlugin)('metrics', _metrics.default, {
  config: _config.default
});
(0, _webexCore.registerInternalPlugin)('newMetrics', _newMetrics.default, {
  config: _config.default
});
