// @webex/internal-plugin-metrics/dist/prelogin-metrics.js
"use strict";

var _Reflect$construct = require("@babel/runtime-corejs2/core-js/reflect/construct");
var _Object$keys = require("@babel/runtime-corejs2/core-js/object/keys");
var _Object$getOwnPropertySymbols = require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols");
var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor");
var _Object$getOwnPropertyDescriptors = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptors");
var _Object$defineProperties = require("@babel/runtime-corejs2/core-js/object/define-properties");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));
var _genericMetrics = _interopRequireDefault(require("./generic-metrics"));
function ownKeys(e, r) { var t = _Object$keys(e); if (_Object$getOwnPropertySymbols) { var o = _Object$getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return _Object$getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(e, _Object$getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { _Object$defineProperty(e, r, _Object$getOwnPropertyDescriptor(t, r)); }); } return e; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
/**
 * @description Util class to handle PreLogin Metrics
 * @export
 * @class PreLoginMetrics
 */
var PreLoginMetrics = exports.default = /*#__PURE__*/function (_GenericMetrics) {
  /**
   * Constructor
   * @param {PreLoginMetricsBatcher} preLoginMetricsBatcher - Pre-login metrics batcher
   * @param {any} attrs - Attributes
   * @param {any} options - Options
   * @constructor
   */
  function PreLoginMetrics(preLoginMetricsBatcher) {
    var _this;
    var attrs = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
    (0, _classCallCheck2.default)(this, PreLoginMetrics);
    _this = _callSuper(this, PreLoginMetrics, [attrs, options]);
    (0, _defineProperty2.default)(_this, "preLoginMetricsBatcher", void 0);
    _this.preLoginMetricsBatcher = preLoginMetricsBatcher;
    return _this;
  }

  /**
   * Submit a business metric to our metrics endpoint.
   * Routes to the correct table with the correct schema payload by table.
   * @see https://confluence-eng-gpk2.cisco.com/conf/display/WAP/Business+metrics++-%3E+ROMA
   * @param {Object} options - The options object
   * @param {string} options.name - Name of the metric
   * @param {string} options.preLoginId - ID to identify pre-login user
   * @param {EventPayload} options.payload - User payload of the metric
   * @param {EventPayload} [options.metadata] - Optional metadata to include outside of eventPayload.value
   * @returns {Promise<void>} Promise that resolves when the metric is submitted
   */
  (0, _inherits2.default)(PreLoginMetrics, _GenericMetrics);
  return (0, _createClass2.default)(PreLoginMetrics, [{
    key: "submitPreLoginEvent",
    value: function submitPreLoginEvent(_ref) {
      var name = _ref.name,
        preLoginId = _ref.preLoginId,
        payload = _ref.payload,
        metadata = _ref.metadata;
      if (!metadata) {
        metadata = {};
      }
      if (!metadata.appType) {
        metadata.appType = 'Web Client';
      }
      var finalEvent = this.buildEvent(name, preLoginId, payload, metadata);
      this.preLoginMetricsBatcher.savePreLoginId(preLoginId);
      return this.preLoginMetricsBatcher.request(finalEvent);
    }

    /**
     * Builds a formatted event object for metrics submission.
     * @param {string} metricName - Metric name
     * @param {string} preLoginId - Pre-login user identifier
     * @param {EventPayload} payload - Metric payload data
     * @param {EventPayload} metadata - Additional metadata to include in the event
     * @returns {object} Formatted metrics event object with type, eventPayload, and timestamp
     */
  }, {
    key: "buildEvent",
    value: function buildEvent(metricName, preLoginId, payload, metadata) {
      return {
        type: ['business'],
        eventPayload: {
          metricName: metricName,
          browserDetails: this.getBrowserDetails(),
          context: this.getContext(),
          timestamp: new Date().getTime(),
          value: _objectSpread(_objectSpread({
            preLoginId: preLoginId
          }, metadata), payload)
        }
      };
    }
  }]);
}(_genericMetrics.default);
