// @webex/calling/dist/Metrics/types.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.VOICEMAIL_ACTION = exports.UPLOAD_LOGS_ACTION = exports.TRANSFER_ACTION = exports.REG_ACTION = exports.MOBIUS_SERVER_ACTION = exports.METRIC_TYPE = exports.METRIC_EVENT = exports.CONNECTION_ACTION = void 0;
var METRIC_TYPE = exports.METRIC_TYPE = /*#__PURE__*/function (METRIC_TYPE) {
  METRIC_TYPE["OPERATIONAL"] = "operational";
  METRIC_TYPE["BEHAVIORAL"] = "behavioral";
  return METRIC_TYPE;
}({});
var METRIC_EVENT = exports.METRIC_EVENT = /*#__PURE__*/function (METRIC_EVENT) {
  METRIC_EVENT["BNR_ENABLED"] = "web-calling-sdk-bnr-enabled";
  METRIC_EVENT["BNR_DISABLED"] = "web-calling-sdk-bnr-disabled";
  METRIC_EVENT["CALL"] = "web-calling-sdk-callcontrol";
  METRIC_EVENT["CALL_ERROR"] = "web-calling-sdk-callcontrol-error";
  METRIC_EVENT["CONNECTION_ERROR"] = "web-calling-sdk-connection";
  METRIC_EVENT["MEDIA"] = "web-calling-sdk-media";
  METRIC_EVENT["MEDIA_ERROR"] = "web-calling-sdk-media-error";
  METRIC_EVENT["REGISTRATION"] = "web-calling-sdk-registration";
  METRIC_EVENT["REGISTRATION_ERROR"] = "web-calling-sdk-registration-error";
  METRIC_EVENT["KEEPALIVE_ERROR"] = "web-calling-sdk-keepalive-error";
  METRIC_EVENT["VOICEMAIL"] = "web-calling-sdk-voicemail";
  METRIC_EVENT["VOICEMAIL_ERROR"] = "web-calling-sdk-voicemail-error";
  METRIC_EVENT["UPLOAD_LOGS_SUCCESS"] = "web-calling-sdk-upload-logs-success";
  METRIC_EVENT["UPLOAD_LOGS_FAILED"] = "web-calling-sdk-upload-logs-failed";
  METRIC_EVENT["MOBIUS_DISCOVERY"] = "web-calling-sdk-mobius-discovery";
  return METRIC_EVENT;
}({});
var MOBIUS_SERVER_ACTION = exports.MOBIUS_SERVER_ACTION = /*#__PURE__*/function (MOBIUS_SERVER_ACTION) {
  MOBIUS_SERVER_ACTION["REGION_INFO"] = "region-info";
  MOBIUS_SERVER_ACTION["MOBIUS_SERVERS"] = "mobius-servers";
  return MOBIUS_SERVER_ACTION;
}({});
var REG_ACTION = exports.REG_ACTION = /*#__PURE__*/function (REG_ACTION) {
  REG_ACTION["REGISTER"] = "register";
  REG_ACTION["DEREGISTER"] = "deregister";
  REG_ACTION["KEEPALIVE_FAILURE"] = "keepaliveFailure";
  return REG_ACTION;
}({});
var TRANSFER_ACTION = exports.TRANSFER_ACTION = /*#__PURE__*/function (TRANSFER_ACTION) {
  TRANSFER_ACTION["BLIND"] = "TRANSFER_BLIND";
  TRANSFER_ACTION["CONSULT"] = "TRANSFER_CONSULT";
  return TRANSFER_ACTION;
}({});
var VOICEMAIL_ACTION = exports.VOICEMAIL_ACTION = /*#__PURE__*/function (VOICEMAIL_ACTION) {
  VOICEMAIL_ACTION["GET_VOICEMAILS"] = "get_voicemails";
  VOICEMAIL_ACTION["GET_VOICEMAIL_CONTENT"] = "get_voicemail_content";
  VOICEMAIL_ACTION["GET_VOICEMAIL_SUMMARY"] = "get_voicemail_summary";
  VOICEMAIL_ACTION["MARK_READ"] = "mark_read";
  VOICEMAIL_ACTION["MARK_UNREAD"] = "mark_unread";
  VOICEMAIL_ACTION["DELETE"] = "delete";
  VOICEMAIL_ACTION["TRANSCRIPT"] = "transcript";
  return VOICEMAIL_ACTION;
}({});
var UPLOAD_LOGS_ACTION = exports.UPLOAD_LOGS_ACTION = 'upload_logs';
var CONNECTION_ACTION = exports.CONNECTION_ACTION = /*#__PURE__*/function (CONNECTION_ACTION) {
  CONNECTION_ACTION["NETWORK_FLAP"] = "network_flap";
  CONNECTION_ACTION["MERCURY_DOWN"] = "mercury_down";
  CONNECTION_ACTION["MERCURY_UP"] = "mercury_up";
  return CONNECTION_ACTION;
}({});
