// @webex/calling/dist/CallingClient/windowsChromiumIceWarmupUtils.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = windowsChromiumIceWarmup;
var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs2/regenerator"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/toConsumableArray"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/asyncToGenerator"));
var _promise = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/promise"));
// This file contains a workaround for a known issue in Windows Chromium-based browsers
// See: https://issues.chromium.org/issues/332933530
// There is a rare condition where WebRTC doesn't work on the first try due to ICE gathering issues.
// This function attempts to "warm up" the ICE gathering process by creating a pair of peer connections
// and forcing ICE candidate gathering and exchange before the actual call setup.
// We need to try establishing the connection through the srflx candidates since the host candidate won't work
// This is intended to be called once at the start of the calling SDK to reduce the likelihood of ICE issues later.
function waitForIceComplete(pc, timeoutMs) {
  return new _promise.default(function (resolve) {
    if (!pc) {
      resolve({
        ok: false,
        reason: 'no-pc'
      });
      return;
    }
    if (pc.iceGatheringState === 'complete') {
      resolve({
        ok: true,
        reached: 'already'
      });
      return;
    }
    var done = false;
    var _onChange = function onChange() {
      if (pc.iceGatheringState === 'complete' && !done) {
        done = true;
        pc.removeEventListener('icegatheringstatechange', _onChange);
        resolve({
          ok: true,
          reached: 'event'
        });
      }
    };
    pc.addEventListener('icegatheringstatechange', _onChange);

    // timeout fallback
    setTimeout(function () {
      if (done) return;
      done = true;
      pc.removeEventListener('icegatheringstatechange', _onChange);
      resolve({
        ok: pc.iceGatheringState === 'complete',
        reached: 'timeout'
      });
    }, timeoutMs);
  });
}
function windowsChromiumIceWarmup(_x) {
  return _windowsChromiumIceWarmup.apply(this, arguments);
}
function _windowsChromiumIceWarmup() {
  _windowsChromiumIceWarmup = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee(_ref) {
    var _ref$iceServers, iceServers, _ref$timeoutMs, timeoutMs, pc1, pc2, candidates, offer, answer;
    return _regenerator.default.wrap(function (_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          _ref$iceServers = _ref.iceServers, iceServers = _ref$iceServers === void 0 ? [] : _ref$iceServers, _ref$timeoutMs = _ref.timeoutMs, timeoutMs = _ref$timeoutMs === void 0 ? 1000 : _ref$timeoutMs;
          pc1 = new RTCPeerConnection({
            iceServers: iceServers,
            iceCandidatePoolSize: 1
          });
          pc2 = new RTCPeerConnection({
            iceServers: iceServers,
            iceCandidatePoolSize: 1
          });
          candidates = {
            pc1: [],
            pc2: []
          };
          pc1.onicecandidate = function (e) {
            if (e.candidate && e.candidate.type !== 'host') {
              candidates.pc1.push(e.candidate);
            }
          };
          _context.next = 1;
          return pc1.createDataChannel('warmup');
        case 1:
          _context.next = 2;
          return pc1.addTransceiver('audio');
        case 2:
          _context.next = 3;
          return pc1.createOffer();
        case 3:
          offer = _context.sent;
          _context.next = 4;
          return pc1.setLocalDescription(offer);
        case 4:
          _context.next = 5;
          return waitForIceComplete(pc1, timeoutMs);
        case 5:
          _context.next = 6;
          return pc2.setRemoteDescription(offer);
        case 6:
          pc2.onicecandidate = function (e) {
            if (e.candidate && e.candidate.type !== 'host') {
              candidates.pc2.push(e.candidate);
            }
          };
          _context.next = 7;
          return pc2.createAnswer();
        case 7:
          answer = _context.sent;
          _context.next = 8;
          return pc2.setLocalDescription(answer);
        case 8:
          _context.next = 9;
          return pc1.setRemoteDescription(answer);
        case 9:
          _context.next = 10;
          return waitForIceComplete(pc2, timeoutMs);
        case 10:
          _context.next = 11;
          return _promise.default.all([].concat((0, _toConsumableArray2.default)(candidates.pc1.map(function (candidate) {
            return pc2.addIceCandidate(candidate).catch(console.error);
          })), (0, _toConsumableArray2.default)(candidates.pc2.map(function (candidate) {
            return pc1.addIceCandidate(candidate).catch(console.error);
          }))));
        case 11:
          pc1.close();
          pc2.close();
        case 12:
        case "end":
          return _context.stop();
      }
    }, _callee);
  }));
  return _windowsChromiumIceWarmup.apply(this, arguments);
}
