// @webex/webex-core/dist/interceptors/request-timing.js
"use strict";

var _Reflect$construct = require("@babel/runtime-corejs2/core-js/reflect/construct");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _now = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/date/now"));
var _promise = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/promise"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));
var _httpCore = require("@webex/http-core");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /*!
 * Copyright (c) 2015-2020 Cisco Systems, Inc. See LICENSE file.
 */
/**
 * @class
 */
var RequestTimingInterceptor = exports.default = /*#__PURE__*/function (_Interceptor) {
  function RequestTimingInterceptor() {
    (0, _classCallCheck2.default)(this, RequestTimingInterceptor);
    return _callSuper(this, RequestTimingInterceptor, arguments);
  }
  (0, _inherits2.default)(RequestTimingInterceptor, _Interceptor);
  return (0, _createClass2.default)(RequestTimingInterceptor, [{
    key: "onRequest",
    value:
    /**
     * @see Interceptor#onRequest
     * @param {Object} options
     * @returns {Object}
     */
    function onRequest(options) {
      options.$timings = options.$timings || {};
      options.$timings.requestStart = (0, _now.default)();
      return options;
    }

    /**
     * @see Interceptor#onRequestError
     * @param {Object} options
     * @returns {Object}
     */
  }, {
    key: "onRequestError",
    value: function onRequestError(options) {
      options.$timings.requestEnd = options.$timings.requestFail = (0, _now.default)();
      return _promise.default.reject(options);
    }

    /**
     * @see Interceptor#onResponse
     * @param {Object} options
     * @param {HttpResponse} response
     * @returns {Object}
     */
  }, {
    key: "onResponse",
    value: function onResponse(options, response) {
      options.$timings.requestEnd = (0, _now.default)();
      return _promise.default.resolve(response);
    }

    /**
     * @see Interceptor#onResponseError
     * @param {Object} options
     * @param {Error} reason
     * @returns {Object}
     */
  }, {
    key: "onResponseError",
    value: function onResponseError(options, reason) {
      options.$timings.requestEnd = options.$timings.requestFail = (0, _now.default)();
      return _promise.default.reject(reason);
    }
  }], [{
    key: "create",
    value:
    /**
     * @param {Object} options
     * @returns {RequestTimingInterceptor}
     */
    function create(options) {
      return new RequestTimingInterceptor(this, options);
    }
  }]);
}(_httpCore.Interceptor);
