// @webex/internal-plugin-metrics/dist/call-diagnostic/call-diagnostic-metrics-latencies.js
"use strict";

var _Reflect$construct = require("@babel/runtime-corejs2/core-js/reflect/construct");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _map = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/map"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));
var _webexCore = require("@webex/webex-core");
var _lodash = require("lodash");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /* eslint-disable class-methods-use-this */ /* eslint-disable valid-jsdoc */
// we only care about client event and feature event for now
/**
 * @description Helper class to store latencies timestamp and to calculate various latencies for CA.
 * @exports
 * @class CallDiagnosticLatencies
 */
var CallDiagnosticLatencies = exports.default = /*#__PURE__*/function (_WebexPlugin) {
  /**
   * @constructor
   */
  function CallDiagnosticLatencies() {
    var _this;
    (0, _classCallCheck2.default)(this, CallDiagnosticLatencies);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _callSuper(this, CallDiagnosticLatencies, [].concat(args));
    (0, _defineProperty2.default)(_this, "latencyTimestamps", void 0);
    (0, _defineProperty2.default)(_this, "precomputedLatencies", void 0);
    // meetingId that the current latencies are for
    (0, _defineProperty2.default)(_this, "meetingId", void 0);
    (0, _defineProperty2.default)(_this, "MAX_INTEGER", 2147483647);
    _this.latencyTimestamps = new _map.default();
    _this.precomputedLatencies = new _map.default();
    return _this;
  }

  /**
   * Clear timestamps
   */
  (0, _inherits2.default)(CallDiagnosticLatencies, _WebexPlugin);
  return (0, _createClass2.default)(CallDiagnosticLatencies, [{
    key: "clearTimestamps",
    value: function clearTimestamps() {
      this.latencyTimestamps.clear();
      this.precomputedLatencies.clear();
    }

    /**
     * Associate current latencies with a meeting id
     * @param meetingId
     */
  }, {
    key: "setMeetingId",
    value: function setMeetingId(meetingId) {
      this.meetingId = meetingId;
    }

    /**
     * Returns the meeting object associated with current latencies
     * @returns meeting object
     */
  }, {
    key: "getMeeting",
    value: function getMeeting() {
      if (this.meetingId) {
        // @ts-ignore
        return this.webex.meetings.getBasicMeetingInformation(this.meetingId);
      }
      return undefined;
    }

    /**
     * Store timestamp value
     * @param key - key
     * @param value - value
     * @param options - store options
     * @throws
     * @returns
     */
  }, {
    key: "saveTimestamp",
    value: function saveTimestamp(_ref) {
      var key = _ref.key,
        _ref$value = _ref.value,
        value = _ref$value === void 0 ? new Date().getTime() : _ref$value,
        _ref$options = _ref.options,
        options = _ref$options === void 0 ? {} : _ref$options;
      // save the meetingId so we can use the meeting object in latency calculations if needed
      var meetingId = options.meetingId;
      if (meetingId) {
        this.setMeetingId(meetingId);
      }
      // for some events we're only interested in the first timestamp not last
      // as these events can happen multiple times
      if (key === 'client.media.rx.start' || key === 'client.media.tx.start' || key === 'internal.client.meetinginfo.request' || key === 'internal.client.meetinginfo.response' || key === 'client.media-engine.remote-sdp-received') {
        this.saveFirstTimestampOnly(key, value);
      } else {
        this.latencyTimestamps.set(key, value);
        // new offer/answer so reset the remote SDP timestamp
        if (key === 'client.media-engine.local-sdp-generated') {
          this.latencyTimestamps.delete('client.media-engine.remote-sdp-received');
        }
      }
    }

    /**
     * Store precomputed latency value
     * @param key - key
     * @param value - value
     * @param accumulate - when it is true, it overwrites existing value with sum of the current value and the new measurement otherwise just store the new measurement
     * @throws
     * @returns
     */
  }, {
    key: "saveLatency",
    value: function saveLatency(key, value) {
      var accumulate = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      var existingValue = accumulate ? this.precomputedLatencies.get(key) || 0 : 0;
      this.precomputedLatencies.set(key, value + existingValue);
    }

    /**
     * Measure latency for a request
     * @param callback - callback for which you would like to measure latency
     * @param key - key
     * @param accumulate - when it is true, it overwrites existing value with sum of the current value and the new measurement otherwise just store the new measurement
     * @returns
     */
  }, {
    key: "measureLatency",
    value: function measureLatency(callback, key) {
      var _this2 = this;
      var accumulate = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
      var start = performance.now();
      return callback().finally(function () {
        _this2.saveLatency(key, performance.now() - start, accumulate);
      });
    }

    /**
     * Store only the first timestamp value for the given key
     * @param key - key
     * @param  value -value
     * @throws
     * @returns
     */
  }, {
    key: "saveFirstTimestampOnly",
    value: function saveFirstTimestampOnly(key) {
      var value = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : new Date().getTime();
      if (this.latencyTimestamps.has(key)) {
        return;
      }
      this.latencyTimestamps.set(key, value);
    }

    /**
     * Helper to calculate end - start
     * @param a start
     * @param b end
     * @returns latency
     */
  }, {
    key: "getDiffBetweenTimestamps",
    value: function getDiffBetweenTimestamps(a, b, clampValues) {
      var start = this.latencyTimestamps.get(a);
      var end = this.latencyTimestamps.get(b);
      if (typeof start !== 'number' || typeof end !== 'number') {
        return undefined;
      }
      var diff = end - start;
      var _ref2 = clampValues || {},
        _ref2$minimum = _ref2.minimum,
        minimum = _ref2$minimum === void 0 ? 0 : _ref2$minimum,
        _ref2$maximum = _ref2.maximum,
        maximum = _ref2$maximum === void 0 ? this.MAX_INTEGER : _ref2$maximum;
      return (0, _lodash.clamp)(diff, minimum, maximum);
    }

    /**
     * Meeting Info Request
     * @note Meeting Info request happen not just in the join phase. CA requires
     * metrics around meeting info request that are only part of join phase.
     * This internal.* event is used to track the real timestamps
     * (when the actual request/response happen). This is because the actual CA event is
     * sent inside the join method on the meeting object based on some logic, but that's not exactly when
     * those events are actually fired. The logic only confirms that they have happened, and we send them over.
     * @returns - latency
     */
  }, {
    key: "getMeetingInfoReqResp",
    value: function getMeetingInfoReqResp() {
      return this.getDiffBetweenTimestamps('internal.client.meetinginfo.request', 'internal.client.meetinginfo.response', {
        maximum: 1200000
      });
    }

    /**
     * Interstitial Time
     * @returns - latency
     */
  }, {
    key: "getShowInterstitialTime",
    value: function getShowInterstitialTime() {
      return this.getDiffBetweenTimestamps('client.interstitial-window.start-launch', 'internal.client.interstitial-window.click.joinbutton');
    }

    /**
     * getU2CTime
     * @returns - latency
     */
  }, {
    key: "getU2CTime",
    value: function getU2CTime() {
      var u2cLatency = this.precomputedLatencies.get('internal.get.u2c.time');
      return typeof u2cLatency === 'number' ? Math.floor(u2cLatency) : undefined;
    }

    /**
     * Device Register Time
     * @returns - latency
     */
  }, {
    key: "getRegisterWDMDeviceJMT",
    value: function getRegisterWDMDeviceJMT() {
      return this.getDiffBetweenTimestamps('internal.register.device.request', 'internal.register.device.response');
    }

    /**
     * Call Init Join Request
     * @returns - latency
     */
  }, {
    key: "getCallInitJoinReq",
    value: function getCallInitJoinReq() {
      return this.getDiffBetweenTimestamps('internal.client.interstitial-window.click.joinbutton', 'client.locus.join.request', {
        maximum: 1200000
      });
    }

    /**
     * Locus Join Request
     * @returns - latency
     */
  }, {
    key: "getJoinReqResp",
    value: function getJoinReqResp() {
      return this.getDiffBetweenTimestamps('client.locus.join.request', 'client.locus.join.response', {
        maximum: 1200000
      });
    }

    /**
     * Time taken to do turn discovery
     * @returns - latency
     */
  }, {
    key: "getTurnDiscoveryTime",
    value: function getTurnDiscoveryTime() {
      return this.getDiffBetweenTimestamps('internal.client.add-media.turn-discovery.start', 'internal.client.add-media.turn-discovery.end');
    }

    /**
     * Local SDP Generated Remote SDP REceived
     * @returns - latency
     */
  }, {
    key: "getLocalSDPGenRemoteSDPRecv",
    value: function getLocalSDPGenRemoteSDPRecv() {
      return this.getDiffBetweenTimestamps('client.media-engine.local-sdp-generated', 'client.media-engine.remote-sdp-received', {
        maximum: 1200000
      });
    }

    /**
     * ICE Setup Time
     * @returns - latency
     */
  }, {
    key: "getICESetupTime",
    value: function getICESetupTime() {
      return this.getDiffBetweenTimestamps('client.ice.start', 'client.ice.end', {
        maximum: 1200000
      });
    }

    /**
     * Audio ICE time
     * @returns - latency
     */
  }, {
    key: "getAudioICESetupTime",
    value: function getAudioICESetupTime() {
      return this.getDiffBetweenTimestamps('client.ice.start', 'client.ice.end');
    }

    /**
     * Video ICE Time
     * @returns - latency
     */
  }, {
    key: "getVideoICESetupTime",
    value: function getVideoICESetupTime() {
      return this.getDiffBetweenTimestamps('client.ice.start', 'client.ice.end');
    }

    /**
     * Share ICE Time
     * @returns - latency
     */
  }, {
    key: "getShareICESetupTime",
    value: function getShareICESetupTime() {
      return this.getDiffBetweenTimestamps('client.ice.start', 'client.ice.end');
    }

    /**
     * Stay Lobby Time
     * @returns - latency
     */
  }, {
    key: "getStayLobbyTime",
    value: function getStayLobbyTime() {
      return this.getDiffBetweenTimestamps('client.locus.join.response', 'client.lobby.exited');
    }

    /**
     * Page JMT
     * @returns - latency
     */
  }, {
    key: "getPageJMT",
    value: function getPageJMT() {
      var latency = this.precomputedLatencies.get('internal.client.pageJMT');
      return typeof latency === 'number' ? (0, _lodash.clamp)(latency, 0, this.MAX_INTEGER) : undefined;
    }

    /**
     * Download Time JMT
     * @returns - latency
     */
  }, {
    key: "getDownloadTimeJMT",
    value: function getDownloadTimeJMT() {
      var latency = this.precomputedLatencies.get('internal.download.time');
      return typeof latency === 'number' ? (0, _lodash.clamp)(latency, 0, this.MAX_INTEGER) : undefined;
    }

    /**
     * Click To Interstitial
     * @returns - latency
     */
  }, {
    key: "getClickToInterstitial",
    value: function getClickToInterstitial() {
      // for normal join (where green join button exists before interstitial, i.e reminder, space list etc)
      if (this.latencyTimestamps.get('internal.client.meeting.click.joinbutton')) {
        return this.getDiffBetweenTimestamps('internal.client.meeting.click.joinbutton', 'internal.client.meeting.interstitial-window.showed');
      }
      var clickToInterstitialLatency = this.precomputedLatencies.get('internal.click.to.interstitial');
      if (typeof clickToInterstitialLatency === 'number') {
        return (0, _lodash.clamp)(clickToInterstitialLatency, 0, this.MAX_INTEGER);
      }
      return undefined;
    }

    /**
     * Click To Interstitial With User Delay
     * @returns - latency
     */
  }, {
    key: "getClickToInterstitialWithUserDelay",
    value: function getClickToInterstitialWithUserDelay() {
      // for normal join (where green join button exists before interstitial, i.e reminder, space list etc)
      if (this.latencyTimestamps.get('internal.client.meeting.click.joinbutton')) {
        return this.getDiffBetweenTimestamps('internal.client.meeting.click.joinbutton', 'internal.client.meeting.interstitial-window.showed');
      }
      var clickToInterstitialWithUserDelayLatency = this.precomputedLatencies.get('internal.click.to.interstitial.with.user.delay');
      if (typeof clickToInterstitialWithUserDelayLatency === 'number') {
        return (0, _lodash.clamp)(clickToInterstitialWithUserDelayLatency, 0, this.MAX_INTEGER);
      }
      return undefined;
    }

    /**
     * Interstitial To Join Ok
     * @returns - latency
     */
  }, {
    key: "getInterstitialToJoinOK",
    value: function getInterstitialToJoinOK() {
      return this.getDiffBetweenTimestamps('internal.client.interstitial-window.click.joinbutton', 'client.locus.join.response');
    }

    /**
     * Call Init To MediaEngineReady
     * @returns - latency
     */
  }, {
    key: "getCallInitMediaEngineReady",
    value: function getCallInitMediaEngineReady() {
      return this.getDiffBetweenTimestamps('internal.client.interstitial-window.click.joinbutton', 'client.media-engine.ready', {
        maximum: 1200000
      });
    }

    /**
     * Interstitial To Media Ok
     * @returns - latency
     */
  }, {
    key: "getInterstitialToMediaOKJMT",
    value: function getInterstitialToMediaOKJMT() {
      var interstitialJoinClickTimestamp = this.latencyTimestamps.get('internal.client.interstitial-window.click.joinbutton');

      // get the first timestamp
      var connectedMedia = this.latencyTimestamps.get('client.ice.end');
      var lobbyTimeLatency = this.getStayLobbyTime();
      var lobbyTime = typeof lobbyTimeLatency === 'number' ? lobbyTimeLatency : 0;
      if (interstitialJoinClickTimestamp && connectedMedia) {
        var interstitialToMediaOKJmt = connectedMedia - interstitialJoinClickTimestamp - lobbyTime;
        return (0, _lodash.clamp)(interstitialToMediaOKJmt, 0, this.MAX_INTEGER);
      }
      return undefined;
    }

    /**
     * Total JMT
     * @returns - latency
     */
  }, {
    key: "getTotalJMT",
    value: function getTotalJMT() {
      var clickToInterstitial = this.getClickToInterstitial();
      var interstitialToJoinOk = this.getInterstitialToJoinOK();
      if (typeof clickToInterstitial === 'number' && typeof interstitialToJoinOk === 'number') {
        return (0, _lodash.clamp)(clickToInterstitial + interstitialToJoinOk, 0, this.MAX_INTEGER);
      }
      return undefined;
    }

    /**
     * Total JMT With User Delay
     * @returns - latency
     */
  }, {
    key: "getTotalJMTWithUserDelay",
    value: function getTotalJMTWithUserDelay() {
      var clickToInterstitialWithUserDelay = this.getClickToInterstitialWithUserDelay();
      var interstitialToJoinOk = this.getInterstitialToJoinOK();
      if (typeof clickToInterstitialWithUserDelay === 'number' && typeof interstitialToJoinOk === 'number') {
        return (0, _lodash.clamp)(clickToInterstitialWithUserDelay + interstitialToJoinOk, 0, this.MAX_INTEGER);
      }
      return undefined;
    }

    /**
     * Join Conf JMT
     * @returns - latency
     */
  }, {
    key: "getJoinConfJMT",
    value: function getJoinConfJMT() {
      var joinReqResp = this.getJoinReqResp();
      var ICESetupTime = this.getICESetupTime();
      if (joinReqResp && ICESetupTime) {
        return (0, _lodash.clamp)(joinReqResp + ICESetupTime, 0, this.MAX_INTEGER);
      }
      return undefined;
    }

    /**
     * Total Media JMT
     * @returns - latency
     */
  }, {
    key: "getTotalMediaJMT",
    value: function getTotalMediaJMT() {
      var clickToInterstitial = this.getClickToInterstitial();
      var interstitialToJoinOk = this.getInterstitialToJoinOK();
      var joinConfJMT = this.getJoinConfJMT();
      var lobbyTimeLatency = this.getStayLobbyTime();
      var lobbyTime = typeof lobbyTimeLatency === 'number' ? lobbyTimeLatency : 0;
      if (clickToInterstitial && interstitialToJoinOk && joinConfJMT) {
        var _this$getMeeting;
        var totalMediaJMT = (0, _lodash.clamp)(clickToInterstitial + interstitialToJoinOk + joinConfJMT, 0, Infinity);
        if ((_this$getMeeting = this.getMeeting()) !== null && _this$getMeeting !== void 0 && _this$getMeeting.allowMediaInLobby) {
          return (0, _lodash.clamp)(totalMediaJMT, 0, this.MAX_INTEGER);
        }
        return (0, _lodash.clamp)(totalMediaJMT - lobbyTime, 0, this.MAX_INTEGER);
      }
      return undefined;
    }

    /**
     * Total Media JMT With User Delay
     * @returns - latency
     */
  }, {
    key: "getTotalMediaJMTWithUserDelay",
    value: function getTotalMediaJMTWithUserDelay() {
      var clickToInterstitialWithUserDelay = this.getClickToInterstitialWithUserDelay();
      var interstitialToJoinOk = this.getInterstitialToJoinOK();
      var joinConfJMT = this.getJoinConfJMT();
      if (clickToInterstitialWithUserDelay && interstitialToJoinOk && joinConfJMT) {
        return (0, _lodash.clamp)(clickToInterstitialWithUserDelay + interstitialToJoinOk + joinConfJMT, 0, this.MAX_INTEGER);
      }
      return undefined;
    }

    /**
     * Client JMT
     * @returns - latency
     */
  }, {
    key: "getClientJMT",
    value: function getClientJMT() {
      var interstitialToJoinOk = this.getInterstitialToJoinOK();
      var joinConfJMT = this.getJoinConfJMT();
      if (typeof interstitialToJoinOk === 'number' && typeof joinConfJMT === 'number') {
        return (0, _lodash.clamp)(interstitialToJoinOk - joinConfJMT, 0, this.MAX_INTEGER);
      }
      return undefined;
    }

    /**
     * Audio setup delay receive
     */
  }, {
    key: "getAudioJoinRespRxStart",
    value: function getAudioJoinRespRxStart() {
      return this.getDiffBetweenTimestamps('client.locus.join.response', 'client.media.rx.start');
    }

    /**
     * Video setup delay receive
     */
  }, {
    key: "getVideoJoinRespRxStart",
    value: function getVideoJoinRespRxStart() {
      return this.getDiffBetweenTimestamps('client.locus.join.response', 'client.media.rx.start');
    }

    /**
     * Total latency for all get cluster request.
     */
  }, {
    key: "getReachabilityClustersReqResp",
    value: function getReachabilityClustersReqResp() {
      var reachablityClusterReqResp = this.precomputedLatencies.get('internal.get.cluster.time');
      return typeof reachablityClusterReqResp === 'number' ? (0, _lodash.clamp)(Math.floor(reachablityClusterReqResp), 0, this.MAX_INTEGER) : undefined;
    }

    /**
     * Audio setup delay transmit
     */
  }, {
    key: "getAudioJoinRespTxStart",
    value: function getAudioJoinRespTxStart() {
      return this.getDiffBetweenTimestamps('client.locus.join.response', 'client.media.tx.start');
    }

    /**
     * Video setup delay transmit
     */
  }, {
    key: "getVideoJoinRespTxStart",
    value: function getVideoJoinRespTxStart() {
      return this.getDiffBetweenTimestamps('client.locus.join.response', 'client.media.tx.start');
    }

    /**
     * Time from share initiation to share stop (ms).
     */
  }, {
    key: "getShareDuration",
    value: function getShareDuration() {
      return this.getDiffBetweenTimestamps('internal.client.share.initiated', 'internal.client.share.stopped');
    }

    /**
     * Total latency for all exchange ci token.
     */
  }, {
    key: "getExchangeCITokenJMT",
    value: function getExchangeCITokenJMT() {
      var exchangeCITokenJMT = this.precomputedLatencies.get('internal.exchange.ci.token.time');
      return typeof exchangeCITokenJMT === 'number' ? (0, _lodash.clamp)(Math.floor(exchangeCITokenJMT), 0, this.MAX_INTEGER) : undefined;
    }

    /**
     * Total latency for all refresh captcha requests.
     */
  }, {
    key: "getRefreshCaptchaReqResp",
    value: function getRefreshCaptchaReqResp() {
      var refreshCaptchaReqResp = this.precomputedLatencies.get('internal.refresh.captcha.time');
      return typeof refreshCaptchaReqResp === 'number' ? (0, _lodash.clamp)(Math.floor(refreshCaptchaReqResp), 0, this.MAX_INTEGER) : undefined;
    }

    /**
     * Get the latency for downloading intelligence models.
     * @returns - latency
     */
  }, {
    key: "getDownloadIntelligenceModelsReqResp",
    value: function getDownloadIntelligenceModelsReqResp() {
      var downloadIntelligenceModelsReqResp = this.precomputedLatencies.get('internal.api.fetch.intelligence.models');
      return typeof downloadIntelligenceModelsReqResp === 'number' ? (0, _lodash.clamp)(Math.floor(downloadIntelligenceModelsReqResp), 0, this.MAX_INTEGER) : undefined;
    }

    /**
     * Get the total latency for all other app API requests.
     * Excludes meeting info, because it's measured separately.
     * @returns - latency
     */
  }, {
    key: "getOtherAppApiReqResp",
    value: function getOtherAppApiReqResp() {
      var otherAppApiJMT = this.precomputedLatencies.get('internal.other.app.api.time');
      return otherAppApiJMT > 0 ? (0, _lodash.clamp)(Math.floor(otherAppApiJMT), 0, this.MAX_INTEGER) : undefined;
    }
  }]);
}(_webexCore.WebexPlugin);
