// @webex/calling/dist/Errors/catalog/LineError.js
"use strict";

var _Reflect$construct = require("@babel/runtime-corejs2/core-js/reflect/construct");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.createLineError = exports.LineError = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));
var _types = require("../../common/types");
var _ExtendedError2 = _interopRequireDefault(require("./ExtendedError"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /* eslint-disable valid-jsdoc */
/**
 * Any error reported from Line class should be stored here.
 */
var LineError = exports.LineError = /*#__PURE__*/function (_ExtendedError) {
  /**
   * Instantiate the Error class with these parameters.
   *
   * @param msg - Custom error message.
   * @param context - Error context.
   * @param type - Error Type.
   * @param status - Line Status, should be inactive.
   */
  function LineError(msg, context, type, status) {
    var _this;
    (0, _classCallCheck2.default)(this, LineError);
    _this = _callSuper(this, LineError, [msg, context, type]);
    (0, _defineProperty2.default)(_this, "status", _types.RegistrationStatus.INACTIVE);
    _this.status = status;
    return _this;
  }

  /**
   *  Class method exposed to allow storing of error object.
   *
   * @param error - Error Object.
   */
  (0, _inherits2.default)(LineError, _ExtendedError);
  return (0, _createClass2.default)(LineError, [{
    key: "setError",
    value: function setError(error) {
      this.message = error.message;
      this.context = error.context;
      this.type = error.type;
      this.status = error.status;
    }

    /**
     *  Class method exposed to retrieve error object.
     *
     * @returns Error.
     */
  }, {
    key: "getError",
    value: function getError() {
      return {
        message: this.message,
        context: this.context,
        type: this.type,
        status: this.status
      };
    }
  }]);
}(_ExtendedError2.default);
/**
 * Instantiate LineError.
 *
 * @param msg - Custom error message.
 * @param context - Error context.
 * @param type - Error Type.
 * @param status - Line Status, should be inactive.
 * @returns LineError instance.
 */
var createLineError = exports.createLineError = function createLineError(msg, context, type, status) {
  return new LineError(msg, context, type, status);
};
