// @webex/calling/dist/Events/impl/index.js
"use strict";

var _Reflect$construct = require("@babel/runtime-corejs2/core-js/reflect/construct");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.Eventing = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));
var _get2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/get"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));
var _events = _interopRequireDefault(require("events"));
var _Logger = _interopRequireDefault(require("../../Logger"));
var _types = require("../../Logger/types");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _superPropGet(t, o, e, r) { var p = (0, _get2.default)((0, _getPrototypeOf2.default)(1 & r ? t.prototype : t), o, e); return 2 & r && "function" == typeof p ? function (t) { return p.apply(e, t); } : p; } /* eslint-disable valid-jsdoc */ /* eslint-disable @typescript-eslint/no-shadow */ // eslint-disable-line import/no-extraneous-dependencies
/**
 *
 */
var Eventing = exports.Eventing = /*#__PURE__*/function (_ref) {
  function Eventing() {
    (0, _classCallCheck2.default)(this, Eventing);
    return _callSuper(this, Eventing, arguments);
  }
  (0, _inherits2.default)(Eventing, _ref);
  return (0, _createClass2.default)(Eventing, [{
    key: "emit",
    value:
    /**
     * @event
     *
     * @param event - Event that is going ot be emitted.
     * @param args - Parameters that are emitted with the event.
     */
    function emit(event) {
      var timestamp = new Date().toUTCString();
      for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }
      _Logger.default.info("".concat(timestamp, " ").concat(_types.LOG_PREFIX.EVENT, ": ").concat(event.toString(), " - event emitted with parameters -> ").concat(args), {
        file: 'Events/impl/index.ts',
        method: 'emit'
      });
      return _superPropGet(Eventing, "emit", this, 3)([event].concat(args));
    }

    /**
     * .
     * @event
     *
     * @param event - Event to listen to.
     * @param listener - Callback for event.
     */
  }, {
    key: "on",
    value: function on(event, listener) {
      return _superPropGet(Eventing, "on", this, 3)([event, listener]);
    }

    /**
     * .
     * @event
     *
     * @param event - Event to remove listener on.
     * @param listener - Callback for event.
     */
  }, {
    key: "off",
    value: function off(event, listener) {
      return _superPropGet(Eventing, "off", this, 3)([event, listener]);
    }
  }]);
}(_events.default);
