// @webex/ladon-ts/build/module/lib/renderer/replacement-renderer.js
import { logger as g } from "../logger.js";
import { compileShader as a, linkProgram as h } from "../util.js";
import { WebGLRenderer as d } from "./webgl-renderer.js";
class T extends d {
  /**
   * Create a new replacement renderer instance.
   *
   * @param outputCanvas - The canvas element where the rendered output will be displayed.
   * @param config - Pipeline configuration containing render settings and background source.
   */
  constructor(t, e) {
    super(t, e), this.outputCanvas = t, this.config = e, this.lastVideoFrameTime = 0;
    const s = this.config.render.background;
    if (s instanceof HTMLVideoElement) {
      const n = this.getVideoFrameRate(s) || 30;
      this.videoFrameInterval = 1e3 / n;
    } else
      this.videoFrameInterval = 0;
    this.fragmentShader = this.buildFragmentShader();
    const i = a(this.gl, this.gl.VERTEX_SHADER, this.vertex_shader_src), o = a(this.gl, this.gl.FRAGMENT_SHADER, this.fragmentShader);
    this.program = h(this.gl, i, o), this.gl.useProgram(this.program), this.setSharedUniforms(this.program), this.setupBackgroundTexture();
  }
  /**
   * Build the fragment shader source code for the replacement effect.
   * Creates a shader that blends the background with the input video based on the mask.
   *
   * @returns The complete fragment shader source code.
   */
  buildFragmentShader() {
    return `${this.fragment_shader_header}
    
uniform sampler2D u_background;

${this.shared_shader_functions}

void main() {
    ${this.getUVCalculation()}

    vec3 background = texture(u_background, st).rgb;
    vec3 image = texture(u_image, st).rgb;
    float mask = get_mask(st);

    vec3 color = mix(background, image, mask);
    outColor = vec4(color, 1.0);
}`;
  }
  /**
   * Set up the WebGL texture for the background image or video.
   * Creates and configures the texture that will be used as the replacement background.
   */
  setupBackgroundTexture() {
    const e = this.config.render.background;
    this.backgroundTexture = this.createTexture(
      this.gl.RGBA,
      e.width,
      e.height,
      () => {
        this.gl.texImage2D(
          this.gl.TEXTURE_2D,
          0,
          this.gl.RGBA,
          this.gl.RGBA,
          this.gl.UNSIGNED_BYTE,
          e
        );
      }
    );
    const r = this.gl.getUniformLocation(this.program, "u_background");
    r || g.errorAndThrow('Error finding "u_background" uniform'), this.gl.activeTexture(this.gl.TEXTURE2), this.gl.bindTexture(this.gl.TEXTURE_2D, this.backgroundTexture), this.gl.uniform1i(r, 2);
  }
  /**
   * Render a frame with the replacement background effect.
   * Combines the input video stream with the background based on the provided mask.
   * Includes frame rate throttling for video backgrounds to prevent stuttering.
   * Canvas backgrounds are updated every frame to support dynamic content.
   *
   * @param input - The input frame data (video element or image data).
   * @param mask - Optional mask data for determining foreground/background blend.
   * @returns Promise that resolves when rendering is complete.
   */
  async render(t, e) {
    this.uploadVideoStream(t), this.uploadMask(e);
    const r = this.config.render, s = !r.static, i = r.background;
    s && (i instanceof HTMLVideoElement ? await this.updateVideoBackgroundTexture(i) : (this.gl.activeTexture(this.gl.TEXTURE2), this.gl.bindTexture(this.gl.TEXTURE_2D, this.backgroundTexture), this.gl.texImage2D(
      this.gl.TEXTURE_2D,
      0,
      this.gl.RGBA,
      this.gl.RGBA,
      this.gl.UNSIGNED_BYTE,
      i
    ))), this.gl.useProgram(this.program), this.gl.drawArrays(this.gl.TRIANGLES, 0, this.verts.length / 3), this.postDraw();
  }
  /**
   * Throttle video background updates and upload texture only when needed.
   * This prevents stuttering and unnecessary GPU uploads when inference is skipped.
   *
   * @param backgroundSource - The video element to potentially upload.
   * @returns Promise that resolves after appropriate throttling and upload.
   */
  async updateVideoBackgroundTexture(t) {
    if (this.videoFrameInterval <= 0) return;
    const e = performance.now();
    e - this.lastVideoFrameTime >= this.videoFrameInterval && (this.gl.activeTexture(this.gl.TEXTURE2), this.gl.bindTexture(this.gl.TEXTURE_2D, this.backgroundTexture), this.gl.texImage2D(
      this.gl.TEXTURE_2D,
      0,
      this.gl.RGBA,
      this.gl.RGBA,
      this.gl.UNSIGNED_BYTE,
      t
    ), this.lastVideoFrameTime = e);
  }
  /**
   * Attempt to determine the frame rate of a video element.
   * Tries to extract frame rate from video track settings if available.
   *
   * @param video - The video element to analyze.
   * @returns The video's frame rate in frames per second, or undefined if not determinable.
   */
  getVideoFrameRate(t) {
    if (t.srcObject instanceof MediaStream) {
      const e = t.srcObject.getVideoTracks()[0];
      if (e != null && e.getSettings)
        return e.getSettings().frameRate;
    }
  }
  /**
   * Load and initialize the renderer.
   * Currently a no-op for this renderer implementation.
   *
   * @returns Promise that resolves when loading is complete.
   */
  async load() {
  }
  /**
   * Clean up resources and destroy the renderer.
   * Currently a no-op for this renderer implementation.
   */
  destroy() {
  }
}
export {
  T as ReplacementRenderer
};
