// @webex/webex-core/dist/lib/services-v2/services-v2.js
"use strict";

var _Object$keys2 = require("@babel/runtime-corejs2/core-js/object/keys");
var _Object$getOwnPropertySymbols = require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols");
var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor");
var _Object$getOwnPropertyDescriptors = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptors");
var _Object$defineProperties = require("@babel/runtime-corejs2/core-js/object/define-properties");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DEFAULT_CLUSTER_SERVICE = exports.DEFAULT_CLUSTER = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs2/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/asyncToGenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/slicedToArray"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));
var _weakMap = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/weak-map"));
var _isArray = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/array/is-array"));
var _keys = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/keys"));
var _promise = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/promise"));
var _values = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/values"));
var _now = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/date/now"));
var _stringify = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/json/stringify"));
var _sha = _interopRequireDefault(require("crypto-js/sha256"));
var _lodash = require("lodash");
var _webexPlugin = _interopRequireDefault(require("../webex-plugin"));
var _metrics = _interopRequireDefault(require("../metrics"));
var _serviceCatalog = _interopRequireDefault(require("./service-catalog"));
var _serviceFedRamp = _interopRequireDefault(require("./service-fed-ramp"));
var _constants = require("../constants");
function ownKeys(e, r) { var t = _Object$keys2(e); if (_Object$getOwnPropertySymbols) { var o = _Object$getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return _Object$getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(e, _Object$getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { _Object$defineProperty(e, r, _Object$getOwnPropertyDescriptor(t, r)); }); } return e; }
var trailingSlashes = /(?:^\/)|(?:\/$)/;

// The default cluster when one is not provided (usually as 'US' from hydra)
var DEFAULT_CLUSTER = exports.DEFAULT_CLUSTER = 'urn:TEAM:us-east-2_a';
// The default service name for convo (currently identityLookup due to some weird CSB issue)
var DEFAULT_CLUSTER_SERVICE = exports.DEFAULT_CLUSTER_SERVICE = 'identityLookup';
var CLUSTER_SERVICE = process.env.WEBEX_CONVERSATION_CLUSTER_SERVICE || DEFAULT_CLUSTER_SERVICE;
var DEFAULT_CLUSTER_IDENTIFIER = process.env.WEBEX_CONVERSATION_DEFAULT_CLUSTER || "".concat(DEFAULT_CLUSTER, ":").concat(CLUSTER_SERVICE);
var CATALOG_CACHE_KEY_V2 = 'services.v2.u2cHostMap';
var CATALOG_TTL_MS = 24 * 60 * 60 * 1000; // 24 hours

/* eslint-disable no-underscore-dangle */
/**
 * @class
 */
var Services = _webexPlugin.default.extend({
  namespace: 'Services',
  props: {
    validateDomains: ['boolean', false, true],
    initFailed: ['boolean', false, false]
  },
  _catalogs: new _weakMap.default(),
  _activeServices: {},
  _services: [],
  /**
   * @private
   * Get the current catalog based on the assocaited
   * webex instance.
   * @returns {IServiceCatalog}
   */
  _getCatalog: function _getCatalog() {
    return this._catalogs.get(this.webex);
  },
  /**
   * Safely access localStorage if available; returns the Storage or null.
   * @returns {Storage | null}
   */
  _getLocalStorageSafe: function _getLocalStorageSafe() {
    if (typeof window !== 'undefined' && window.localStorage) {
      return window.localStorage;
    }
    return null;
  },
  /**
   * Determine the intended preauth selection based on the current context.
   * @param {string} [currentOrgId]
   * @returns {{selectionType: string, selectionValue: string}}
   */
  getIntendedPreauthSelection: function getIntendedPreauthSelection(currentOrgId) {
    var _this$webex$credentia;
    if ((_this$webex$credentia = this.webex.credentials) !== null && _this$webex$credentia !== void 0 && _this$webex$credentia.canAuthorize) {
      if (currentOrgId) {
        return {
          selectionType: 'orgId',
          selectionValue: currentOrgId
        };
      }
    }
    var emailConfig = this.webex.config && this.webex.config.email;
    if (typeof emailConfig === 'string' && emailConfig.trim()) {
      return {
        selectionType: 'emailhash',
        selectionValue: (0, _sha.default)(emailConfig.toLowerCase()).toString()
      };
    }

    // fall back to proximity mode when no orgId or email available
    return {
      selectionType: 'mode',
      selectionValue: 'DEFAULT_BY_PROXIMITY'
    };
  },
  /**
   * Get a service url from the current services list by name
   * from the associated instance catalog.
   * @param {string} name
   * @param {ServiceGroup} [serviceGroup]
   * @returns {string|undefined}
   */
  get: function get(name, serviceGroup) {
    var catalog = this._getCatalog();
    var clusterId = this._activeServices[name];
    var urlById = catalog.get(clusterId, serviceGroup);
    var urlByName = catalog.get(name, serviceGroup);

    // if both are undefined, then we cannot find the service
    if (!urlById && !urlByName) {
      return undefined;
    }
    return urlById || urlByName;
  },
  /**
   * Determine if a whilelist exists in the service catalog.
   *
   * @returns {boolean} - True if a allowed domains list exists.
   */
  hasAllowedDomains: function hasAllowedDomains() {
    var catalog = this._getCatalog();
    return catalog.getAllowedDomains().length > 0;
  },
  /**
   * Mark a priority host service url as failed.
   * This will mark the service url associated with the
   * `ServiceDetail` to be removed from the its
   * respective service url array, and then return the next
   * viable service url from the `ServiceDetail` service url array.
   * @param {string} url
   * @returns {string}
   */
  markFailedUrl: function markFailedUrl(url) {
    var catalog = this._getCatalog();
    return catalog.markFailedServiceUrl(url);
  },
  /**
   * Get all Mobius cluster host entries from the v2 services list.
   * @returns {Array<ServiceHost>} - An array of `ServiceHost` objects.
   */
  getMobiusClusters: function getMobiusClusters() {
    this.logger.info('services: fetching mobius clusters');
    var clusters = [];
    var services = this._services || [];
    services.filter(function (service) {
      return (service === null || service === void 0 ? void 0 : service.serviceName) === 'mobius' && (0, _isArray.default)(service.serviceUrls) && service.serviceUrls.length > 0;
    }).forEach(function (service) {
      service.serviceUrls.forEach(function (serviceUrl) {
        var modifiedHost = serviceUrl.baseUrl.replace('https://', '').replace('/api/v1', '');
        if (!clusters.find(function (c) {
          return c && c.host === modifiedHost;
        })) {
          clusters.push({
            host: modifiedHost,
            priority: serviceUrl.priority,
            id: service.id,
            ttl: 0
          });
        }
      });
    });
    return clusters;
  },
  /**
   * Check is valid host from services list.
   * @param {string} host
   * @returns {Boolean}
   */
  isValidHost: function isValidHost(host) {
    var services = this._services || [];
    return services.some(function (service) {
      return service.serviceUrls.some(function (serviceUrl) {
        var _URL;
        var serviceHost = (serviceUrl === null || serviceUrl === void 0 ? void 0 : serviceUrl.baseUrl) && ((_URL = new URL(serviceUrl.baseUrl)) === null || _URL === void 0 ? void 0 : _URL.host);
        return serviceHost === host;
      });
    });
  },
  /**
   * Checks if the current environment is an integration (INT) environment
   * by examining the u2c discovery URL from webex config.
   * INT environments use discovery URLs containing 'intb' (e.g., u2c-intb.ciscospark.com).
   * @returns {boolean} True if INT environment, false otherwise
   */
  isIntegrationEnvironment: function isIntegrationEnvironment() {
    try {
      var _this$webex, _this$webex$config, _this$webex$config$se, _this$webex$config$se2;
      var u2cUrl = ((_this$webex = this.webex) === null || _this$webex === void 0 ? void 0 : (_this$webex$config = _this$webex.config) === null || _this$webex$config === void 0 ? void 0 : (_this$webex$config$se = _this$webex$config.services) === null || _this$webex$config$se === void 0 ? void 0 : (_this$webex$config$se2 = _this$webex$config$se.discovery) === null || _this$webex$config$se2 === void 0 ? void 0 : _this$webex$config$se2.u2c) || '';
      var isInt = u2cUrl.includes('intb');
      this.logger.info("services: isIntegrationEnvironment: ".concat(isInt));
      return isInt;
    } catch (error) {
      this.logger.error('services: failed to determine integration environment', error);
      return false;
    }
  },
  /**
   * saves all the services from the pre and post catalog service
   * @param {ActiveServices} activeServices
   * @returns {void}
   */
  _updateActiveServices: function _updateActiveServices(activeServices) {
    this._activeServices = _objectSpread(_objectSpread({}, this._activeServices), activeServices);
  },
  /**
   * saves the hostCatalog object
   * @param {Array<Service>} services
   * @returns {void}
   */
  _updateServices: function _updateServices(services) {
    this._services = (0, _lodash.unionBy)(services, this._services, 'id');
  },
  /**
   * Update a list of `serviceUrls` to the most current
   * catalog via the defined `discoveryUrl` then returns the current
   * list of services.
   * @param {object} [param]
   * @param {string} [param.from] - This accepts `limited` or `signin`
   * @param {object} [param.query] - This accepts `email`, `orgId` or `userId` key values
   * @param {string} [param.query.email] - must be a standard-format email
   * @param {string} [param.query.orgId] - must be an organization id
   * @param {string} [param.query.userId] - must be a user id
   * @param {string} [param.token] - used for signin catalog
   * @returns {Promise<object>}
   */
  updateServices: function updateServices() {
    var _catalog$status,
      _catalog$status$servi,
      _this = this;
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      from = _ref.from,
      query = _ref.query,
      token = _ref.token,
      forceRefresh = _ref.forceRefresh;
    var catalog = this._getCatalog();
    var formattedQuery;
    var serviceGroup;

    // map catalog name to service group name.
    switch (from) {
      case 'limited':
        serviceGroup = 'preauth';
        break;
      case 'signin':
        serviceGroup = 'signin';
        break;
      default:
        serviceGroup = 'postauth';
        break;
    }
    // confirm catalog update for group is not in progress.
    if ((_catalog$status = catalog.status) !== null && _catalog$status !== void 0 && (_catalog$status$servi = _catalog$status[serviceGroup]) !== null && _catalog$status$servi !== void 0 && _catalog$status$servi.collecting) {
      return this.waitForCatalog(serviceGroup);
    }
    catalog.status[serviceGroup].collecting = true;
    if (serviceGroup === 'preauth') {
      var queryKey = query && (0, _keys.default)(query)[0];
      if (!['email', 'emailhash', 'userId', 'orgId', 'mode'].includes(queryKey)) {
        return _promise.default.reject(new Error('a query param of email, emailhash, userId, orgId, or mode is required'));
      }
    }
    // encode email when query key is email
    if (serviceGroup === 'preauth' || serviceGroup === 'signin') {
      var _queryKey = (0, _keys.default)(query)[0];
      formattedQuery = {};
      if (_queryKey === 'email' && query.email) {
        formattedQuery.emailhash = (0, _sha.default)(query.email.toLowerCase()).toString();
      } else {
        formattedQuery[_queryKey] = query[_queryKey];
      }
    }
    return this._fetchNewServiceHostmap({
      from: from,
      token: token,
      query: formattedQuery,
      forceRefresh: forceRefresh
    }).then(function (serviceHostMap) {
      catalog.updateServiceGroups(serviceGroup, serviceHostMap === null || serviceHostMap === void 0 ? void 0 : serviceHostMap.services, serviceHostMap === null || serviceHostMap === void 0 ? void 0 : serviceHostMap.timestamp);
      // Build selection metadata for caching discrimination (preauth/signin)
      var selectionMeta;
      if (serviceGroup === 'preauth' || serviceGroup === 'signin') {
        var key = formattedQuery && (0, _keys.default)(formattedQuery || {})[0];
        if (key) {
          selectionMeta = {
            selectionType: key,
            selectionValue: formattedQuery[key]
          };
        }
      }
      _this._cacheCatalog(serviceGroup, serviceHostMap, selectionMeta);
      _this.updateCredentialsConfig();
      catalog.status[serviceGroup].collecting = false;
    }).catch(function (error) {
      catalog.status[serviceGroup].collecting = false;
      return _promise.default.reject(error);
    });
  },
  /**
   * User validation parameter transfer object for {@link validateUser}.
   * @param {object} ValidateUserPTO
   * @property {string} ValidateUserPTO.email - The email of the user.
   * @property {string} [ValidateUserPTO.reqId] - The activation requester.
   * @property {object} [ValidateUserPTO.activationOptions] - Extra options to pass when sending the activation
   * @property {object} [ValidateUserPTO.preloginUserId] - The prelogin user id to set when sending the activation.
   */
  /**
   * User validation return transfer object for {@link validateUser}.
   * @param {object} ValidateUserRTO
   * @property {boolean} ValidateUserRTO.activated - If the user is activated.
   * @property {boolean} ValidateUserRTO.exists - If the user exists.
   * @property {string} ValidateUserRTO.details - A descriptive status message.
   * @property {object} ValidateUserRTO.user - **License** service user object.
   */
  /**
   * Validate if a user is activated and update the service catalogs as needed
   * based on the user's activation status.
   *
   * @param {ValidateUserPTO} - The parameter transfer object.
   * @returns {ValidateUserRTO} - The return transfer object.
   */
  validateUser: function validateUser(_ref2) {
    var _this2 = this;
    var email = _ref2.email,
      _ref2$reqId = _ref2.reqId,
      reqId = _ref2$reqId === void 0 ? 'WEBCLIENT' : _ref2$reqId,
      _ref2$forceRefresh = _ref2.forceRefresh,
      forceRefresh = _ref2$forceRefresh === void 0 ? false : _ref2$forceRefresh,
      _ref2$activationOptio = _ref2.activationOptions,
      activationOptions = _ref2$activationOptio === void 0 ? {} : _ref2$activationOptio,
      preloginUserId = _ref2.preloginUserId;
    this.logger.info('services: validating a user');

    // Validate that an email parameter key was provided.
    if (!email) {
      return _promise.default.reject(new Error('`email` is required'));
    }

    // Destructure the credentials object.
    var canAuthorize = this.webex.credentials.canAuthorize;

    // Validate that the user is already authorized.
    if (canAuthorize) {
      return this.updateServices({
        forceRefresh: forceRefresh
      }).then(function () {
        return _this2.webex.credentials.getUserToken();
      }).then(function (token) {
        return _this2.sendUserActivation({
          email: email,
          reqId: reqId,
          token: token.toString(),
          activationOptions: activationOptions,
          preloginUserId: preloginUserId
        });
      }).then(function (userObj) {
        return {
          activated: true,
          exists: true,
          details: 'user is authorized via a user token',
          user: userObj
        };
      });
    }

    // Destructure the client authorization details.
    /* eslint-disable camelcase */
    var _this$webex$credentia2 = this.webex.credentials.config,
      client_id = _this$webex$credentia2.client_id,
      client_secret = _this$webex$credentia2.client_secret;

    // Validate that client authentication details exist.
    if (!client_id || !client_secret) {
      return _promise.default.reject(new Error('client authentication details are not available'));
    }
    /* eslint-enable camelcase */

    // Declare a class-memeber-scoped token for usage within the promise chain.
    var token;

    // Begin client authentication user validation.
    return this.collectPreauthCatalog({
      email: email
    }).then(function () {
      // Retrieve the service url from the updated catalog. This is required
      // since `WebexCore` is usually not fully initialized at the time this
      // request completes.
      var idbrokerService = _this2.get('idbroker');

      // Collect the client auth token.
      return _this2.webex.credentials.getClientToken({
        uri: "".concat(idbrokerService, "idb/oauth2/v1/access_token"),
        scope: 'webexsquare:admin webexsquare:get_conversation Identity:SCIM'
      });
    }).then(function (tokenObj) {
      // Generate the token string.
      token = tokenObj.toString();

      // Collect the signin catalog using the client auth information.
      return _this2.collectSigninCatalog({
        email: email,
        token: token,
        forceRefresh: forceRefresh
      });
    })
    // Validate if collecting the signin catalog failed and populate the RTO
    // with the appropriate content.
    .catch(function (error) {
      return {
        exists: error.name !== 'NotFound',
        activated: false,
        details: error.name !== 'NotFound' ? 'user exists but is not activated' : 'user does not exist and is not activated'
      };
    })
    // Validate if the previous promise resolved with an RTO and populate the
    // new RTO accordingly.
    .then(function (rto) {
      return _promise.default.all([rto || {
        activated: true,
        exists: true,
        details: 'user exists and is activated'
      }, _this2.sendUserActivation({
        email: email,
        reqId: reqId,
        token: token,
        activationOptions: activationOptions,
        preloginUserId: preloginUserId
      })]);
    }).then(function (_ref3) {
      var _ref4 = (0, _slicedToArray2.default)(_ref3, 2),
        rto = _ref4[0],
        user = _ref4[1];
      return _objectSpread(_objectSpread({}, rto), {}, {
        user: user
      });
    }).catch(function (error) {
      var response = {
        statusCode: error.statusCode,
        responseText: error.body && error.body.message,
        body: error.body
      };
      return _promise.default.reject(response);
    });
  },
  /**
   * Update cluster id via mercury service update. If the cluster id does not exist,
   * fetch new catalog.
   *
   * @param {ActiveServices} newActiveClusters - The new active clusters to switch to.
   * @returns {Promsie<void>}
   * */
  switchActiveClusterIds: function switchActiveClusterIds(newActiveClusters) {
    var _this3 = this;
    this.logger.info('services: switching active cluster ids');
    var newActiveClusterIds = (0, _values.default)(newActiveClusters);
    var missingClusterIds = newActiveClusterIds.some(function (clusterId) {
      // if the clusterId does not exist in the catalog, fetch the catalog
      return !_this3._services.find(function (service) {
        return service.id === clusterId;
      });
    });
    if (missingClusterIds) {
      this.logger.warn('services: some cluster ids do not exist in the catalog, fetching the catalog');

      // fetch the catalog
      return this.initServiceCatalogs(true);
    }
    // update the active services
    this._updateActiveServices(newActiveClusters);
    this.logger.info('services: active cluster ids updated successfully');
    return _promise.default.resolve();
  },
  /**
   * Invalidate cache via mercury notification. If the timestamp is newer than current,
   * refetch catalog services.
   *
   * @param {string} timestamp - The timestamp of invalidation notification.
   * @returns {Promsie<void>}
   * */
  invalidateCache: function invalidateCache(timestamp) {
    var _this$_getCatalog;
    this.logger.info('services: invalidate cache, timestamp:', timestamp);
    var lastTime = (0, _lodash.toNumber)((_this$_getCatalog = this._getCatalog()) === null || _this$_getCatalog === void 0 ? void 0 : _this$_getCatalog.timestamp) || 0;
    var invalidateTime = (0, _lodash.toNumber)(timestamp) || 0;
    if (invalidateTime > lastTime) {
      this.logger.info('services: invalidateCache, refresh services');
      this.initServiceCatalogs(true);
    }
    return _promise.default.resolve();
  },
  /**
   * Get user meeting preferences (preferred webex site).
   *
   * @returns {object} - User Information including user preferrences .
   */
  getMeetingPreferences: function getMeetingPreferences() {
    var _this4 = this;
    return this.request({
      method: 'GET',
      service: 'hydra',
      resource: 'meetingPreferences'
    }).then(function (res) {
      _this4.logger.info('services: received user region info');
      return res.body;
    }).catch(function (err) {
      _this4.logger.info('services: was not able to fetch user login information', err);
      // resolve successfully even if request failed
    });
  },
  /**
   * Fetches client region info such as countryCode and timezone.
   *
   * @returns {object} - The region info object.
   */
  fetchClientRegionInfo: function fetchClientRegionInfo() {
    var _this5 = this;
    var services = this.webex.config.services;
    return this.request({
      uri: services.discovery.sqdiscovery,
      addAuthHeader: false,
      headers: {
        'spark-user-agent': null
      },
      timeout: 5000
    }).then(function (res) {
      _this5.logger.info('services: received user region info');
      return res.body;
    }).catch(function (err) {
      _this5.logger.info('services: was not able to get user region info', err);
      // resolve successfully even if request failed
    });
  },
  /**
   * User activation parameter transfer object for {@link sendUserActivation}.
   * @typedef {object} SendUserActivationPTO
   * @property {string} SendUserActivationPTO.email - The email of the user.
   * @property {string} SendUserActivationPTO.reqId - The activation requester.
   * @property {string} SendUserActivationPTO.token - The client auth token.
   * @property {object} SendUserActivationPTO.activationOptions - Extra options to pass when sending the activation.
   * @property {object} SendUserActivationPTO.preloginUserId - The prelogin user id to set when sending the activation.
   */
  /**
   * Send a request to activate a user using a client token.
   *
   * @param {SendUserActivationPTO} - The Parameter transfer object.
   * @returns {LicenseDTO} - The DTO returned from the **License** service.
   */
  sendUserActivation: function sendUserActivation(_ref5) {
    var _this6 = this;
    var email = _ref5.email,
      reqId = _ref5.reqId,
      token = _ref5.token,
      activationOptions = _ref5.activationOptions,
      preloginUserId = _ref5.preloginUserId;
    this.logger.info('services: sending user activation request');
    var countryCode;
    var timezone;

    // try to fetch client region info first
    return this.fetchClientRegionInfo().then(function (clientRegionInfo) {
      var _this6$webex$config$s;
      if (clientRegionInfo) {
        countryCode = clientRegionInfo.countryCode;
        timezone = clientRegionInfo.timezone;
      }

      // Send the user activation request.
      // Use user-onboarding service if configured, otherwise use license service.
      var useUserOnboarding = (_this6$webex$config$s = _this6.webex.config.services) === null || _this6$webex$config$s === void 0 ? void 0 : _this6$webex$config$s.useUserOnboardingServiceForActivations;
      return _this6.request({
        service: useUserOnboarding ? 'user-onboarding' : 'license',
        resource: useUserOnboarding ? 'api/v1/users/activations' : 'users/activations',
        method: 'POST',
        headers: {
          accept: 'application/json',
          authorization: token,
          'x-prelogin-userid': preloginUserId
        },
        body: _objectSpread({
          email: email,
          reqId: reqId,
          countryCode: countryCode,
          timeZone: timezone
        }, activationOptions),
        shouldRefreshAccessToken: false
      });
    })
    // On success, return the **License** user object.
    .then(function (_ref6) {
      var body = _ref6.body;
      return body;
    })
    // On failure, reject with error from **License**.
    .catch(function (error) {
      return _promise.default.reject(error);
    });
  },
  /**
   * Updates a given service group i.e. preauth, signin, postauth with a new hostmap.
   * @param {ServiceGroup} serviceGroup - preauth, signin, postauth
   * @param {ServiceHostmap} hostMap - The new hostmap to update the service group with.
   * @returns {Promise<void>}
   */
  updateCatalog: function updateCatalog(serviceGroup, hostMap) {
    var catalog = this._getCatalog();
    var serviceHostMap = this._formatReceivedHostmap(hostMap || {});
    return catalog.updateServiceGroups(serviceGroup, serviceHostMap === null || serviceHostMap === void 0 ? void 0 : serviceHostMap.services, serviceHostMap === null || serviceHostMap === void 0 ? void 0 : serviceHostMap.timestamp);
  },
  /**
   * simplified method to update the preauth catalog via email
   *
   * @param {object} query
   * @param {string} query.email - A standard format email.
   * @param {string} query.orgId - The user's OrgId.
   * @param {boolean} forceRefresh - Boolean to bypass u2c cache control header
   * @returns {Promise<void>}
   */
  collectPreauthCatalog: function collectPreauthCatalog(query) {
    var forceRefresh = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
    if (!query) {
      return this.updateServices({
        from: 'limited',
        query: {
          mode: 'DEFAULT_BY_PROXIMITY'
        },
        forceRefresh: forceRefresh
      });
    }
    return this.updateServices({
      from: 'limited',
      query: query,
      forceRefresh: forceRefresh
    });
  },
  /**
   * simplified method to update the signin catalog via email and token
   * @param {object} param
   * @param {string} param.email - must be a standard-format email
   * @param {string} param.token - must be a client token
   * @returns {Promise<void>}
   */
  collectSigninCatalog: function collectSigninCatalog() {
    var _ref7 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      email = _ref7.email,
      token = _ref7.token,
      forceRefresh = _ref7.forceRefresh;
    if (!email) {
      return _promise.default.reject(new Error('`email` is required'));
    }
    if (!token) {
      return _promise.default.reject(new Error('`token` is required'));
    }
    return this.updateServices({
      from: 'signin',
      query: {
        email: email
      },
      token: token,
      forceRefresh: forceRefresh
    });
  },
  /**
   * Updates credentials config to utilize u2c catalog
   * urls.
   * @returns {void}
   */
  updateCredentialsConfig: function updateCredentialsConfig() {
    var idbrokerUrl = this.get('idbroker');
    var identityUrl = this.get('identity');
    if (idbrokerUrl && identityUrl) {
      var _this$webex$config$cr = this.webex.config.credentials,
        authorizationString = _this$webex$config$cr.authorizationString,
        authorizeUrl = _this$webex$config$cr.authorizeUrl;

      // This must be set outside of the setConfig method used to assign the
      // idbroker and identity url values.
      this.webex.config.credentials.authorizeUrl = authorizationString ? authorizeUrl : "".concat(idbrokerUrl.replace(trailingSlashes, ''), "/idb/oauth2/v1/authorize");
      this.webex.setConfig({
        credentials: {
          idbroker: {
            url: idbrokerUrl.replace(trailingSlashes, '') // remove trailing slash
          },
          identity: {
            url: identityUrl.replace(trailingSlashes, '') // remove trailing slash
          }
        }
      });
    }
  },
  /**
   * Wait until the service catalog is available,
   * or reject afte ra timeout of 60 seconds.
   * @param {ServiceGroup} serviceGroup
   * @param {number} [timeout] - in seconds
   * @returns {Promise<void>}
   */
  waitForCatalog: function waitForCatalog(serviceGroup, timeout) {
    var catalog = this._getCatalog();
    var supertoken = this.webex.credentials.supertoken;
    if (serviceGroup === 'postauth' && supertoken && supertoken.access_token && !catalog.status.postauth.collecting && !catalog.status.postauth.ready) {
      if (!catalog.status.preauth.ready) {
        return this.initServiceCatalogs();
      }
      return this.updateServices();
    }
    return catalog.waitForCatalog(serviceGroup, timeout);
  },
  /**
   * Service waiting parameter transfer object for {@link waitForService}.
   *
   * @typedef {object} WaitForServicePTO
   * @property {string} [WaitForServicePTO.name] - The service name.
   * @property {string} [WaitForServicePTO.url] - The service url.
   * @property {string} [WaitForServicePTO.timeout] - wait duration in seconds.
   */
  /**
   * Wait until the service has been ammended to any service catalog. This
   * method prioritizes the service name over the service url when searching.
   *
   * @param {WaitForServicePTO} - The parameter transfer object.
   * @returns {Promise<string>} - Resolves to the priority host of a service.
   */
  waitForService: function waitForService(_ref8) {
    var _this7 = this;
    var name = _ref8.name,
      _ref8$timeout = _ref8.timeout,
      timeout = _ref8$timeout === void 0 ? 5 : _ref8$timeout,
      url = _ref8.url;
    var services = this.webex.config.services;

    // Save memory by grabbing the catalog after there isn't a priortyURL
    var catalog = this._getCatalog();
    var fetchFromServiceUrl = services.servicesNotNeedValidation.find(function (service) {
      return service === name;
    });
    if (fetchFromServiceUrl) {
      var clusterId = this._activeServices[name];
      return _promise.default.resolve(this.get(clusterId));
    }
    var priorityUrl = this.get(name);
    var priorityUrlObj = this.getServiceFromUrl(url);
    if (priorityUrl || priorityUrlObj) {
      return _promise.default.resolve(priorityUrl || priorityUrlObj.priorityUrl);
    }
    if (catalog.isReady) {
      if (url) {
        return _promise.default.resolve(url);
      }
      this.webex.internal.metrics.submitClientMetrics(_metrics.default.JS_SDK_SERVICE_NOT_FOUND, {
        fields: {
          service_name: name
        }
      });
      return _promise.default.reject(new Error("services: service '".concat(name, "' was not found in any of the catalogs")));
    }
    return new _promise.default(function (resolve, reject) {
      var groupsToCheck = ['preauth', 'signin', 'postauth'];
      var checkCatalog = function checkCatalog(catalogGroup) {
        return catalog.waitForCatalog(catalogGroup, timeout).then(function () {
          var scopedPriorityUrl = _this7.get(name);
          var scopedPrioriryUrlObj = _this7.getServiceFromUrl(url);
          if (scopedPriorityUrl || scopedPrioriryUrlObj) {
            resolve(scopedPriorityUrl || scopedPrioriryUrlObj.priorityUrl);
          }
        }).catch(function () {
          return undefined;
        });
      };
      _promise.default.all(groupsToCheck.map(function (group) {
        return checkCatalog(group);
      })).then(function () {
        _this7.webex.internal.metrics.submitClientMetrics(_metrics.default.JS_SDK_SERVICE_NOT_FOUND, {
          fields: {
            service_name: name
          }
        });
        reject(new Error("services: service '".concat(name, "' was not found after waiting")));
      });
    });
  },
  /**
   * Looks up the hostname in the host catalog
   * and replaces it with the first host if it finds it
   * @param {string} uri
   * @returns {string} uri with the host replaced
   */
  replaceHostFromHostmap: function replaceHostFromHostmap(uri) {
    try {
      return this.convertUrlToPriorityHostUrl(uri);
    } catch (_unused) {
      return uri;
    }
  },
  /**
   * Formats a host map entry for use in service catalog.
   *
   * @param {Object} entry - The host map entry to format.
   * @param {string} entry.serviceName - i.e. conversation, identity, etc.
   * @param {string} entry.id - The unique identifier for the service, usually clusterId.
   * @param {Array<IServiceDetail>} entry.serviceUrls - The group to which the service belongs.
   * @returns {Object} - The formatted host map entry.
   */
  _formatHostMapEntry: function _formatHostMapEntry(_ref9) {
    var id = _ref9.id,
      serviceName = _ref9.serviceName,
      serviceUrls = _ref9.serviceUrls;
    var formattedServiceUrls = serviceUrls.map(function (serviceUrl) {
      return _objectSpread({
        host: new URL(serviceUrl.baseUrl).host
      }, serviceUrl);
    });
    return {
      id: id,
      serviceName: serviceName,
      serviceUrls: formattedServiceUrls
    };
  },
  /**
   * @private
   * Organize a received hostmap from a service
   * @param {ServiceHostmap} serviceHostmap
   * catalog endpoint.
   * @returns {Array<Service>}
   */
  _formatReceivedHostmap: function _formatReceivedHostmap(_ref0) {
    var _this8 = this;
    var services = _ref0.services,
      activeServices = _ref0.activeServices,
      timestamp = _ref0.timestamp,
      orgId = _ref0.orgId,
      format = _ref0.format;
    var formattedHostmap = {
      activeServices: activeServices,
      services: services === null || services === void 0 ? void 0 : services.map(function (service) {
        return _this8._formatHostMapEntry(service);
      }),
      timestamp: timestamp,
      orgId: orgId,
      format: format
    };
    this._updateActiveServices(activeServices);
    this._updateServices(services);
    return formattedHostmap;
  },
  /**
   * Get the clusterId associated with a URL string.
   * @param {string} url
   * @returns {string | undefined} - Cluster ID of url provided
   */
  getClusterId: function getClusterId(url) {
    var catalog = this._getCatalog();
    return catalog.findClusterId(url);
  },
  /**
   * Get a service value from a provided clusterId. This method will
   * return an object containing both the name and url of a found service.
   * @param {object} params
   * @param {string} params.clusterId - clusterId of found service
   * @param {ServiceGroup} [params.serviceGroup] - specify service group
   * @returns {object} service
   * @returns {string} service.name
   * @returns {string} service.url
   */
  getServiceFromClusterId: function getServiceFromClusterId(params) {
    var catalog = this._getCatalog();
    return catalog.findServiceFromClusterId(params);
  },
  /**
   * @param {String} cluster the cluster containing the id
   * @param {UUID} [id] the id of the conversation.
   *  If empty, just return the base URL.
   * @returns {String} url of the service
   */
  getServiceUrlFromClusterId: function getServiceUrlFromClusterId() {
    var _ref1 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      _ref1$cluster = _ref1.cluster,
      cluster = _ref1$cluster === void 0 ? 'us' : _ref1$cluster;
    var clusterId = cluster === 'us' ? DEFAULT_CLUSTER_IDENTIFIER : cluster;

    // Determine if cluster has service name (non-US clusters from hydra do not)
    if (clusterId.split(':').length < 4) {
      // Add Service to cluster identifier
      clusterId = "".concat(cluster, ":").concat(CLUSTER_SERVICE);
    }
    var _ref10 = this.getServiceFromClusterId({
        clusterId: clusterId
      }) || {},
      url = _ref10.url;
    if (!url) {
      throw Error("Could not find service for cluster [".concat(cluster, "]"));
    }
    return url;
  },
  /**
   * Get a service object from a service url if the service url exists in the
   * catalog.
   *
   * @param {string} url - The url to be validated.
   * @returns {object} - Service object.
   * @returns {object.name} - The name of the service found.
   * @returns {object.priorityUrl} - The default url of the found service.
   * @returns {object.defaultUrl} - The default url of the found service.
   */
  getServiceFromUrl: function getServiceFromUrl() {
    var url = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    var service = this._getCatalog().findServiceDetailFromUrl(url);
    if (!service) {
      return undefined;
    }
    var priorityUrl = service.get();
    var defaultUrl = new URL(service.serviceUrls.find(function (serviceUrl) {
      return url.startsWith(serviceUrl.baseUrl);
    }).baseUrl).href;
    return {
      name: service.serviceName,
      priorityUrl: priorityUrl,
      defaultUrl: defaultUrl
    };
  },
  /**
   * Determine if a provided url is in the catalog's allowed domains.
   *
   * @param {string} url - The url to match allowed domains against.
   * @returns {boolean} - True if the url provided is allowed.
   */
  isAllowedDomainUrl: function isAllowedDomainUrl(url) {
    var catalog = this._getCatalog();
    return !!catalog.findAllowedDomain(url);
  },
  /**
   * Converts the host portion of the url from default host
   * to a priority host
   *
   * @param {string} url a service url that contains a default host
   * @returns {string} a service url that contains the top priority host.
   * @throws if url isn't a service url
   */
  convertUrlToPriorityHostUrl: function convertUrlToPriorityHostUrl() {
    var url = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    var data = this.getServiceFromUrl(url);
    if (!data) {
      throw Error("No service associated with url: [".concat(url, "]"));
    }
    return url.replace(data.defaultUrl, data.priorityUrl);
  },
  /**
   * @private
   * Cache the catalog in the bounded storage.
   * @param {ServiceGroup} serviceGroup - preauth, signin, postauth
   * @param {ServiceHostmap} hostMap - The hostmap to cache
   * @param {object} [meta] - Optional selection metadata for cache discrimination
   * @returns {Promise<void>}
   */
  _cacheCatalog: function _cacheCatalog(serviceGroup, hostMap, meta) {
    var _this9 = this;
    return (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee() {
      var current, orgId, _this9$webex$config, _this9$webex$config$c, _this9$webex, _this9$webex$config2, _this9$webex2, _this9$webex2$config, _this9$webex2$config$, _this9$webex2$config$2, _ls, cachedJson, credentials, _current, env, fedramp, u2cDiscoveryUrl, updated, ls, _t;
      return _regenerator.default.wrap(function (_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            current = {};
            _context.prev = 1;
            if ((_this9$webex$config = _this9.webex.config) !== null && _this9$webex$config !== void 0 && (_this9$webex$config$c = _this9$webex$config.calling) !== null && _this9$webex$config$c !== void 0 && _this9$webex$config$c.cacheU2C) {
              _context.next = 2;
              break;
            }
            _this9.logger.info("services: skipping cache write for ".concat(serviceGroup, " as per the config"));
            return _context.abrupt("return");
          case 2:
            try {
              _ls = _this9._getLocalStorageSafe();
              cachedJson = _ls ? _ls.getItem(CATALOG_CACHE_KEY_V2) : null;
              current = cachedJson ? JSON.parse(cachedJson) : {};
            } catch (_unused2) {
              current = {};
            }
            try {
              credentials = _this9.webex.credentials;
              orgId = credentials.getOrgId();
            } catch (_unused3) {
              orgId = current.orgId;
            }

            // Capture environment fingerprint to invalidate cache across env changes
            _current = current, env = _current.env;
            fedramp = !!((_this9$webex = _this9.webex) !== null && _this9$webex !== void 0 && (_this9$webex$config2 = _this9$webex.config) !== null && _this9$webex$config2 !== void 0 && _this9$webex$config2.fedramp);
            u2cDiscoveryUrl = (_this9$webex2 = _this9.webex) === null || _this9$webex2 === void 0 ? void 0 : (_this9$webex2$config = _this9$webex2.config) === null || _this9$webex2$config === void 0 ? void 0 : (_this9$webex2$config$ = _this9$webex2$config.services) === null || _this9$webex2$config$ === void 0 ? void 0 : (_this9$webex2$config$2 = _this9$webex2$config$.discovery) === null || _this9$webex2$config$2 === void 0 ? void 0 : _this9$webex2$config$2.u2c;
            env = {
              fedramp: fedramp,
              u2cDiscoveryUrl: u2cDiscoveryUrl
            };
            updated = _objectSpread(_objectSpread({}, current), {}, (0, _defineProperty2.default)((0, _defineProperty2.default)({
              orgId: orgId || current.orgId,
              env: env || current.env
            }, serviceGroup, meta ? {
              hostMap: hostMap,
              meta: meta
            } : hostMap), "cachedAt", (0, _now.default)()));
            ls = _this9._getLocalStorageSafe();
            if (ls) {
              ls.setItem(CATALOG_CACHE_KEY_V2, (0, _stringify.default)(updated));
            }
            _context.next = 4;
            break;
          case 3:
            _context.prev = 3;
            _t = _context["catch"](1);
            _this9.logger.warn('services: error caching catalog', _t);
          case 4:
          case "end":
            return _context.stop();
        }
      }, _callee, null, [[1, 3]]);
    }))();
  },
  /**
   * @private
   * Load the catalog from cache and hydrate the in-memory ServiceCatalog.
   * @returns {Promise<boolean>} true if cache was loaded, false otherwise
   */
  _loadCatalogFromCache: function _loadCatalogFromCache() {
    var _this0 = this;
    return (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee2() {
      var currentOrgId, _this0$webex$config, _this0$webex$config$c, _this0$webex$config2, _this0$webex$config3, _this0$webex$config3$, _this0$webex$config3$2, ls, cachedJson, cached, cachedAt, _this0$webex$credenti, credentials, fedramp, u2cDiscoveryUrl, currentEnv, sameEnv, catalog, groups, _t2, _t3;
      return _regenerator.default.wrap(function (_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            _context2.prev = 0;
            if ((_this0$webex$config = _this0.webex.config) !== null && _this0$webex$config !== void 0 && (_this0$webex$config$c = _this0$webex$config.calling) !== null && _this0$webex$config$c !== void 0 && _this0$webex$config$c.cacheU2C) {
              _context2.next = 1;
              break;
            }
            _this0.logger.info('services: skipping cache warm-up as per the cache config');
            return _context2.abrupt("return", false);
          case 1:
            ls = _this0._getLocalStorageSafe();
            if (ls) {
              _context2.next = 2;
              break;
            }
            _this0.logger.info('services: skipping cache warm-up as no localStorage is available');
            return _context2.abrupt("return", false);
          case 2:
            cachedJson = ls.getItem(CATALOG_CACHE_KEY_V2);
            cached = cachedJson ? JSON.parse(cachedJson) : undefined;
            if (cached) {
              _context2.next = 3;
              break;
            }
            return _context2.abrupt("return", false);
          case 3:
            // TTL enforcement
            cachedAt = Number(cached.cachedAt) || 0;
            if (!(!cachedAt || (0, _now.default)() - cachedAt > CATALOG_TTL_MS)) {
              _context2.next = 4;
              break;
            }
            _this0.clearCatalogCache();
            return _context2.abrupt("return", false);
          case 4:
            _context2.prev = 4;
            if (!((_this0$webex$credenti = _this0.webex.credentials) !== null && _this0$webex$credenti !== void 0 && _this0$webex$credenti.canAuthorize)) {
              _context2.next = 5;
              break;
            }
            credentials = _this0.webex.credentials;
            currentOrgId = credentials.getOrgId();
            if (!(cached.orgId && cached.orgId !== currentOrgId)) {
              _context2.next = 5;
              break;
            }
            return _context2.abrupt("return", false);
          case 5:
            _context2.next = 7;
            break;
          case 6:
            _context2.prev = 6;
            _t2 = _context2["catch"](4);
            _this0.logger.warn('services: error checking orgId', _t2);
          case 7:
            // Ensure cached environment matches current environment
            fedramp = !!((_this0$webex$config2 = _this0.webex.config) !== null && _this0$webex$config2 !== void 0 && _this0$webex$config2.fedramp);
            u2cDiscoveryUrl = (_this0$webex$config3 = _this0.webex.config) === null || _this0$webex$config3 === void 0 ? void 0 : (_this0$webex$config3$ = _this0$webex$config3.services) === null || _this0$webex$config3$ === void 0 ? void 0 : (_this0$webex$config3$2 = _this0$webex$config3$.discovery) === null || _this0$webex$config3$2 === void 0 ? void 0 : _this0$webex$config3$2.u2c;
            currentEnv = {
              fedramp: fedramp,
              u2cDiscoveryUrl: u2cDiscoveryUrl
            };
            if (!cached.env) {
              _context2.next = 8;
              break;
            }
            sameEnv = cached.env.fedramp === currentEnv.fedramp && cached.env.u2cDiscoveryUrl === currentEnv.u2cDiscoveryUrl;
            if (sameEnv) {
              _context2.next = 8;
              break;
            }
            return _context2.abrupt("return", false);
          case 8:
            catalog = _this0._getCatalog();
            groups = ['preauth', 'signin', 'postauth'];
            groups.forEach(function (serviceGroup) {
              var cachedGroup = cached[serviceGroup];
              if (!cachedGroup) {
                return;
              }

              // Support legacy (hostMap) and new ({hostMap, meta}) shapes
              var hostMap = cachedGroup && cachedGroup.hostMap ? cachedGroup.hostMap : cachedGroup;
              var meta = cachedGroup === null || cachedGroup === void 0 ? void 0 : cachedGroup.meta;
              if (serviceGroup === 'preauth' && meta) {
                // For proximity-based selection, always fetch fresh to respect IP/region changes
                if (meta.selectionType === 'mode') {
                  return;
                }
                var intended = _this0.getIntendedPreauthSelection(currentOrgId);
                var matches = intended && intended.selectionType === meta.selectionType && intended.selectionValue === meta.selectionValue;
                if (!matches) {
                  return;
                }
              }
              if (hostMap) {
                catalog.updateServiceGroups(serviceGroup, hostMap === null || hostMap === void 0 ? void 0 : hostMap.services, hostMap === null || hostMap === void 0 ? void 0 : hostMap.timestamp);
              }
            });
            _this0.updateCredentialsConfig();
            return _context2.abrupt("return", true);
          case 9:
            _context2.prev = 9;
            _t3 = _context2["catch"](0);
            return _context2.abrupt("return", false);
          case 10:
          case "end":
            return _context2.stop();
        }
      }, _callee2, null, [[0, 9], [4, 6]]);
    }))();
  },
  /**
   * Clear the catalog cache from the bounded storage (v2).
   * @returns {Promise<void>}
   */
  clearCatalogCache: function clearCatalogCache() {
    try {
      var ls = this._getLocalStorageSafe();
      if (ls) {
        ls.removeItem(CATALOG_CACHE_KEY_V2);
      }
    } catch (e) {
      this.logger.warn('services: error clearing catalog cache', e);
    }
    return _promise.default.resolve();
  },
  /**
   * @private
   * Simplified method wrapper for sending a request to get
   * an updated service hostmap.
   * @param {object} [param]
   * @param {string} [param.from] - This accepts `limited` or `signin`
   * @param {object} [param.query] - This accepts `email`, `orgId` or `userId` key values
   * @param {string} [param.query.email] - must be a standard-format email
   * @param {string} [param.query.orgId] - must be an organization id
   * @param {string} [param.query.userId] - must be a user id
   * @param {string} [param.token] - used for signin catalog
   * @returns {Promise<object>}
   */
  _fetchNewServiceHostmap: function _fetchNewServiceHostmap() {
    var _this1 = this;
    var _ref11 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      from = _ref11.from,
      query = _ref11.query,
      token = _ref11.token,
      forceRefresh = _ref11.forceRefresh;
    var service = 'u2c';
    var resource = from ? "/".concat(from, "/catalog") : '/catalog';
    var qs = _objectSpread(_objectSpread({}, query || {}), {}, {
      format: 'U2CV2'
    });
    if (forceRefresh) {
      qs.timestamp = new Date().getTime();
    }
    var requestObject = {
      method: 'GET',
      service: service,
      resource: resource,
      qs: qs,
      headers: {}
    };
    if (token) {
      requestObject.headers = {
        authorization: token
      };
    }
    return this.webex.internal.newMetrics.callDiagnosticLatencies.measureLatency(function () {
      return _this1.request(requestObject);
    }, 'internal.get.u2c.time').then(function (_ref12) {
      var body = _ref12.body;
      return _this1._formatReceivedHostmap(body || {});
    });
  },
  /**
   * Initialize the discovery services and the whitelisted services.
   *
   * @returns {void}
   */
  initConfig: function initConfig() {
    var _this10 = this;
    // Get the catalog and destructure the services config.
    var catalog = this._getCatalog();
    var _this$webex$config2 = this.webex.config,
      services = _this$webex$config2.services,
      fedramp = _this$webex$config2.fedramp;

    // Validate that the services configuration exists.
    if (services) {
      if (fedramp) {
        services.discovery = _serviceFedRamp.default;
      }
      // Check for discovery services.
      if (services.discovery) {
        // Format the discovery configuration into an injectable array.
        var formattedDiscoveryServices = (0, _keys.default)(services.discovery).map(function (key) {
          return _this10._formatHostMapEntry({
            id: key,
            serviceName: key,
            serviceUrls: [{
              baseUrl: services.discovery[key],
              priority: 1
            }]
          });
        });

        // Inject formatted discovery services into services catalog.
        catalog.updateServiceGroups('discovery', formattedDiscoveryServices);
      }
      if (services.override) {
        // Format the override configuration into an injectable array.
        var formattedOverrideServices = (0, _keys.default)(services.override).map(function (key) {
          return _this10._formatHostMapEntry({
            id: key,
            serviceName: key,
            serviceUrls: [{
              baseUrl: services.override[key],
              priority: 1
            }]
          });
        });

        // Inject formatted override services into services catalog.
        catalog.updateServiceGroups('override', formattedOverrideServices);
      }

      // if not fedramp, append on the commercialAllowedDomains
      if (!fedramp) {
        services.allowedDomains = (0, _lodash.union)(services.allowedDomains, _constants.COMMERCIAL_ALLOWED_DOMAINS);
      }

      // Check for allowed host domains.
      if (services.allowedDomains) {
        // Store the allowed domains as a property of the catalog.
        catalog.setAllowedDomains(services.allowedDomains);
      }

      // Set `validateDomains` property to match configuration
      this.validateDomains = services.validateDomains;
    }
  },
  /**
   * Make the initial requests to collect the root catalogs.
   * @param {boolean} refresh - Is need force update
   * @returns {Promise<void, Error>} - Errors if the token is unavailable.
   */
  initServiceCatalogs: function initServiceCatalogs() {
    var _this11 = this;
    var refresh = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
    this.logger.info('services: initializing initial service catalogs');

    // Destructure the credentials plugin.
    var credentials = this.webex.credentials;

    // Init a promise chain. Must be done as a Promise.resolve() to allow
    // credentials#getOrgId() to properly throw.
    return _promise.default.resolve()
    // Get the user's OrgId.
    .then(function () {
      return credentials.getOrgId();
    })
    // Begin collecting the preauth/limited catalog.
    .then(function (orgId) {
      return _this11.collectPreauthCatalog({
        orgId: orgId
      }, refresh);
    }).then(function () {
      // Validate if the token is authorized.
      if (credentials.canAuthorize) {
        // Attempt to collect the postauth catalog.
        return _this11.updateServices({
          forceRefresh: refresh
        }).catch(function () {
          _this11.initFailed = true;
          _this11.logger.warn('services: cannot retrieve postauth catalog');
        });
      }

      // Return a resolved promise for consistent return value.
      return _promise.default.resolve();
    });
  },
  /**
   * Initializer
   *
   * @instance
   * @memberof Services
   * @returns {Services}
   */
  initialize: function initialize() {
    var _this12 = this;
    var catalog = new _serviceCatalog.default();
    this._catalogs.set(this.webex, catalog);

    // Listen for configuration changes once.
    this.listenToOnce(this.webex, 'change:config', function () {
      _this12.initConfig();
    });

    // wait for webex instance to be ready before attempting
    // to update the service catalogs
    this.listenToOnce(this.webex, 'ready', /*#__PURE__*/(0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee3() {
      var warmed, supertoken, email;
      return _regenerator.default.wrap(function (_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 1;
            return _this12._loadCatalogFromCache();
          case 1:
            warmed = _context3.sent;
            if (!warmed) {
              _context3.next = 2;
              break;
            }
            catalog.isReady = true;
            return _context3.abrupt("return");
          case 2:
            supertoken = _this12.webex.credentials.supertoken; // Validate if the supertoken exists.
            if (supertoken && supertoken.access_token) {
              _this12.initServiceCatalogs().then(function () {
                catalog.isReady = true;
              }).catch(function (error) {
                _this12.initFailed = true;
                _this12.logger.error("services: failed to init initial services when credentials available, ".concat(error === null || error === void 0 ? void 0 : error.message));
              });
            } else {
              email = _this12.webex.config.email;
              _this12.collectPreauthCatalog(email ? {
                email: email
              } : undefined).catch(function (error) {
                _this12.initFailed = true;
                _this12.logger.error("services: failed to init initial services when no credentials available, ".concat(error === null || error === void 0 ? void 0 : error.message));
              });
            }
          case 3:
          case "end":
            return _context3.stop();
        }
      }, _callee3);
    })));
  },
  version: "3.12.0"
});
/* eslint-enable no-underscore-dangle */
var _default = exports.default = Services;
