// @webex/calling/dist/CallSettings/UcmBackendConnector.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.UcmBackendConnector = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs2/regenerator"));
var _promise = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/promise"));
var _stringify = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/json/stringify"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));
var _Logger = _interopRequireDefault(require("../Logger"));
var _SDKConnector = _interopRequireDefault(require("../SDKConnector"));
var _Utils = require("../common/Utils");
var _constants = require("../common/constants");
var _types = require("../common/types");
var _constants2 = require("./constants");
/**
 * This Connector class will implement child interface of ICallSettings and
 * has methods for the CCUC with UCM backend.
 */
var UcmBackendConnector = exports.UcmBackendConnector = /*#__PURE__*/function () {
  /**
   * @param useProdWebexApis - default value is true
   */
  function UcmBackendConnector(webex, logger) {
    var useProdWebexApis = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
    (0, _classCallCheck2.default)(this, UcmBackendConnector);
    (0, _defineProperty2.default)(this, "sdkConnector", void 0);
    (0, _defineProperty2.default)(this, "webex", void 0);
    (0, _defineProperty2.default)(this, "userId", void 0);
    (0, _defineProperty2.default)(this, "orgId", void 0);
    (0, _defineProperty2.default)(this, "useProdWebexApis", void 0);
    this.sdkConnector = _SDKConnector.default;
    if (!this.sdkConnector.getWebex()) {
      _SDKConnector.default.setWebex(webex);
    }
    this.webex = this.sdkConnector.getWebex();
    _Logger.default.setLogger(logger.level, _constants.UCM_CONNECTOR_FILE);
    this.userId = this.webex.internal.device.userId;
    this.orgId = this.webex.internal.device.orgId;
    this.useProdWebexApis = useProdWebexApis;
  }

  // Removed header caching to align with review: fetch per request instead.

  /**
   * Reads call waiting setting at the backend.
   */
  return (0, _createClass2.default)(UcmBackendConnector, [{
    key: "getCallWaitingSetting",
    value: function getCallWaitingSetting() {
      var loggerContext = {
        file: _constants.UCM_CONNECTOR_FILE,
        method: _constants2.METHODS.GET_CALL_WAITING_SETTING
      };
      _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
      return this.getMethodNotSupportedResponse();
    }

    /**
     * Reads DND setting at the backend.
     */
  }, {
    key: "getDoNotDisturbSetting",
    value: function getDoNotDisturbSetting() {
      var loggerContext = {
        file: _constants.UCM_CONNECTOR_FILE,
        method: _constants2.METHODS.GET_DO_NOT_DISTURB_SETTING
      };
      _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
      return this.getMethodNotSupportedResponse();
    }

    /**
     * Updates DND setting at the backend.
     */
  }, {
    key: "setDoNotDisturbSetting",
    value: function setDoNotDisturbSetting() {
      var loggerContext = {
        file: _constants.UCM_CONNECTOR_FILE,
        method: _constants2.METHODS.SET_DO_NOT_DISTURB_SETTING
      };
      _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
      return this.getMethodNotSupportedResponse();
    }

    /**
     * Reads Call Forward setting at the backend.
     */
  }, {
    key: "getCallForwardSetting",
    value: function getCallForwardSetting() {
      var loggerContext = {
        file: _constants.UCM_CONNECTOR_FILE,
        method: _constants2.METHODS.GET_CALL_FORWARD_SETTING
      };
      _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
      return this.getMethodNotSupportedResponse();
    }

    /**
     * Updates Call Forward setting at the backend.
     */
  }, {
    key: "setCallForwardSetting",
    value: function setCallForwardSetting() {
      var loggerContext = {
        file: _constants.UCM_CONNECTOR_FILE,
        method: _constants2.METHODS.SET_CALL_FORWARD_SETTING
      };
      _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
      return this.getMethodNotSupportedResponse();
    }

    /**
     * Reads Voicemail setting at the backend.
     */
  }, {
    key: "getVoicemailSetting",
    value: function getVoicemailSetting() {
      var loggerContext = {
        file: _constants.UCM_CONNECTOR_FILE,
        method: _constants2.METHODS.GET_VOICEMAIL_SETTING
      };
      _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
      return this.getMethodNotSupportedResponse();
    }

    /**
     * Updates Voicemail setting at the backend.
     */
  }, {
    key: "setVoicemailSetting",
    value: function setVoicemailSetting() {
      var loggerContext = {
        file: _constants.UCM_CONNECTOR_FILE,
        method: _constants2.METHODS.SET_VOICEMAIL_SETTING
      };
      _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
      return this.getMethodNotSupportedResponse();
    }

    /**
     * Returns a default error response for unsupported methods.
     */
  }, {
    key: "getMethodNotSupportedResponse",
    value: function getMethodNotSupportedResponse() {
      var loggerContext = {
        file: _constants.UCM_CONNECTOR_FILE,
        method: _constants2.METHODS.GET_METHOD_NOT_SUPPORTED_RESPONSE
      };
      _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
      var response = (0, _Utils.serviceErrorCodeHandler)({
        statusCode: 501
      }, loggerContext);
      return _promise.default.resolve(response);
    }

    /**
     * Reads the Call Forwarding Always settings at the backend.
     * This will also check if CFA is set to Voicemail.
     * If CFA is set to destination, that will take precedence.
     * For UCM backend, relevant fields in the response are `enabled` & `destination`.
     *
     * @param directoryNumber - Directory number for which CFA needs to returned.
     */
  }, {
    key: "getCallForwardAlwaysSetting",
    value: (function () {
      var _getCallForwardAlwaysSetting = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee(directoryNumber) {
        var _this$webex, _this$webex$config;
        var loggerContext, headers, webexApisUrl, requestOptions, resp, _ref, callForwarding, cfa, _response, _response2, response, errorInfo, errorStatus, _t;
        return _regenerator.default.wrap(function (_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              loggerContext = {
                file: _constants.UCM_CONNECTOR_FILE,
                method: _constants2.METHODS.GET_CALL_FORWARD_ALWAYS_SETTING
              };
              _Logger.default.info(directoryNumber ? "".concat(_constants.METHOD_START_MESSAGE, " with ").concat(directoryNumber) : _constants.METHOD_START_MESSAGE, loggerContext);

              // Fetch auth headers per request (no caching) to ensure freshness.
              _context.next = 1;
              return this.getAuthHeaders();
            case 1:
              headers = _context.sent;
              if ((_this$webex = this.webex) !== null && _this$webex !== void 0 && (_this$webex$config = _this$webex.config) !== null && _this$webex$config !== void 0 && _this$webex$config.fedramp) {
                webexApisUrl = _constants.WEBEX_API_CONFIG_FEDRAMP_URL;
              } else {
                webexApisUrl = this.useProdWebexApis ? _constants.WEBEX_API_CONFIG_PROD_URL : _constants.WEBEX_API_CONFIG_INT_URL;
              }
              _context.prev = 2;
              if (!directoryNumber) {
                _context.next = 5;
                break;
              }
              requestOptions = {
                uri: "".concat(webexApisUrl, "/").concat(_constants2.PEOPLE_ENDPOINT, "/").concat(this.userId, "/").concat(_constants2.CF_ENDPOINT.toLowerCase(), "?").concat(_constants2.ORG_ENDPOINT, "=").concat(this.orgId),
                method: _types.HTTP_METHODS.GET,
                headers: headers
              };
              _context.next = 3;
              return this.webex.request(requestOptions);
            case 3:
              resp = _context.sent;
              _Logger.default.log("Response code: ".concat(resp.statusCode), loggerContext);
              _ref = resp.body, callForwarding = _ref.callForwarding;
              cfa = callForwarding.always.find(function (item) {
                return item.dn.endsWith(directoryNumber) || item.e164Number.endsWith(directoryNumber);
              });
              if (!cfa) {
                _context.next = 4;
                break;
              }
              _response = {
                statusCode: Number(resp[_constants.STATUS_CODE]),
                message: _constants.SUCCESS_MESSAGE,
                data: {
                  callSetting: {
                    enabled: cfa.destinationVoicemailEnabled || !!cfa.destination,
                    destination: cfa.destinationVoicemailEnabled ? _constants.VOICEMAIL : cfa.destination
                  }
                }
              };
              _Logger.default.log("Successfully retrieved call forward always setting for directory number: ".concat(directoryNumber), loggerContext);
              return _context.abrupt("return", _response);
            case 4:
              _response2 = {
                statusCode: 404,
                message: _constants.FAILURE_MESSAGE,
                data: {
                  error: 'Directory Number is not assigned to the user'
                }
              };
              return _context.abrupt("return", _response2);
            case 5:
              response = {
                statusCode: 400,
                message: _constants.FAILURE_MESSAGE,
                data: {
                  error: 'Directory Number is mandatory for UCM backend'
                }
              };
              return _context.abrupt("return", response);
            case 6:
              _context.prev = 6;
              _t = _context["catch"](2);
              errorInfo = _t;
              _Logger.default.error("Failed to get call forward always setting: ".concat((0, _stringify.default)(_t)), loggerContext);
              _context.next = 7;
              return (0, _Utils.uploadLogs)();
            case 7:
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context.abrupt("return", errorStatus);
            case 8:
            case "end":
              return _context.stop();
          }
        }, _callee, this, [[2, 6]]);
      }));
      function getCallForwardAlwaysSetting(_x) {
        return _getCallForwardAlwaysSetting.apply(this, arguments);
      }
      return getCallForwardAlwaysSetting;
    }())
  }, {
    key: "getAuthHeaders",
    value: function () {
      var _getAuthHeaders = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee2() {
        var _this$webex2, _this$webex2$config;
        var headers;
        return _regenerator.default.wrap(function (_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              headers = {}; // Match behavior from WxCallBackendConnector:
              // Only add authorization for FedRAMP, else rely on implicit auth/session.
              // Note: Use lowercase 'authorization' to match SDK's auth interceptor check
              if (!((_this$webex2 = this.webex) !== null && _this$webex2 !== void 0 && (_this$webex2$config = _this$webex2.config) !== null && _this$webex2$config !== void 0 && _this$webex2$config.fedramp)) {
                _context2.next = 2;
                break;
              }
              _context2.next = 1;
              return this.webex.credentials.getUserToken();
            case 1:
              headers.authorization = _context2.sent;
            case 2:
              return _context2.abrupt("return", headers);
            case 3:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function getAuthHeaders() {
        return _getAuthHeaders.apply(this, arguments);
      }
      return getAuthHeaders;
    }()
  }]);
}();
