// @webex/calling/dist/CallSettings/WxCallBackendConnector.js
"use strict";

var _Object$keys = require("@babel/runtime-corejs2/core-js/object/keys");
var _Object$getOwnPropertySymbols = require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols");
var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor");
var _Object$getOwnPropertyDescriptors = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptors");
var _Object$defineProperties = require("@babel/runtime-corejs2/core-js/object/define-properties");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.WxCallBackendConnector = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs2/regenerator"));
var _stringify = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/json/stringify"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));
var _SDKConnector = _interopRequireDefault(require("../SDKConnector"));
var _Logger = _interopRequireDefault(require("../Logger"));
var _types = require("../common/types");
var _constants = require("../common/constants");
var _constants2 = require("./constants");
var _Utils = require("../common/Utils");
function ownKeys(e, r) { var t = _Object$keys(e); if (_Object$getOwnPropertySymbols) { var o = _Object$getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return _Object$getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(e, _Object$getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { _Object$defineProperty(e, r, _Object$getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * This Connector class will implement child interface of ICallSettings and
 * has methods for the Webex Calling backend.
 */
var WxCallBackendConnector = exports.WxCallBackendConnector = /*#__PURE__*/function () {
  function WxCallBackendConnector(webex, logger) {
    var _this$webex$internal$;
    (0, _classCallCheck2.default)(this, WxCallBackendConnector);
    (0, _defineProperty2.default)(this, "sdkConnector", void 0);
    (0, _defineProperty2.default)(this, "webex", void 0);
    (0, _defineProperty2.default)(this, "userId", void 0);
    (0, _defineProperty2.default)(this, "personId", void 0);
    (0, _defineProperty2.default)(this, "orgId", void 0);
    (0, _defineProperty2.default)(this, "xsiEndpoint", void 0);
    (0, _defineProperty2.default)(this, "hydraEndpoint", void 0);
    (0, _defineProperty2.default)(this, "VOICEMAIL", 'VOICEMAIL');
    this.sdkConnector = _SDKConnector.default;
    if (!this.sdkConnector.getWebex()) {
      _SDKConnector.default.setWebex(webex);
    }
    this.webex = this.sdkConnector.getWebex();
    /* eslint no-underscore-dangle: 0 */
    this.hydraEndpoint = ((_this$webex$internal$ = this.webex.internal.services._serviceUrls) === null || _this$webex$internal$ === void 0 ? void 0 : _this$webex$internal$.hydra) || this.webex.internal.services.get(this.webex.internal.services._activeServices.hydra);
    _Logger.default.setLogger(logger.level, _constants.WEBEX_CALLING_CONNECTOR_FILE);
    this.userId = this.webex.internal.device.userId;
    this.personId = (0, _Utils.inferIdFromUuid)(this.webex.internal.device.userId, _types.DecodeType.PEOPLE);
    this.orgId = (0, _Utils.inferIdFromUuid)(this.webex.internal.device.orgId, _types.DecodeType.ORGANIZATION);
  }

  /**
   * Reads call waiting setting at the backend.
   */
  return (0, _createClass2.default)(WxCallBackendConnector, [{
    key: "getCallWaitingSetting",
    value: (function () {
      var _getCallWaitingSetting = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee() {
        var loggerContext, _response$headers, callWaitingUrl, response, xmlData, parser, xmlDOM, callWaitingDetails, status, toggleSetting, responseDetails, errorInfo, errorStatus, _t, _t2, _t3, _t4, _t5, _t6;
        return _regenerator.default.wrap(function (_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: _constants2.METHODS.GET_CALL_WAITING_SETTING
              };
              _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
              _context.prev = 1;
              if (this.xsiEndpoint) {
                _context.next = 3;
                break;
              }
              _context.next = 2;
              return (0, _Utils.getXsiActionEndpoint)(this.webex, loggerContext, _types.CALLING_BACKEND.WXC);
            case 2:
              this.xsiEndpoint = _context.sent;
              _Logger.default.info("xsiEndpoint: ".concat(this.xsiEndpoint), loggerContext);
            case 3:
              callWaitingUrl = "".concat(this.xsiEndpoint, "/").concat(_constants2.XSI_VERSION, "/").concat(_constants2.USER_ENDPOINT, "/").concat(this.userId, "/").concat(_constants.SERVICES_ENDPOINT, "/").concat(_constants2.CALL_WAITING_ENDPOINT);
              _t = fetch;
              _t2 = "".concat(callWaitingUrl);
              _t3 = _types.HTTP_METHODS.GET;
              _context.next = 4;
              return this.webex.credentials.getUserToken();
            case 4:
              _t4 = _context.sent;
              _t5 = {
                Authorization: _t4
              };
              _context.next = 5;
              return _t(_t2, {
                method: _t3,
                headers: _t5
              });
            case 5:
              response = _context.sent;
              if (response.ok) {
                _context.next = 6;
                break;
              }
              throw new Error("".concat(response.status));
            case 6:
              _Logger.default.log("Response trackingId: ".concat((_response$headers = response.headers) === null || _response$headers === void 0 ? void 0 : _response$headers.get('trackingid')), loggerContext);
              _context.next = 7;
              return response.text();
            case 7:
              xmlData = _context.sent;
              parser = new DOMParser();
              xmlDOM = parser.parseFromString(xmlData, _constants.XML_TYPE);
              callWaitingDetails = xmlDOM.getElementsByTagName('active');
              status = callWaitingDetails[0].childNodes[0].textContent;
              toggleSetting = {
                enabled: status === 'true'
              };
              responseDetails = {
                statusCode: 200,
                data: {
                  callSetting: toggleSetting
                },
                message: _constants.SUCCESS_MESSAGE
              };
              _Logger.default.log("Successfully got Call Waiting settings, personId: ".concat(this.personId), loggerContext);
              return _context.abrupt("return", responseDetails);
            case 8:
              _context.prev = 8;
              _t6 = _context["catch"](1);
              _Logger.default.error("Failed to get call waiting setting: ".concat((0, _stringify.default)(_t6)), loggerContext);
              _context.next = 9;
              return (0, _Utils.uploadLogs)();
            case 9:
              errorInfo = {
                statusCode: _t6 instanceof Error ? Number(_t6.message) : ''
              };
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context.abrupt("return", errorStatus);
            case 10:
            case "end":
              return _context.stop();
          }
        }, _callee, this, [[1, 8]]);
      }));
      function getCallWaitingSetting() {
        return _getCallWaitingSetting.apply(this, arguments);
      }
      return getCallWaitingSetting;
    }()
    /**
     * Reads DND setting at the backend.
     */
    )
  }, {
    key: "getDoNotDisturbSetting",
    value: (function () {
      var _getDoNotDisturbSetting = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee2() {
        var loggerContext, _resp$headers, resp, dndSettingResponse, responseDetails, errorInfo, errorStatus, _t7;
        return _regenerator.default.wrap(function (_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: _constants2.METHODS.GET_DO_NOT_DISTURB_SETTING
              };
              _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
              _context2.prev = 1;
              _context2.next = 2;
              return this.webex.request({
                uri: "".concat(this.hydraEndpoint, "/").concat(_constants2.PEOPLE_ENDPOINT, "/").concat(this.personId, "/").concat(_constants2.DND_ENDPOINT, "?").concat(_constants2.ORG_ENDPOINT, "=").concat(this.orgId),
                method: _types.HTTP_METHODS.GET
              });
            case 2:
              resp = _context2.sent;
              _Logger.default.log("Response trackingId: ".concat(resp === null || resp === void 0 ? void 0 : (_resp$headers = resp.headers) === null || _resp$headers === void 0 ? void 0 : _resp$headers.trackingid), loggerContext);
              dndSettingResponse = resp.body;
              responseDetails = {
                statusCode: Number(resp[_constants.STATUS_CODE]),
                data: {
                  callSetting: dndSettingResponse
                },
                message: _constants.SUCCESS_MESSAGE
              };
              _Logger.default.log("Successfully got Do Not Disturb settings, personId: ".concat(this.personId), loggerContext);
              return _context2.abrupt("return", responseDetails);
            case 3:
              _context2.prev = 3;
              _t7 = _context2["catch"](1);
              _Logger.default.error("Failed to get DoNotDisturb setting: ".concat((0, _stringify.default)(_t7)), loggerContext);
              _context2.next = 4;
              return (0, _Utils.uploadLogs)();
            case 4:
              errorInfo = _t7;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context2.abrupt("return", errorStatus);
            case 5:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this, [[1, 3]]);
      }));
      function getDoNotDisturbSetting() {
        return _getDoNotDisturbSetting.apply(this, arguments);
      }
      return getDoNotDisturbSetting;
    }()
    /**
     * Updates DND setting at the backend.
     */
    )
  }, {
    key: "setDoNotDisturbSetting",
    value: (function () {
      var _setDoNotDisturbSetting = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee3(enabled) {
        var loggerContext, _resp$headers2, dndRequestBody, resp, responseDetails, errorInfo, errorStatus, _t8;
        return _regenerator.default.wrap(function (_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: _constants2.METHODS.SET_DO_NOT_DISTURB_SETTING
              };
              _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
              _context3.prev = 1;
              dndRequestBody = {
                enabled: enabled,
                ringSplashEnabled: false
              };
              _context3.next = 2;
              return this.webex.request({
                uri: "".concat(this.hydraEndpoint, "/").concat(_constants2.PEOPLE_ENDPOINT, "/").concat(this.personId, "/").concat(_constants2.DND_ENDPOINT, "?").concat(_constants2.ORG_ENDPOINT, "=").concat(this.orgId),
                method: _types.HTTP_METHODS.PUT,
                body: dndRequestBody
              });
            case 2:
              resp = _context3.sent;
              _Logger.default.log("Response trackingId: ".concat(resp === null || resp === void 0 ? void 0 : (_resp$headers2 = resp.headers) === null || _resp$headers2 === void 0 ? void 0 : _resp$headers2.trackingid), loggerContext);
              responseDetails = {
                statusCode: Number(resp[_constants.STATUS_CODE]),
                data: {
                  callSetting: dndRequestBody
                },
                message: _constants.SUCCESS_MESSAGE
              };
              _Logger.default.log("Successfully set Do Not Disturb settings, personId: ".concat(this.personId), loggerContext);
              return _context3.abrupt("return", responseDetails);
            case 3:
              _context3.prev = 3;
              _t8 = _context3["catch"](1);
              _Logger.default.error("Failed to set DoNotDisturb setting: ".concat((0, _stringify.default)(_t8)), loggerContext);
              _context3.next = 4;
              return (0, _Utils.uploadLogs)();
            case 4:
              errorInfo = _t8;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context3.abrupt("return", errorStatus);
            case 5:
            case "end":
              return _context3.stop();
          }
        }, _callee3, this, [[1, 3]]);
      }));
      function setDoNotDisturbSetting(_x) {
        return _setDoNotDisturbSetting.apply(this, arguments);
      }
      return setDoNotDisturbSetting;
    }()
    /**
     * Reads Call Forward setting at the backend.
     */
    )
  }, {
    key: "getCallForwardSetting",
    value: (function () {
      var _getCallForwardSetting = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee4() {
        var loggerContext, _resp$headers3, resp, cfResponse, responseDetails, errorInfo, errorStatus, _t9;
        return _regenerator.default.wrap(function (_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: _constants2.METHODS.GET_CALL_FORWARD_SETTING
              };
              _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
              _context4.prev = 1;
              _context4.next = 2;
              return this.webex.request({
                uri: "".concat(this.hydraEndpoint, "/").concat(_constants2.PEOPLE_ENDPOINT, "/").concat(this.personId, "/").concat(_constants2.CF_ENDPOINT, "?").concat(_constants2.ORG_ENDPOINT, "=").concat(this.orgId),
                method: _types.HTTP_METHODS.GET
              });
            case 2:
              resp = _context4.sent;
              _Logger.default.log("Response trackingId: ".concat(resp === null || resp === void 0 ? void 0 : (_resp$headers3 = resp.headers) === null || _resp$headers3 === void 0 ? void 0 : _resp$headers3.trackingid), loggerContext);
              cfResponse = resp.body;
              responseDetails = {
                statusCode: Number(resp[_constants.STATUS_CODE]),
                data: {
                  callSetting: cfResponse
                },
                message: _constants.SUCCESS_MESSAGE
              };
              _Logger.default.log("Successfully got Call Forward settings, personId: ".concat(this.personId), loggerContext);
              return _context4.abrupt("return", responseDetails);
            case 3:
              _context4.prev = 3;
              _t9 = _context4["catch"](1);
              _Logger.default.error("Failed to get Call Forward setting: ".concat((0, _stringify.default)(_t9)), loggerContext);
              _context4.next = 4;
              return (0, _Utils.uploadLogs)();
            case 4:
              errorInfo = _t9;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context4.abrupt("return", errorStatus);
            case 5:
            case "end":
              return _context4.stop();
          }
        }, _callee4, this, [[1, 3]]);
      }));
      function getCallForwardSetting() {
        return _getCallForwardSetting.apply(this, arguments);
      }
      return getCallForwardSetting;
    }()
    /**
     * Updates Call Forward setting at the backend.
     */
    )
  }, {
    key: "setCallForwardSetting",
    value: (function () {
      var _setCallForwardSetting = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee5(callForwardingRequest) {
        var loggerContext, _resp$headers4, resp, responseDetails, errorInfo, errorStatus, _t0;
        return _regenerator.default.wrap(function (_context5) {
          while (1) switch (_context5.prev = _context5.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: _constants2.METHODS.SET_CALL_FORWARD_SETTING
              };
              _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
              _context5.prev = 1;
              _context5.next = 2;
              return this.webex.request({
                uri: "".concat(this.hydraEndpoint, "/").concat(_constants2.PEOPLE_ENDPOINT, "/").concat(this.personId, "/").concat(_constants2.CF_ENDPOINT, "?").concat(_constants2.ORG_ENDPOINT, "=").concat(this.orgId),
                method: _types.HTTP_METHODS.PUT,
                body: callForwardingRequest
              });
            case 2:
              resp = _context5.sent;
              _Logger.default.log("Response trackingId: ".concat(resp === null || resp === void 0 ? void 0 : (_resp$headers4 = resp.headers) === null || _resp$headers4 === void 0 ? void 0 : _resp$headers4.trackingid), loggerContext);
              responseDetails = {
                statusCode: Number(resp[_constants.STATUS_CODE]),
                data: {
                  callSetting: callForwardingRequest
                },
                message: _constants.SUCCESS_MESSAGE
              };
              _Logger.default.log("Successfully set Call Forward settings, personId: ".concat(this.personId), loggerContext);
              return _context5.abrupt("return", responseDetails);
            case 3:
              _context5.prev = 3;
              _t0 = _context5["catch"](1);
              _Logger.default.error("Failed to set Call Forward setting: ".concat((0, _stringify.default)(_t0)), loggerContext);
              _context5.next = 4;
              return (0, _Utils.uploadLogs)();
            case 4:
              errorInfo = _t0;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context5.abrupt("return", errorStatus);
            case 5:
            case "end":
              return _context5.stop();
          }
        }, _callee5, this, [[1, 3]]);
      }));
      function setCallForwardSetting(_x2) {
        return _setCallForwardSetting.apply(this, arguments);
      }
      return setCallForwardSetting;
    }()
    /**
     * Reads Voicemail setting at the backend.
     */
    )
  }, {
    key: "getVoicemailSetting",
    value: (function () {
      var _getVoicemailSetting = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee6() {
        var loggerContext, _resp$headers5, resp, vmResponse, responseDetails, errorInfo, errorStatus, _t1;
        return _regenerator.default.wrap(function (_context6) {
          while (1) switch (_context6.prev = _context6.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: _constants2.METHODS.GET_VOICEMAIL_SETTING
              };
              _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
              _context6.prev = 1;
              _context6.next = 2;
              return this.webex.request({
                uri: "".concat(this.hydraEndpoint, "/").concat(_constants2.PEOPLE_ENDPOINT, "/").concat(this.personId, "/").concat(_constants2.VM_ENDPOINT, "?").concat(_constants2.ORG_ENDPOINT, "=").concat(this.orgId),
                method: _types.HTTP_METHODS.GET
              });
            case 2:
              resp = _context6.sent;
              _Logger.default.log("Response trackingId: ".concat(resp === null || resp === void 0 ? void 0 : (_resp$headers5 = resp.headers) === null || _resp$headers5 === void 0 ? void 0 : _resp$headers5.trackingid), loggerContext);
              vmResponse = resp.body;
              responseDetails = {
                statusCode: Number(resp[_constants.STATUS_CODE]),
                data: {
                  callSetting: vmResponse
                },
                message: _constants.SUCCESS_MESSAGE
              };
              _Logger.default.log("Successfully got Voicemail settings, personId: ".concat(this.personId), loggerContext);
              return _context6.abrupt("return", responseDetails);
            case 3:
              _context6.prev = 3;
              _t1 = _context6["catch"](1);
              _Logger.default.error("Failed to get Voicemail setting: ".concat((0, _stringify.default)(_t1)), loggerContext);
              _context6.next = 4;
              return (0, _Utils.uploadLogs)();
            case 4:
              errorInfo = _t1;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context6.abrupt("return", errorStatus);
            case 5:
            case "end":
              return _context6.stop();
          }
        }, _callee6, this, [[1, 3]]);
      }));
      function getVoicemailSetting() {
        return _getVoicemailSetting.apply(this, arguments);
      }
      return getVoicemailSetting;
    }()
    /**
     * Updates Voicemail setting at the backend.
     */
    )
  }, {
    key: "setVoicemailSetting",
    value: (function () {
      var _setVoicemailSetting = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee7(voicemailRequest) {
        var loggerContext, _resp$headers6, resp, responseDetails, errorInfo, errorStatus, _t10;
        return _regenerator.default.wrap(function (_context7) {
          while (1) switch (_context7.prev = _context7.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: _constants2.METHODS.SET_VOICEMAIL_SETTING
              };
              _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
              _context7.prev = 1;
              _context7.next = 2;
              return this.webex.request({
                uri: "".concat(this.hydraEndpoint, "/").concat(_constants2.PEOPLE_ENDPOINT, "/").concat(this.personId, "/").concat(_constants2.VM_ENDPOINT, "?").concat(_constants2.ORG_ENDPOINT, "=").concat(this.orgId),
                method: _types.HTTP_METHODS.PUT,
                body: voicemailRequest
              });
            case 2:
              resp = _context7.sent;
              _Logger.default.log("Response trackingId: ".concat(resp === null || resp === void 0 ? void 0 : (_resp$headers6 = resp.headers) === null || _resp$headers6 === void 0 ? void 0 : _resp$headers6.trackingid), loggerContext);
              responseDetails = {
                statusCode: Number(resp[_constants.STATUS_CODE]),
                data: {
                  callSetting: voicemailRequest
                },
                message: _constants.SUCCESS_MESSAGE
              };
              _Logger.default.log("Successfully set Voicemail settings, personId: ".concat(this.personId), loggerContext);
              return _context7.abrupt("return", responseDetails);
            case 3:
              _context7.prev = 3;
              _t10 = _context7["catch"](1);
              _Logger.default.error("Failed to set Voicemail setting: ".concat((0, _stringify.default)(_t10)), loggerContext);
              _context7.next = 4;
              return (0, _Utils.uploadLogs)();
            case 4:
              errorInfo = _t10;
              errorStatus = (0, _Utils.serviceErrorCodeHandler)(errorInfo, loggerContext);
              return _context7.abrupt("return", errorStatus);
            case 5:
            case "end":
              return _context7.stop();
          }
        }, _callee7, this, [[1, 3]]);
      }));
      function setVoicemailSetting(_x3) {
        return _setVoicemailSetting.apply(this, arguments);
      }
      return setVoicemailSetting;
    }()
    /**
     * Reads the Call Forwarding Always settings at the backend.
     * This will also check if CFA is set to Voicemail.
     * If CFA is set to destination, that will take precedence.
     */
    )
  }, {
    key: "getCallForwardAlwaysSetting",
    value: (function () {
      var _getCallForwardAlwaysSetting = (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee8() {
        var loggerContext, cfResponse, cfa, response, vmResponse, _vm$sendAllCalls, vm, _response, _response2;
        return _regenerator.default.wrap(function (_context8) {
          while (1) switch (_context8.prev = _context8.next) {
            case 0:
              loggerContext = {
                file: _constants.WEBEX_CALLING_CONNECTOR_FILE,
                method: _constants2.METHODS.GET_CALL_FORWARD_ALWAYS_SETTING
              };
              _Logger.default.info(_constants.METHOD_START_MESSAGE, loggerContext);
              _context8.next = 1;
              return this.getCallForwardSetting();
            case 1:
              cfResponse = _context8.sent;
              if (!(cfResponse.statusCode === 200)) {
                _context8.next = 7;
                break;
              }
              cfa = cfResponse.data.callSetting.callForwarding.always;
              /** CFA is set to destination */
              if (!cfa.enabled) {
                _context8.next = 3;
                break;
              }
              if (!cfa.destination) {
                _context8.next = 2;
                break;
              }
              response = _objectSpread(_objectSpread({}, cfResponse), {}, {
                data: {
                  callSetting: cfa
                }
              });
              return _context8.abrupt("return", response);
            case 2:
              _Logger.default.warn("CFA is enabled, but destination is not set", loggerContext);
            case 3:
              _context8.next = 4;
              return this.getVoicemailSetting();
            case 4:
              vmResponse = _context8.sent;
              if (!(vmResponse.statusCode === 200)) {
                _context8.next = 6;
                break;
              }
              vm = vmResponse.data.callSetting;
              /** CFA is enabled to voicemail */
              if (!(vm.enabled && (_vm$sendAllCalls = vm.sendAllCalls) !== null && _vm$sendAllCalls !== void 0 && _vm$sendAllCalls.enabled)) {
                _context8.next = 5;
                break;
              }
              _response = _objectSpread(_objectSpread({}, cfResponse), {}, {
                data: {
                  callSetting: _objectSpread(_objectSpread({}, cfa), {}, {
                    enabled: true,
                    destination: this.VOICEMAIL
                  })
                }
              });
              return _context8.abrupt("return", _response);
            case 5:
              /** No CFA is set */
              _response2 = _objectSpread(_objectSpread({}, cfResponse), {}, {
                data: {
                  callSetting: _objectSpread(_objectSpread({}, cfa), {}, {
                    enabled: false,
                    destination: undefined
                  })
                }
              });
              return _context8.abrupt("return", _response2);
            case 6:
              _Logger.default.warn("Unable to retrieve voicemail settings.", loggerContext);
              vmResponse.data.error = 'Unable to retrieve voicemail settings.';
              return _context8.abrupt("return", vmResponse);
            case 7:
              return _context8.abrupt("return", cfResponse);
            case 8:
            case "end":
              return _context8.stop();
          }
        }, _callee8, this);
      }));
      function getCallForwardAlwaysSetting() {
        return _getCallForwardAlwaysSetting.apply(this, arguments);
      }
      return getCallForwardAlwaysSetting;
    }())
  }]);
}();
