// @webex/webex-core/dist/lib/credentials/grant-errors.js
"use strict";

var _Reflect$construct = require("@babel/runtime-corejs2/core-js/reflect/construct");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.OAuthError = void 0;
var _defineProperties = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/define-properties"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));
var _common = require("@webex/common");
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); } /*!
 * Copyright (c) 2015-2020 Cisco Systems, Inc. See LICENSE file.
 */ /* eslint camelcase: [0] */
/**
 * Error thrown during oauth flow
 */
var OAuthError = exports.OAuthError = /*#__PURE__*/function (_Exception) {
  function OAuthError() {
    (0, _classCallCheck2.default)(this, OAuthError);
    return _callSuper(this, OAuthError, arguments);
  }
  (0, _inherits2.default)(OAuthError, _Exception);
  return (0, _createClass2.default)(OAuthError, [{
    key: "parse",
    value:
    /**
     * @param {HttpResponse} res
     * @returns {string}
     */
    function parse(res) {
      // This error may be returned in an HTTPResponse or as part of a querystring
      var body = res.body || res;
      (0, _defineProperties.default)(this, {
        error: {
          enumerable: true,
          value: body.error
        },
        errorDescription: {
          enumerable: true,
          value: body.error_description
        },
        errorUri: {
          enumerable: true,
          value: body.error_uri
        },
        res: {
          enumerable: false,
          value: res
        }
      });
      return this.errorDescription;
    }
  }]);
}(_common.Exception);
/**
 * InvalidRequestError
 */
var InvalidRequestError = /*#__PURE__*/function (_OAuthError) {
  function InvalidRequestError() {
    (0, _classCallCheck2.default)(this, InvalidRequestError);
    return _callSuper(this, InvalidRequestError, arguments);
  }
  (0, _inherits2.default)(InvalidRequestError, _OAuthError);
  return (0, _createClass2.default)(InvalidRequestError);
}(OAuthError);
/**
 * InvalidClientError
 */
var InvalidClientError = /*#__PURE__*/function (_OAuthError2) {
  function InvalidClientError() {
    (0, _classCallCheck2.default)(this, InvalidClientError);
    return _callSuper(this, InvalidClientError, arguments);
  }
  (0, _inherits2.default)(InvalidClientError, _OAuthError2);
  return (0, _createClass2.default)(InvalidClientError);
}(OAuthError);
/**
 * InvalidGrantError
 */
var InvalidGrantError = /*#__PURE__*/function (_OAuthError3) {
  function InvalidGrantError() {
    (0, _classCallCheck2.default)(this, InvalidGrantError);
    return _callSuper(this, InvalidGrantError, arguments);
  }
  (0, _inherits2.default)(InvalidGrantError, _OAuthError3);
  return (0, _createClass2.default)(InvalidGrantError);
}(OAuthError);
/**
 * UnauthorizedClientError
 */
var UnauthorizedClientError = /*#__PURE__*/function (_OAuthError4) {
  function UnauthorizedClientError() {
    (0, _classCallCheck2.default)(this, UnauthorizedClientError);
    return _callSuper(this, UnauthorizedClientError, arguments);
  }
  (0, _inherits2.default)(UnauthorizedClientError, _OAuthError4);
  return (0, _createClass2.default)(UnauthorizedClientError);
}(OAuthError);
/**
 * UnsupportGrantTypeError
 */
var UnsupportGrantTypeError = /*#__PURE__*/function (_OAuthError5) {
  function UnsupportGrantTypeError() {
    (0, _classCallCheck2.default)(this, UnsupportGrantTypeError);
    return _callSuper(this, UnsupportGrantTypeError, arguments);
  }
  (0, _inherits2.default)(UnsupportGrantTypeError, _OAuthError5);
  return (0, _createClass2.default)(UnsupportGrantTypeError);
}(OAuthError);
/**
 * InvalidScopeError
 */
var InvalidScopeError = /*#__PURE__*/function (_OAuthError6) {
  function InvalidScopeError() {
    (0, _classCallCheck2.default)(this, InvalidScopeError);
    return _callSuper(this, InvalidScopeError, arguments);
  }
  (0, _inherits2.default)(InvalidScopeError, _OAuthError6);
  return (0, _createClass2.default)(InvalidScopeError);
}(OAuthError);
var errors = {
  OAuthError: OAuthError,
  InvalidRequestError: InvalidRequestError,
  InvalidClientError: InvalidClientError,
  InvalidGrantError: InvalidGrantError,
  UnauthorizedClientError: UnauthorizedClientError,
  UnsupportGrantTypeError: UnsupportGrantTypeError,
  InvalidScopeError: InvalidScopeError,
  invalid_request: InvalidRequestError,
  invalid_client: InvalidClientError,
  invalid_grant: InvalidGrantError,
  unauthorized_client: UnauthorizedClientError,
  unsupported_grant_type: UnsupportGrantTypeError,
  invalid_scope: InvalidScopeError,
  select: function select(errorString) {
    return errors[errorString] || OAuthError;
  }
};
var _default = exports.default = errors;
