// @webex/webex-core/dist/lib/services/index.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
_Object$defineProperty(exports, "ServiceCatalog", {
  enumerable: true,
  get: function get() {
    return _serviceCatalog.default;
  }
});
_Object$defineProperty(exports, "ServiceHost", {
  enumerable: true,
  get: function get() {
    return _serviceHost.default;
  }
});
_Object$defineProperty(exports, "ServiceRegistry", {
  enumerable: true,
  get: function get() {
    return _serviceRegistry.default;
  }
});
_Object$defineProperty(exports, "ServiceState", {
  enumerable: true,
  get: function get() {
    return _serviceState.default;
  }
});
_Object$defineProperty(exports, "ServiceUrl", {
  enumerable: true,
  get: function get() {
    return _serviceUrl.default;
  }
});
_Object$defineProperty(exports, "Services", {
  enumerable: true,
  get: function get() {
    return _services.default;
  }
});
var _webexCore = require("../../webex-core");
var _services = _interopRequireDefault(require("./services"));
var _serverError = _interopRequireDefault(require("../interceptors/server-error"));
var _service = _interopRequireDefault(require("../interceptors/service"));
var _serviceCatalog = _interopRequireDefault(require("./service-catalog"));
var _serviceRegistry = _interopRequireDefault(require("./service-registry"));
var _serviceState = _interopRequireDefault(require("./service-state"));
var _serviceHost = _interopRequireDefault(require("./service-host"));
var _serviceUrl = _interopRequireDefault(require("./service-url"));
/*!
 * Copyright (c) 2015-2020 Cisco Systems, Inc. See LICENSE file.
 */

(0, _webexCore.registerInternalPlugin)('services', _services.default, {
  interceptors: {
    ServiceInterceptor: _service.default.create,
    ServerErrorInterceptor: _serverError.default.create
  }
});
