// @webex/webex-core/dist/lib/services/services.js
"use strict";

var _Object$keys2 = require("@babel/runtime-corejs2/core-js/object/keys");
var _Object$getOwnPropertySymbols = require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols");
var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor");
var _Object$getOwnPropertyDescriptors = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptors");
var _Object$defineProperties = require("@babel/runtime-corejs2/core-js/object/define-properties");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DEFAULT_CLUSTER_SERVICE = exports.DEFAULT_CLUSTER = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs2/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/asyncToGenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/slicedToArray"));
var _weakMap = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/weak-map"));
var _entries = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/entries"));
var _keys = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/keys"));
var _promise = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/promise"));
var _now = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/date/now"));
var _stringify = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/json/stringify"));
var _sha = _interopRequireDefault(require("crypto-js/sha256"));
var _lodash = require("lodash");
var _webexPlugin = _interopRequireDefault(require("../webex-plugin"));
var _metrics = _interopRequireDefault(require("../metrics"));
var _serviceCatalog = _interopRequireDefault(require("./service-catalog"));
var _serviceRegistry = _interopRequireDefault(require("./service-registry"));
var _serviceState = _interopRequireDefault(require("./service-state"));
var _serviceFedRamp = _interopRequireDefault(require("./service-fed-ramp"));
var _constants = require("../constants");
function ownKeys(e, r) { var t = _Object$keys2(e); if (_Object$getOwnPropertySymbols) { var o = _Object$getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return _Object$getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(e, _Object$getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { _Object$defineProperty(e, r, _Object$getOwnPropertyDescriptor(t, r)); }); } return e; }
var trailingSlashes = /(?:^\/)|(?:\/$)/;

// The default cluster when one is not provided (usually as 'US' from hydra)
var DEFAULT_CLUSTER = exports.DEFAULT_CLUSTER = 'urn:TEAM:us-east-2_a';
// The default service name for convo (currently identityLookup due to some weird CSB issue)
var DEFAULT_CLUSTER_SERVICE = exports.DEFAULT_CLUSTER_SERVICE = 'identityLookup';
var CLUSTER_SERVICE = process.env.WEBEX_CONVERSATION_CLUSTER_SERVICE || DEFAULT_CLUSTER_SERVICE;
var DEFAULT_CLUSTER_IDENTIFIER = process.env.WEBEX_CONVERSATION_DEFAULT_CLUSTER || "".concat(DEFAULT_CLUSTER, ":").concat(CLUSTER_SERVICE);
var CATALOG_CACHE_KEY_V1 = 'services.v1.u2cHostMap';
var CATALOG_TTL_MS = 24 * 60 * 60 * 1000; // 24 hours

/* eslint-disable no-underscore-dangle */
/**
 * @class
 */
var Services = _webexPlugin.default.extend({
  namespace: 'Services',
  /**
   * The {@link WeakMap} of {@link ServiceRegistry} class instances that are
   * keyed with WebexCore instances.
   *
   * @instance
   * @type {WeakMap<WebexCore, ServiceRegistry>}
   * @private
   * @memberof Services
   */
  registries: new _weakMap.default(),
  /**
   * The {@link WeakMap} of {@link ServiceState} class instances that are
   * keyed with WebexCore instances.
   *
   * @instance
   * @type {WeakMap<WebexCore, ServiceState>}
   * @private
   * @memberof Services
   */
  states: new _weakMap.default(),
  props: {
    validateDomains: ['boolean', false, true],
    initFailed: ['boolean', false, false]
  },
  _catalogs: new _weakMap.default(),
  _serviceUrls: null,
  _hostCatalog: null,
  // Map of active cluster ids per service, e.g. { wdm: 'urn:TEAM:ap-southeast-2_m:wdm' }
  _activeServices: {},
  /**
   * Get the registry associated with this webex instance.
   *
   * @private
   * @memberof Services
   * @returns {ServiceRegistry} - The associated {@link ServiceRegistry}.
   */
  getRegistry: function getRegistry() {
    return this.registries.get(this.webex);
  },
  /**
   * Get the state associated with this webex instance.
   *
   * @private
   * @memberof Services
   * @returns {ServiceState} - The associated {@link ServiceState}.
   */
  getState: function getState() {
    return this.states.get(this.webex);
  },
  /**
   * @private
   * Get the current catalog based on the assocaited
   * webex instance.
   * @returns {ServiceCatalog}
   */
  _getCatalog: function _getCatalog() {
    return this._catalogs.get(this.webex);
  },
  /**
   * Safely access localStorage if available; returns the Storage or null.
   * @returns {Storage|null}
   */
  _getLocalStorageSafe: function _getLocalStorageSafe() {
    if (typeof window !== 'undefined' && window.localStorage) {
      return window.localStorage;
    }
    return null;
  },
  /**
   * Determine the intended preauth selection based on the current context.
   * @param {string|undefined} currentOrgId
   * @returns {{selectionType: string, selectionValue: string}}
   */
  getIntendedPreauthSelection: function getIntendedPreauthSelection(currentOrgId) {
    var _this$webex$credentia;
    if ((_this$webex$credentia = this.webex.credentials) !== null && _this$webex$credentia !== void 0 && _this$webex$credentia.canAuthorize) {
      if (currentOrgId) {
        return {
          selectionType: 'orgId',
          selectionValue: currentOrgId
        };
      }
    }
    var emailConfig = this.webex.config && this.webex.config.email;
    if (typeof emailConfig === 'string' && emailConfig.trim()) {
      return {
        selectionType: 'emailhash',
        selectionValue: (0, _sha.default)(emailConfig.toLowerCase()).toString()
      };
    }

    // fall back to proximity mode when no orgId or email available
    return {
      selectionType: 'mode',
      selectionValue: 'DEFAULT_BY_PROXIMITY'
    };
  },
  /**
   * Get a service url from the current services list by name
   * from the associated instance catalog.
   * @param {string} name
   * @param {boolean} [priorityHost]
   * @param {string} [serviceGroup]
   * @returns {string|undefined}
   */
  get: function get(name, priorityHost, serviceGroup) {
    var catalog = this._getCatalog();
    return catalog.get(name, priorityHost, serviceGroup);
  },
  /**
   * Determine if the catalog contains a specific service
   *
   * @param {string} serviceName - The service name to validate.
   * @returns {boolean} - True if the service exists.
   */
  hasService: function hasService(serviceName) {
    return !!this.get(serviceName);
  },
  /**
   * Determine if a whilelist exists in the service catalog.
   *
   * @returns {boolean} - True if a allowed domains list exists.
   */
  hasAllowedDomains: function hasAllowedDomains() {
    var catalog = this._getCatalog();
    return catalog.getAllowedDomains().length > 0;
  },
  /**
   * Generate a service catalog as an object from
   * the associated instance catalog.
   * @param {boolean} [priorityHost] - use highest priority host if set to `true`
   * @param {string} [serviceGroup]
   * @returns {Record<string, string>}
   */
  list: function list(priorityHost, serviceGroup) {
    var catalog = this._getCatalog();
    return catalog.list(priorityHost, serviceGroup);
  },
  /**
   * Mark a priority host service url as failed.
   * This will mark the host associated with the
   * `ServiceUrl` to be removed from the its
   * respective host array, and then return the next
   * viable host from the `ServiceUrls` host array,
   * or the `ServiceUrls` default url if no other priority
   * hosts are available, or if `noPriorityHosts` is set to
   * `true`.
   * @param {string} url
   * @param {boolean} noPriorityHosts
   * @returns {string}
   */
  markFailedUrl: function markFailedUrl(url, noPriorityHosts) {
    var catalog = this._getCatalog();
    return catalog.markFailedUrl(url, noPriorityHosts);
  },
  /**
   * Get all Mobius cluster host entries from the legacy host catalog.
   * @returns {Array<{host: string, id: string, ttl: number, priority: number}>}
   */
  getMobiusClusters: function getMobiusClusters() {
    this.logger.info('services: fetching mobius clusters');
    var clusters = [];
    var hostCatalog = this._hostCatalog || {};
    (0, _entries.default)(hostCatalog).forEach(function (_ref) {
      var _ref2 = (0, _slicedToArray2.default)(_ref, 2),
        host = _ref2[0],
        entries = _ref2[1];
      (entries || []).forEach(function (entry) {
        if (typeof (entry === null || entry === void 0 ? void 0 : entry.id) === 'string' && entry.id.endsWith(':mobius')) {
          // Ensure host is included; prefer entry.host if present, else use the map key
          var withHost = entry.host ? entry.host : host;
          // Skip duplicates for the same host
          if (!clusters.find(function (c) {
            return c && c.host === withHost;
          })) {
            clusters.push(_objectSpread(_objectSpread({}, entry), {}, {
              host: withHost
            }));
          }
        }
      });
    });
    return clusters;
  },
  /**
   * Check is valid host from the legacy host catalog.
   * @param {string} host
   * @returns {Boolean}
   */
  isValidHost: function isValidHost(host) {
    var _hostCatalog$host;
    var hostCatalog = this._hostCatalog || {};
    return !!((_hostCatalog$host = hostCatalog[host]) !== null && _hostCatalog$host !== void 0 && _hostCatalog$host.length);
  },
  /**
   * Checks if the current environment is an integration (INT) environment
   * by examining the u2c discovery URL from webex config.
   * INT environments use discovery URLs containing 'intb' (e.g., u2c-intb.ciscospark.com).
   * @returns {boolean} True if INT environment, false otherwise
   */
  isIntegrationEnvironment: function isIntegrationEnvironment() {
    try {
      var _this$webex, _this$webex$config, _this$webex$config$se, _this$webex$config$se2;
      var u2cUrl = ((_this$webex = this.webex) === null || _this$webex === void 0 ? void 0 : (_this$webex$config = _this$webex.config) === null || _this$webex$config === void 0 ? void 0 : (_this$webex$config$se = _this$webex$config.services) === null || _this$webex$config$se === void 0 ? void 0 : (_this$webex$config$se2 = _this$webex$config$se.discovery) === null || _this$webex$config$se2 === void 0 ? void 0 : _this$webex$config$se2.u2c) || '';
      var isInt = u2cUrl.includes('intb');
      this.logger.info("services: isIntegrationEnvironment: ".concat(isInt));
      return isInt;
    } catch (error) {
      this.logger.error('services: failed to determine integration environment', error);
      return false;
    }
  },
  /**
   * Merge provided active cluster mappings into current state.
   * @param {Record<string,string>} activeServices
   * @returns {void}
   */
  _updateActiveServices: function _updateActiveServices(activeServices) {
    this._activeServices = _objectSpread(_objectSpread({}, this._activeServices), activeServices);
  },
  /**
   * saves all the services from the pre and post catalog service
   * @param {Object} serviceUrls
   * @returns {void}
   */
  _updateServiceUrls: function _updateServiceUrls(serviceUrls) {
    this._serviceUrls = _objectSpread(_objectSpread({}, this._serviceUrls), serviceUrls);
  },
  /**
   * saves the hostCatalog object
   * @param {Object} hostCatalog
   * @returns {void}
   */
  _updateHostCatalog: function _updateHostCatalog(hostCatalog) {
    this._hostCatalog = _objectSpread(_objectSpread({}, this._hostCatalog), hostCatalog);
  },
  /**
   * Update a list of `serviceUrls` to the most current
   * catalog via the defined `discoveryUrl` then returns the current
   * list of services.
   * @param {object} [param]
   * @param {string} [param.from] - This accepts `limited` or `signin`
   * @param {object} [param.query] - This accepts `email`, `orgId` or `userId` key values
   * @param {string} [param.query.email] - must be a standard-format email
   * @param {string} [param.query.orgId] - must be an organization id
   * @param {string} [param.query.userId] - must be a user id
   * @param {string} [param.token] - used for signin catalog
   * @returns {Promise<object>}
   */
  updateServices: function updateServices() {
    var _arguments = arguments,
      _this = this;
    return (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee() {
      var _ref3, from, query, token, forceRefresh, catalog, formattedQuery, serviceGroup, queryKey, _queryKey, _t;
      return _regenerator.default.wrap(function (_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            _ref3 = _arguments.length > 0 && _arguments[0] !== undefined ? _arguments[0] : {}, from = _ref3.from, query = _ref3.query, token = _ref3.token, forceRefresh = _ref3.forceRefresh;
            catalog = _this._getCatalog();
            _t = from;
            _context.next = _t === 'limited' ? 1 : _t === 'signin' ? 2 : 3;
            break;
          case 1:
            serviceGroup = 'preauth';
            return _context.abrupt("continue", 4);
          case 2:
            serviceGroup = 'signin';
            return _context.abrupt("continue", 4);
          case 3:
            serviceGroup = 'postauth';
            return _context.abrupt("continue", 4);
          case 4:
            if (!catalog.status[serviceGroup].collecting) {
              _context.next = 5;
              break;
            }
            return _context.abrupt("return", _this.waitForCatalog(serviceGroup));
          case 5:
            catalog.status[serviceGroup].collecting = true;
            if (!(serviceGroup === 'preauth')) {
              _context.next = 6;
              break;
            }
            queryKey = query && (0, _keys.default)(query)[0];
            if (['email', 'emailhash', 'userId', 'orgId', 'mode'].includes(queryKey)) {
              _context.next = 6;
              break;
            }
            return _context.abrupt("return", _promise.default.reject(new Error('a query param of email, emailhash, userId, orgId, or mode is required')));
          case 6:
            // encode email when query key is email
            if (serviceGroup === 'preauth' || serviceGroup === 'signin') {
              _queryKey = (0, _keys.default)(query)[0];
              formattedQuery = {};
              if (_queryKey === 'email' && query.email) {
                formattedQuery.emailhash = (0, _sha.default)(query.email.toLowerCase()).toString();
              } else {
                formattedQuery[_queryKey] = query[_queryKey];
              }
            }
            return _context.abrupt("return", _this._fetchNewServiceHostmap({
              from: from,
              token: token,
              query: formattedQuery,
              forceRefresh: forceRefresh
            }).then(function (serviceHostMap) {
              var formattedServiceHostMap = _this._formatReceivedHostmap(serviceHostMap);
              // Build selection metadata for caching discrimination
              var selectionMeta;
              if (serviceGroup === 'preauth' || serviceGroup === 'signin') {
                var key = formattedQuery && (0, _keys.default)(formattedQuery || {})[0];
                if (key) {
                  selectionMeta = {
                    selectionType: key,
                    selectionValue: formattedQuery[key]
                  };
                }
              }
              _this._cacheCatalog(serviceGroup, serviceHostMap, selectionMeta);
              catalog.updateServiceUrls(serviceGroup, formattedServiceHostMap);
              _this.updateCredentialsConfig();
              catalog.status[serviceGroup].collecting = false;
            }).catch(function (error) {
              catalog.status[serviceGroup].collecting = false;
              return _promise.default.reject(error);
            }));
          case 7:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }))();
  },
  /**
   * User validation parameter transfer object for {@link validateUser}.
   * @param {object} ValidateUserPTO
   * @property {string} ValidateUserPTO.email - The email of the user.
   * @property {string} [ValidateUserPTO.reqId] - The activation requester.
   * @property {object} [ValidateUserPTO.activationOptions] - Extra options to pass when sending the activation
   * @property {object} [ValidateUserPTO.preloginUserId] - The prelogin user id to set when sending the activation.
   */
  /**
   * User validation return transfer object for {@link validateUser}.
   * @param {object} ValidateUserRTO
   * @property {boolean} ValidateUserRTO.activated - If the user is activated.
   * @property {boolean} ValidateUserRTO.exists - If the user exists.
   * @property {string} ValidateUserRTO.details - A descriptive status message.
   * @property {object} ValidateUserRTO.user - **License** service user object.
   */
  /**
   * Validate if a user is activated and update the service catalogs as needed
   * based on the user's activation status.
   *
   * @param {ValidateUserPTO} - The parameter transfer object.
   * @returns {ValidateUserRTO} - The return transfer object.
   */
  validateUser: function validateUser(_ref4) {
    var _this2 = this;
    var email = _ref4.email,
      _ref4$reqId = _ref4.reqId,
      reqId = _ref4$reqId === void 0 ? 'WEBCLIENT' : _ref4$reqId,
      _ref4$forceRefresh = _ref4.forceRefresh,
      forceRefresh = _ref4$forceRefresh === void 0 ? false : _ref4$forceRefresh,
      _ref4$activationOptio = _ref4.activationOptions,
      activationOptions = _ref4$activationOptio === void 0 ? {} : _ref4$activationOptio,
      preloginUserId = _ref4.preloginUserId;
    this.logger.info('services: validating a user');

    // Validate that an email parameter key was provided.
    if (!email) {
      return _promise.default.reject(new Error('`email` is required'));
    }

    // Destructure the credentials object.
    var canAuthorize = this.webex.credentials.canAuthorize;

    // Validate that the user is already authorized.
    if (canAuthorize) {
      return this.updateServices({
        forceRefresh: forceRefresh
      }).then(function () {
        return _this2.webex.credentials.getUserToken();
      }).then(function (token) {
        return _this2.sendUserActivation({
          email: email,
          reqId: reqId,
          token: token.toString(),
          activationOptions: activationOptions,
          preloginUserId: preloginUserId
        });
      }).then(function (userObj) {
        return {
          activated: true,
          exists: true,
          details: 'user is authorized via a user token',
          user: userObj
        };
      });
    }

    // Destructure the client authorization details.
    /* eslint-disable camelcase */
    var _this$webex$credentia2 = this.webex.credentials.config,
      client_id = _this$webex$credentia2.client_id,
      client_secret = _this$webex$credentia2.client_secret;

    // Validate that client authentication details exist.
    if (!client_id || !client_secret) {
      return _promise.default.reject(new Error('client authentication details are not available'));
    }
    /* eslint-enable camelcase */

    // Declare a class-memeber-scoped token for usage within the promise chain.
    var token;

    // Begin client authentication user validation.
    return this.collectPreauthCatalog({
      email: email
    }).then(function () {
      // Retrieve the service url from the updated catalog. This is required
      // since `WebexCore` is usually not fully initialized at the time this
      // request completes.
      var idbrokerService = _this2.get('idbroker', true);

      // Collect the client auth token.
      return _this2.webex.credentials.getClientToken({
        uri: "".concat(idbrokerService, "idb/oauth2/v1/access_token"),
        scope: 'webexsquare:admin webexsquare:get_conversation Identity:SCIM'
      });
    }).then(function (tokenObj) {
      // Generate the token string.
      token = tokenObj.toString();

      // Collect the signin catalog using the client auth information.
      return _this2.collectSigninCatalog({
        email: email,
        token: token,
        forceRefresh: forceRefresh
      });
    })
    // Validate if collecting the signin catalog failed and populate the RTO
    // with the appropriate content.
    .catch(function (error) {
      return {
        exists: error.name !== 'NotFound',
        activated: false,
        details: error.name !== 'NotFound' ? 'user exists but is not activated' : 'user does not exist and is not activated'
      };
    })
    // Validate if the previous promise resolved with an RTO and populate the
    // new RTO accordingly.
    .then(function (rto) {
      return _promise.default.all([rto || {
        activated: true,
        exists: true,
        details: 'user exists and is activated'
      }, _this2.sendUserActivation({
        email: email,
        reqId: reqId,
        token: token,
        activationOptions: activationOptions,
        preloginUserId: preloginUserId
      })]);
    }).then(function (_ref5) {
      var _ref6 = (0, _slicedToArray2.default)(_ref5, 2),
        rto = _ref6[0],
        user = _ref6[1];
      return _objectSpread(_objectSpread({}, rto), {}, {
        user: user
      });
    }).catch(function (error) {
      var response = {
        statusCode: error.statusCode,
        responseText: error.body && error.body.message,
        body: error.body
      };
      return _promise.default.reject(response);
    });
  },
  /**
   * Get user meeting preferences (preferred webex site).
   *
   * @returns {object} - User Information including user preferrences .
   */
  getMeetingPreferences: function getMeetingPreferences() {
    var _this3 = this;
    return this.request({
      method: 'GET',
      service: 'hydra',
      resource: 'meetingPreferences'
    }).then(function (res) {
      _this3.logger.info('services: received user region info');
      return res.body;
    }).catch(function (err) {
      _this3.logger.info('services: was not able to fetch user login information', err);
      // resolve successfully even if request failed
    });
  },
  /**
   * Fetches client region info such as countryCode and timezone.
   *
   * @returns {object} - The region info object.
   */
  fetchClientRegionInfo: function fetchClientRegionInfo() {
    var _this4 = this;
    var services = this.webex.config.services;
    return this.request({
      uri: services.discovery.sqdiscovery,
      addAuthHeader: false,
      headers: {
        'spark-user-agent': null
      },
      timeout: 5000
    }).then(function (res) {
      _this4.logger.info('services: received user region info');
      return res.body;
    }).catch(function (err) {
      _this4.logger.info('services: was not able to get user region info', err);
      // resolve successfully even if request failed
    });
  },
  /**
   * User activation parameter transfer object for {@link sendUserActivation}.
   * @typedef {object} SendUserActivationPTO
   * @property {string} SendUserActivationPTO.email - The email of the user.
   * @property {string} SendUserActivationPTO.reqId - The activation requester.
   * @property {string} SendUserActivationPTO.token - The client auth token.
   * @property {object} SendUserActivationPTO.activationOptions - Extra options to pass when sending the activation.
   * @property {object} SendUserActivationPTO.preloginUserId - The prelogin user id to set when sending the activation.
   */
  /**
   * Send a request to activate a user using a client token.
   *
   * @param {SendUserActivationPTO} - The Parameter transfer object.
   * @returns {LicenseDTO} - The DTO returned from the **License** service.
   */
  sendUserActivation: function sendUserActivation(_ref7) {
    var _this5 = this;
    var email = _ref7.email,
      reqId = _ref7.reqId,
      token = _ref7.token,
      activationOptions = _ref7.activationOptions,
      preloginUserId = _ref7.preloginUserId;
    this.logger.info('services: sending user activation request');
    var countryCode;
    var timezone;

    // try to fetch client region info first
    return this.fetchClientRegionInfo().then(function (clientRegionInfo) {
      var _this5$webex$config$s;
      if (clientRegionInfo) {
        countryCode = clientRegionInfo.countryCode;
        timezone = clientRegionInfo.timezone;
      }

      // Send the user activation request.
      // Use user-onboarding service if configured, otherwise use license service.
      var useUserOnboarding = (_this5$webex$config$s = _this5.webex.config.services) === null || _this5$webex$config$s === void 0 ? void 0 : _this5$webex$config$s.useUserOnboardingServiceForActivations;
      return _this5.request({
        service: useUserOnboarding ? 'user-onboarding' : 'license',
        resource: useUserOnboarding ? 'api/v1/users/activations' : 'users/activations',
        method: 'POST',
        headers: {
          accept: 'application/json',
          authorization: token,
          'x-prelogin-userid': preloginUserId
        },
        body: _objectSpread({
          email: email,
          reqId: reqId,
          countryCode: countryCode,
          timeZone: timezone
        }, activationOptions),
        shouldRefreshAccessToken: false
      });
    })
    // On success, return the **License** user object.
    .then(function (_ref8) {
      var body = _ref8.body;
      return body;
    })
    // On failure, reject with error from **License**.
    .catch(function (error) {
      return _promise.default.reject(error);
    });
  },
  /**
   * Updates a given service group i.e. preauth, signin, postauth with a new hostmap.
   * @param {string} serviceGroup - preauth, signin, postauth
   * @param {object} hostMap - The new hostmap to update the service group with.
   * @returns {Promise<void>}
   */
  updateCatalog: function updateCatalog(serviceGroup, hostMap) {
    var catalog = this._getCatalog();
    var serviceHostMap = this._formatReceivedHostmap(hostMap);
    return catalog.updateServiceUrls(serviceGroup, serviceHostMap);
  },
  /**
   * simplified method to update the preauth catalog via email
   *
   * @param {object} query
   * @param {string} query.email - A standard format email.
   * @param {string} query.orgId - The user's OrgId.
   * @param {boolean} forceRefresh - Boolean to bypass u2c cache control header
   * @returns {Promise<void>}
   */
  collectPreauthCatalog: function collectPreauthCatalog(query) {
    var forceRefresh = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
    if (!query) {
      return this.updateServices({
        from: 'limited',
        query: {
          mode: 'DEFAULT_BY_PROXIMITY'
        },
        forceRefresh: forceRefresh
      });
    }
    return this.updateServices({
      from: 'limited',
      query: query,
      forceRefresh: forceRefresh
    });
  },
  /**
   * simplified method to update the signin catalog via email and token
   * @param {object} param
   * @param {string} param.email - must be a standard-format email
   * @param {string} param.token - must be a client token
   * @returns {Promise<void>}
   */
  collectSigninCatalog: function collectSigninCatalog() {
    var _ref9 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      email = _ref9.email,
      token = _ref9.token,
      forceRefresh = _ref9.forceRefresh;
    if (!email) {
      return _promise.default.reject(new Error('`email` is required'));
    }
    if (!token) {
      return _promise.default.reject(new Error('`token` is required'));
    }
    return this.updateServices({
      from: 'signin',
      query: {
        email: email
      },
      token: token,
      forceRefresh: forceRefresh
    });
  },
  /**
   * Updates credentials config to utilize u2c catalog
   * urls.
   * @returns {void}
   */
  updateCredentialsConfig: function updateCredentialsConfig() {
    var _this$list = this.list(true),
      idbroker = _this$list.idbroker,
      identity = _this$list.identity;
    if (idbroker && identity) {
      var _this$webex$config$cr = this.webex.config.credentials,
        authorizationString = _this$webex$config$cr.authorizationString,
        authorizeUrl = _this$webex$config$cr.authorizeUrl;

      // This must be set outside of the setConfig method used to assign the
      // idbroker and identity url values.
      this.webex.config.credentials.authorizeUrl = authorizationString ? authorizeUrl : "".concat(idbroker.replace(trailingSlashes, ''), "/idb/oauth2/v1/authorize");
      this.webex.setConfig({
        credentials: {
          idbroker: {
            url: idbroker.replace(trailingSlashes, '') // remove trailing slash
          },
          identity: {
            url: identity.replace(trailingSlashes, '') // remove trailing slash
          }
        }
      });
    }
  },
  /**
   * Wait until the service catalog is available,
   * or reject afte ra timeout of 60 seconds.
   * @param {string} serviceGroup
   * @param {number} [timeout] - in seconds
   * @returns {Promise<void>}
   */
  waitForCatalog: function waitForCatalog(serviceGroup, timeout) {
    var catalog = this._getCatalog();
    var supertoken = this.webex.credentials.supertoken;
    if (serviceGroup === 'postauth' && supertoken && supertoken.access_token && !catalog.status.postauth.collecting && !catalog.status.postauth.ready) {
      if (!catalog.status.preauth.ready) {
        return this.initServiceCatalogs();
      }
      return this.updateServices();
    }
    return catalog.waitForCatalog(serviceGroup, timeout);
  },
  /**
   * Service waiting parameter transfer object for {@link waitForService}.
   *
   * @typedef {object} WaitForServicePTO
   * @property {string} [WaitForServicePTO.name] - The service name.
   * @property {string} [WaitForServicePTO.url] - The service url.
   * @property {string} [WaitForServicePTO.timeout] - wait duration in seconds.
   */
  /**
   * Wait until the service has been ammended to any service catalog. This
   * method prioritizes the service name over the service url when searching.
   *
   * @param {WaitForServicePTO} - The parameter transfer object.
   * @returns {Promise<string>} - Resolves to the priority host of a service.
   */
  waitForService: function waitForService(_ref0) {
    var _this6 = this;
    var name = _ref0.name,
      _ref0$timeout = _ref0.timeout,
      timeout = _ref0$timeout === void 0 ? 5 : _ref0$timeout,
      url = _ref0.url;
    var services = this.webex.config.services;

    // Save memory by grabbing the catalog after there isn't a priortyURL
    var catalog = this._getCatalog();
    var fetchFromServiceUrl = services.servicesNotNeedValidation.find(function (service) {
      return service === name;
    });
    if (fetchFromServiceUrl) {
      return _promise.default.resolve(this._serviceUrls[name]);
    }
    var priorityUrl = this.get(name, true);
    var priorityUrlObj = this.getServiceFromUrl(url);
    if (priorityUrl || priorityUrlObj) {
      return _promise.default.resolve(priorityUrl || priorityUrlObj.priorityUrl);
    }
    if (catalog.isReady) {
      if (url) {
        return _promise.default.resolve(url);
      }
      this.webex.internal.metrics.submitClientMetrics(_metrics.default.JS_SDK_SERVICE_NOT_FOUND, {
        fields: {
          service_name: name
        }
      });
      return _promise.default.reject(new Error("services: service '".concat(name, "' was not found in any of the catalogs")));
    }
    return new _promise.default(function (resolve, reject) {
      var groupsToCheck = ['preauth', 'signin', 'postauth'];
      var checkCatalog = function checkCatalog(catalogGroup) {
        return catalog.waitForCatalog(catalogGroup, timeout).then(function () {
          var scopedPriorityUrl = _this6.get(name, true);
          var scopedPrioriryUrlObj = _this6.getServiceFromUrl(url);
          if (scopedPriorityUrl || scopedPrioriryUrlObj) {
            resolve(scopedPriorityUrl || scopedPrioriryUrlObj.priorityUrl);
          }
        }).catch(function () {
          return undefined;
        });
      };
      _promise.default.all(groupsToCheck.map(function (group) {
        return checkCatalog(group);
      })).then(function () {
        _this6.webex.internal.metrics.submitClientMetrics(_metrics.default.JS_SDK_SERVICE_NOT_FOUND, {
          fields: {
            service_name: name
          }
        });
        reject(new Error("services: service '".concat(name, "' was not found after waiting")));
      });
    });
  },
  /**
   * Looks up the hostname in the host catalog
   * and replaces it with the first host if it finds it
   * @param {string} uri
   * @returns {string} uri with the host replaced
   */
  replaceHostFromHostmap: function replaceHostFromHostmap(uri) {
    var url = new URL(uri);
    var hostCatalog = this._hostCatalog;
    if (!hostCatalog) {
      return uri;
    }
    var host = hostCatalog[url.host];
    if (host && host[0]) {
      var newHost = host[0].host;
      url.host = newHost;
      return url.toString();
    }
    return uri;
  },
  /**
   * @private
   * Organize a received hostmap from a service
   * catalog endpoint.
   * @param {object} serviceHostmap
   * @returns {object}
   */
  _formatReceivedHostmap: function _formatReceivedHostmap(serviceHostmap) {
    this._updateHostCatalog(serviceHostmap.hostCatalog);
    var extractId = function extractId(entry) {
      return entry.id.split(':')[3];
    };
    var formattedHostmap = [];

    // for each of the services in the serviceLinks, find the matching host in the catalog
    (0, _keys.default)(serviceHostmap.serviceLinks).forEach(function (serviceName) {
      var _formattedHost$hosts;
      var serviceUrl = serviceHostmap.serviceLinks[serviceName];
      var host;
      try {
        host = new URL(serviceUrl).host;
      } catch (e) {
        return;
      }
      var matchingCatalogEntry = serviceHostmap.hostCatalog[host];
      var formattedHost = {
        name: serviceName,
        defaultUrl: serviceUrl,
        defaultHost: host,
        hosts: []
      };
      formattedHostmap.push(formattedHost);

      // If the catalog does not have any hosts we will be unable to find the service ID
      // so can't search for other hosts
      if (!matchingCatalogEntry || !matchingCatalogEntry[0]) {
        return;
      }
      var serviceId = extractId(matchingCatalogEntry[0]);
      (0, _lodash.forEach)(matchingCatalogEntry, function (entry) {
        // The ids for all hosts within a hostCatalog entry should be the same
        // but for safety, only add host entries that have the same id as the first one
        if (extractId(entry) === serviceId) {
          formattedHost.hosts.push(_objectSpread(_objectSpread({}, entry), {}, {
            homeCluster: true
          }));
        }
      });
      var otherHosts = [];

      // find the services in the host catalog that have the same id
      // and add them to the otherHosts
      (0, _lodash.forEach)(serviceHostmap.hostCatalog, function (entry) {
        // exclude the matching catalog entry as we have already added that
        if (entry === matchingCatalogEntry) {
          return;
        }
        (0, _lodash.forEach)(entry, function (entryHost) {
          // only add hosts that have the correct id
          if (extractId(entryHost) === serviceId) {
            otherHosts.push(_objectSpread(_objectSpread({}, entryHost), {}, {
              homeCluster: false
            }));
          }
        });
      });
      (_formattedHost$hosts = formattedHost.hosts).push.apply(_formattedHost$hosts, otherHosts);
    });

    // update all the service urls in the host catalog

    this._updateServiceUrls(serviceHostmap.serviceLinks);
    this._updateHostCatalog(serviceHostmap.hostCatalog);
    return formattedHostmap;
  },
  /**
   * Get the clusterId associated with a URL string.
   * @param {string} url
   * @returns {string} - Cluster ID of url provided
   */
  getClusterId: function getClusterId(url) {
    var catalog = this._getCatalog();
    return catalog.findClusterId(url);
  },
  /**
   * Get a service value from a provided clusterId. This method will
   * return an object containing both the name and url of a found service.
   * @param {object} params
   * @param {string} params.clusterId - clusterId of found service
   * @param {boolean} [params.priorityHost] - returns priority host url if true
   * @param {string} [params.serviceGroup] - specify service group
   * @returns {object} service
   * @returns {string} service.name
   * @returns {string} service.url
   */
  getServiceFromClusterId: function getServiceFromClusterId(params) {
    var catalog = this._getCatalog();
    return catalog.findServiceFromClusterId(params);
  },
  /**
   * @param {String} cluster the cluster containing the id
   * @param {UUID} [id] the id of the conversation.
   *  If empty, just return the base URL.
   * @returns {String} url of the service
   */
  getServiceUrlFromClusterId: function getServiceUrlFromClusterId() {
    var _ref1 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      _ref1$cluster = _ref1.cluster,
      cluster = _ref1$cluster === void 0 ? 'us' : _ref1$cluster;
    var clusterId = cluster === 'us' ? DEFAULT_CLUSTER_IDENTIFIER : cluster;

    // Determine if cluster has service name (non-US clusters from hydra do not)
    if (clusterId.split(':').length < 4) {
      // Add Service to cluster identifier
      clusterId = "".concat(cluster, ":").concat(CLUSTER_SERVICE);
    }
    var _ref10 = this.getServiceFromClusterId({
        clusterId: clusterId
      }) || {},
      url = _ref10.url;
    if (!url) {
      throw Error("Could not find service for cluster [".concat(cluster, "]"));
    }
    return url;
  },
  /**
   * Get a service object from a service url if the service url exists in the
   * catalog.
   *
   * @param {string} url - The url to be validated.
   * @returns {object} - Service object.
   * @returns {object.name} - The name of the service found.
   * @returns {object.priorityUrl} - The priority url of the found service.
   * @returns {object.defaultUrl} - The default url of the found service.
   */
  getServiceFromUrl: function getServiceFromUrl() {
    var url = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    var service = this._getCatalog().findServiceUrlFromUrl(url);
    if (!service) {
      return undefined;
    }
    return {
      name: service.name,
      priorityUrl: service.get(true),
      defaultUrl: service.get()
    };
  },
  /**
   * Verify that a provided url exists in the service
   * catalog.
   * @param {string} url
   * @returns {boolean} - true if exists, false otherwise
   */
  isServiceUrl: function isServiceUrl(url) {
    var catalog = this._getCatalog();
    return !!catalog.findServiceUrlFromUrl(url);
  },
  /**
   * Determine if a provided url is in the catalog's allowed domains.
   *
   * @param {string} url - The url to match allowed domains against.
   * @returns {boolean} - True if the url provided is allowed.
   */
  isAllowedDomainUrl: function isAllowedDomainUrl(url) {
    var catalog = this._getCatalog();
    return !!catalog.findAllowedDomain(url);
  },
  /**
   * Converts the host portion of the url from default host
   * to a priority host
   *
   * @param {string} url a service url that contains a default host
   * @returns {string} a service url that contains the top priority host.
   * @throws if url isn't a service url
   */
  convertUrlToPriorityHostUrl: function convertUrlToPriorityHostUrl() {
    var url = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    var data = this.getServiceFromUrl(url);
    if (!data) {
      throw Error("No service associated with url: [".concat(url, "]"));
    }
    return url.replace(data.defaultUrl, data.priorityUrl);
  },
  /**
   * @private
   * Simplified method wrapper for sending a request to get
   * an updated service hostmap.
   * @param {object} [param]
   * @param {string} [param.from] - This accepts `limited` or `signin`
   * @param {object} [param.query] - This accepts `email`, `orgId` or `userId` key values
   * @param {string} [param.query.email] - must be a standard-format email
   * @param {string} [param.query.orgId] - must be an organization id
   * @param {string} [param.query.userId] - must be a user id
   * @param {string} [param.token] - used for signin catalog
   * @returns {Promise<object>}
   */
  _fetchNewServiceHostmap: function _fetchNewServiceHostmap() {
    var _this7 = this;
    var _ref11 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      from = _ref11.from,
      query = _ref11.query,
      token = _ref11.token,
      forceRefresh = _ref11.forceRefresh;
    var service = 'u2c';
    var resource = from ? "/".concat(from, "/catalog") : '/catalog';
    var qs = _objectSpread(_objectSpread({}, query), {}, {
      format: 'hostmap'
    });
    if (forceRefresh) {
      qs.timestamp = new Date().getTime();
    }
    var requestObject = {
      method: 'GET',
      service: service,
      resource: resource,
      qs: qs
    };
    if (token) {
      requestObject.headers = {
        authorization: token
      };
    }
    return this.webex.internal.newMetrics.callDiagnosticLatencies.measureLatency(function () {
      return _this7.request(requestObject);
    }, 'internal.get.u2c.time').then(function (_ref12) {
      var body = _ref12.body;
      return body;
    });
  },
  /**
   * Cache the catalog in the bounded storage.
   * @param {string} serviceGroup - preauth, signin, postauth
   * @param {object} hostMap - The hostmap to cache
   * @param {object} [meta] - Optional selection metadata used to validate cache reuse
   * @returns {Promise<void>}
   *
   */
  _cacheCatalog: function _cacheCatalog(serviceGroup, hostMap, meta) {
    var _this8 = this;
    return (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee2() {
      var current, orgId, _this8$webex$config, _this8$webex$config$c, _this8$webex, _this8$webex$config2, _this8$webex2, _this8$webex2$config, _this8$webex2$config$, _this8$webex2$config$2, _ls, cachedJson, credentials, _current, env, fedramp, u2cDiscoveryUrl, updated, ls, _t2;
      return _regenerator.default.wrap(function (_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            current = {};
            _context2.prev = 1;
            if ((_this8$webex$config = _this8.webex.config) !== null && _this8$webex$config !== void 0 && (_this8$webex$config$c = _this8$webex$config.calling) !== null && _this8$webex$config$c !== void 0 && _this8$webex$config$c.cacheU2C) {
              _context2.next = 2;
              break;
            }
            _this8.logger.info("services: skipping cache write for ".concat(serviceGroup, " as per the config"));
            return _context2.abrupt("return");
          case 2:
            // Persist to localStorage to survive browser refresh
            try {
              _ls = _this8._getLocalStorageSafe();
              cachedJson = _ls ? _ls.getItem(CATALOG_CACHE_KEY_V1) : null;
              current = cachedJson ? JSON.parse(cachedJson) : {};
            } catch (e) {
              current = {};
            }
            try {
              credentials = _this8.webex.credentials;
              orgId = credentials.getOrgId();
            } catch (e) {
              orgId = current.orgId;
            }

            // Capture environment fingerprint to invalidate cache across env changes
            _current = current, env = _current.env;
            fedramp = !!((_this8$webex = _this8.webex) !== null && _this8$webex !== void 0 && (_this8$webex$config2 = _this8$webex.config) !== null && _this8$webex$config2 !== void 0 && _this8$webex$config2.fedramp);
            u2cDiscoveryUrl = (_this8$webex2 = _this8.webex) === null || _this8$webex2 === void 0 ? void 0 : (_this8$webex2$config = _this8$webex2.config) === null || _this8$webex2$config === void 0 ? void 0 : (_this8$webex2$config$ = _this8$webex2$config.services) === null || _this8$webex2$config$ === void 0 ? void 0 : (_this8$webex2$config$2 = _this8$webex2$config$.discovery) === null || _this8$webex2$config$2 === void 0 ? void 0 : _this8$webex2$config$2.u2c;
            env = {
              fedramp: fedramp,
              u2cDiscoveryUrl: u2cDiscoveryUrl
            };
            updated = _objectSpread(_objectSpread({}, current), {}, (0, _defineProperty2.default)((0, _defineProperty2.default)({
              orgId: orgId || current.orgId,
              env: env || current.env
            }, serviceGroup, meta ? {
              hostMap: hostMap,
              meta: meta
            } : hostMap), "cachedAt", (0, _now.default)()));
            ls = _this8._getLocalStorageSafe();
            if (ls) {
              ls.setItem(CATALOG_CACHE_KEY_V1, (0, _stringify.default)(updated));
            }
            _context2.next = 4;
            break;
          case 3:
            _context2.prev = 3;
            _t2 = _context2["catch"](1);
            _this8.logger.warn('services: error caching catalog', _t2);
          case 4:
          case "end":
            return _context2.stop();
        }
      }, _callee2, null, [[1, 3]]);
    }))();
  },
  /**
   * Load the catalog from cache and hydrate the in-memory ServiceCatalog.
   * @returns {Promise<boolean>} true if cache was loaded, false otherwise
   */
  _loadCatalogFromCache: function _loadCatalogFromCache() {
    var _this9 = this;
    return (0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee3() {
      var currentOrgId, _this9$webex$config, _this9$webex$config$c, _this9$webex$config2, _this9$webex$config3, _this9$webex$config3$, _this9$webex$config3$2, ls, cachedJson, cached, cachedAt, _this9$webex$credenti, credentials, fedramp, u2cDiscoveryUrl, currentEnv, sameEnv, catalog, groups, _t3, _t4;
      return _regenerator.default.wrap(function (_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            _context3.prev = 0;
            if ((_this9$webex$config = _this9.webex.config) !== null && _this9$webex$config !== void 0 && (_this9$webex$config$c = _this9$webex$config.calling) !== null && _this9$webex$config$c !== void 0 && _this9$webex$config$c.cacheU2C) {
              _context3.next = 1;
              break;
            }
            _this9.logger.info('services: skipping cache warm-up as per the cache config');
            return _context3.abrupt("return", false);
          case 1:
            ls = _this9._getLocalStorageSafe();
            if (ls) {
              _context3.next = 2;
              break;
            }
            _this9.logger.info('services: skipping cache warm-up as no localStorage is available');
            return _context3.abrupt("return", false);
          case 2:
            cachedJson = ls.getItem(CATALOG_CACHE_KEY_V1);
            cached = cachedJson ? JSON.parse(cachedJson) : undefined;
            if (cached) {
              _context3.next = 3;
              break;
            }
            return _context3.abrupt("return", false);
          case 3:
            // TTL enforcement: clear if older than 24 hours
            cachedAt = Number(cached.cachedAt) || 0;
            if (!(!cachedAt || (0, _now.default)() - cachedAt > CATALOG_TTL_MS)) {
              _context3.next = 4;
              break;
            }
            _this9.clearCatalogCache();
            return _context3.abrupt("return", false);
          case 4:
            _context3.prev = 4;
            if (!((_this9$webex$credenti = _this9.webex.credentials) !== null && _this9$webex$credenti !== void 0 && _this9$webex$credenti.canAuthorize)) {
              _context3.next = 5;
              break;
            }
            credentials = _this9.webex.credentials;
            currentOrgId = credentials.getOrgId();
            if (!(cached.orgId && cached.orgId !== currentOrgId)) {
              _context3.next = 5;
              break;
            }
            return _context3.abrupt("return", false);
          case 5:
            _context3.next = 7;
            break;
          case 6:
            _context3.prev = 6;
            _t3 = _context3["catch"](4);
            _this9.logger.warn('services: error checking orgId', _t3);
          case 7:
            // Ensure cached environment matches current environment
            fedramp = !!((_this9$webex$config2 = _this9.webex.config) !== null && _this9$webex$config2 !== void 0 && _this9$webex$config2.fedramp);
            u2cDiscoveryUrl = (_this9$webex$config3 = _this9.webex.config) === null || _this9$webex$config3 === void 0 ? void 0 : (_this9$webex$config3$ = _this9$webex$config3.services) === null || _this9$webex$config3$ === void 0 ? void 0 : (_this9$webex$config3$2 = _this9$webex$config3$.discovery) === null || _this9$webex$config3$2 === void 0 ? void 0 : _this9$webex$config3$2.u2c;
            currentEnv = {
              fedramp: fedramp,
              u2cDiscoveryUrl: u2cDiscoveryUrl
            };
            if (!cached.env) {
              _context3.next = 8;
              break;
            }
            sameEnv = cached.env.fedramp === currentEnv.fedramp && cached.env.u2cDiscoveryUrl === currentEnv.u2cDiscoveryUrl;
            if (sameEnv) {
              _context3.next = 8;
              break;
            }
            _this9.logger.info('services: skipping cache warm due to environment mismatch');
            return _context3.abrupt("return", false);
          case 8:
            catalog = _this9._getCatalog(); // Apply any cached groups (with preauth selection validation if available)
            groups = ['preauth', 'signin', 'postauth'];
            groups.forEach(function (serviceGroup) {
              var cachedGroup = cached[serviceGroup];
              if (!cachedGroup) {
                return;
              }

              // Support legacy (hostMap) and new ({hostMap, meta}) shapes
              var hostMap = cachedGroup && cachedGroup.hostMap ? cachedGroup.hostMap : cachedGroup;
              var meta = cachedGroup === null || cachedGroup === void 0 ? void 0 : cachedGroup.meta;
              if (serviceGroup === 'preauth' && meta) {
                // For proximity-based selection, always fetch fresh to respect IP/region changes
                if (meta.selectionType === 'mode') {
                  _this9.logger.info('services: skipping preauth cache warm for proximity mode');
                  return;
                }
                var intended = _this9.getIntendedPreauthSelection(currentOrgId);
                var matches = intended && intended.selectionType === meta.selectionType && intended.selectionValue === meta.selectionValue;
                if (!matches) {
                  _this9.logger.info('services: skipping preauth cache warm due to selection mismatch');
                  return;
                }
              }
              if (hostMap) {
                var formatted = _this9._formatReceivedHostmap(hostMap);
                catalog.updateServiceUrls(serviceGroup, formatted);
              }
            });

            // Align credentials against warmed catalog
            _this9.updateCredentialsConfig();
            return _context3.abrupt("return", true);
          case 9:
            _context3.prev = 9;
            _t4 = _context3["catch"](0);
            _this9.logger.warn('services: error loading catalog from cache', _t4);
            return _context3.abrupt("return", false);
          case 10:
          case "end":
            return _context3.stop();
        }
      }, _callee3, null, [[0, 9], [4, 6]]);
    }))();
  },
  /**
   * Clear the catalog cache from the bounded storage.
   * @returns {Promise<void>}
   */
  clearCatalogCache: function clearCatalogCache() {
    try {
      var ls = this._getLocalStorageSafe();
      if (ls) {
        ls.removeItem(CATALOG_CACHE_KEY_V1);
      }
    } catch (e) {
      this.logger.warn('services: error clearing catalog cache', e);
    }
    return _promise.default.resolve();
  },
  /**
   * Initialize the discovery services and the whitelisted services.
   *
   * @returns {void}
   */
  initConfig: function initConfig() {
    // Get the catalog and destructure the services config.
    var catalog = this._getCatalog();
    var _this$webex$config2 = this.webex.config,
      services = _this$webex$config2.services,
      fedramp = _this$webex$config2.fedramp;

    // Validate that the services configuration exists.
    if (services) {
      if (fedramp) {
        services.discovery = _serviceFedRamp.default;
      }
      // Check for discovery services.
      if (services.discovery) {
        // Format the discovery configuration into an injectable array.
        var formattedDiscoveryServices = (0, _keys.default)(services.discovery).map(function (key) {
          return {
            name: key,
            defaultUrl: services.discovery[key]
          };
        });

        // Inject formatted discovery services into services catalog.
        catalog.updateServiceUrls('discovery', formattedDiscoveryServices);
      }
      if (services.override) {
        // Format the override configuration into an injectable array.
        var formattedOverrideServices = (0, _keys.default)(services.override).map(function (key) {
          return {
            name: key,
            defaultUrl: services.override[key]
          };
        });

        // Inject formatted override services into services catalog.
        catalog.updateServiceUrls('override', formattedOverrideServices);
      }

      // if not fedramp, append on the commercialAllowedDomains
      if (!fedramp) {
        services.allowedDomains = (0, _lodash.union)(services.allowedDomains, _constants.COMMERCIAL_ALLOWED_DOMAINS);
      }

      // Check for allowed host domains.
      if (services.allowedDomains) {
        // Store the allowed domains as a property of the catalog.
        catalog.setAllowedDomains(services.allowedDomains);
      }

      // Set `validateDomains` property to match configuration
      this.validateDomains = services.validateDomains;
    }
  },
  /**
   * Make the initial requests to collect the root catalogs.
   *
   * @returns {Promise<void, Error>} - Errors if the token is unavailable.
   */
  initServiceCatalogs: function initServiceCatalogs() {
    var _this0 = this;
    this.logger.info('services: initializing initial service catalogs');

    // Destructure the credentials plugin.
    var credentials = this.webex.credentials;

    // Init a promise chain. Must be done as a Promise.resolve() to allow
    // credentials#getOrgId() to properly throw.
    return _promise.default.resolve()
    // Get the user's OrgId.
    .then(function () {
      return credentials.getOrgId();
    })
    // Begin collecting the preauth/limited catalog.
    .then(function (orgId) {
      return _this0.collectPreauthCatalog({
        orgId: orgId
      });
    }).then(function () {
      // Validate if the token is authorized.
      if (credentials.canAuthorize) {
        // Attempt to collect the postauth catalog.

        return _this0.updateServices().catch(function () {
          _this0.initFailed = true;
          _this0.logger.warn('services: cannot retrieve postauth catalog');
        });
      }

      // Return a resolved promise for consistent return value.
      return _promise.default.resolve();
    });
  },
  /**
   * Initializer
   *
   * @instance
   * @memberof Services
   * @returns {Services}
   */
  initialize: function initialize() {
    var _this1 = this;
    var catalog = new _serviceCatalog.default();
    var registry = new _serviceRegistry.default();
    var state = new _serviceState.default();
    this._catalogs.set(this.webex, catalog);
    this.registries.set(this.webex, registry);
    this.states.set(this.webex, state);

    // Listen for configuration changes once.
    this.listenToOnce(this.webex, 'change:config', function () {
      _this1.initConfig();
    });

    // wait for webex instance to be ready before attempting
    // to update the service catalogs
    // this can cause a race condition because credentials may
    // not be valid when services is initialized
    this.listenToOnce(this.webex, 'ready', /*#__PURE__*/(0, _asyncToGenerator2.default)(/*#__PURE__*/_regenerator.default.mark(function _callee4() {
      var cachedCatalog, supertoken, email;
      return _regenerator.default.wrap(function (_context4) {
        while (1) switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 1;
            return _this1._loadCatalogFromCache();
          case 1:
            cachedCatalog = _context4.sent;
            if (!cachedCatalog) {
              _context4.next = 2;
              break;
            }
            catalog.isReady = true;
            return _context4.abrupt("return");
          case 2:
            supertoken = _this1.webex.credentials.supertoken; // Validate if the supertoken exists.
            if (supertoken && supertoken.access_token) {
              _this1.initServiceCatalogs().then(function () {
                catalog.isReady = true;
              }).catch(function (error) {
                _this1.initFailed = true;
                _this1.logger.error("services: failed to init initial services when credentials available, ".concat(error === null || error === void 0 ? void 0 : error.message));
              });
            } else {
              email = _this1.webex.config.email;
              _this1.collectPreauthCatalog(email ? {
                email: email
              } : undefined).catch(function (error) {
                _this1.initFailed = true;
                _this1.logger.error("services: failed to init initial services when no credentials available, ".concat(error === null || error === void 0 ? void 0 : error.message));
              });
            }
          case 3:
          case "end":
            return _context4.stop();
        }
      }, _callee4);
    })));
  },
  version: "3.12.0"
});
/* eslint-enable no-underscore-dangle */
var _default = exports.default = Services;
