// @webex/http-core/dist/http-error-subtypes.js
"use strict";

var _Reflect$construct = require("@babel/runtime-corejs2/core-js/reflect/construct");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = makeSubTypes;
var _assign = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/object/assign"));
var _parseInt2 = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/parse-int"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/createClass"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/classCallCheck"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inherits"));
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? _Reflect$construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
/*!
 * Copyright (c) 2015-2020 Cisco Systems, Inc. See LICENSE file.
 */

/**
 * @param {Error} Base
 * @returns {undefined}
 */
function makeSubTypes(Base) {
  /**
   * NetworkOrCORSError
   */
  var NetworkOrCORSError = /*#__PURE__*/function (_Base) {
    function NetworkOrCORSError() {
      (0, _classCallCheck2.default)(this, NetworkOrCORSError);
      return _callSuper(this, NetworkOrCORSError, arguments);
    }
    (0, _inherits2.default)(NetworkOrCORSError, _Base);
    return (0, _createClass2.default)(NetworkOrCORSError);
  }(Base);
  /**
   * BadRequest
   */
  var BadRequest = /*#__PURE__*/function (_Base2) {
    function BadRequest() {
      (0, _classCallCheck2.default)(this, BadRequest);
      return _callSuper(this, BadRequest, arguments);
    }
    (0, _inherits2.default)(BadRequest, _Base2);
    return (0, _createClass2.default)(BadRequest);
  }(Base);
  /**
   * Unauthorized
   */
  var Unauthorized = /*#__PURE__*/function (_BadRequest) {
    function Unauthorized() {
      (0, _classCallCheck2.default)(this, Unauthorized);
      return _callSuper(this, Unauthorized, arguments);
    }
    (0, _inherits2.default)(Unauthorized, _BadRequest);
    return (0, _createClass2.default)(Unauthorized);
  }(BadRequest);
  /**
   * PaymentRequired
   */
  var PaymentRequired = /*#__PURE__*/function (_BadRequest2) {
    function PaymentRequired() {
      (0, _classCallCheck2.default)(this, PaymentRequired);
      return _callSuper(this, PaymentRequired, arguments);
    }
    (0, _inherits2.default)(PaymentRequired, _BadRequest2);
    return (0, _createClass2.default)(PaymentRequired);
  }(BadRequest);
  /**
   * Forbidden
   */
  var Forbidden = /*#__PURE__*/function (_BadRequest3) {
    function Forbidden() {
      (0, _classCallCheck2.default)(this, Forbidden);
      return _callSuper(this, Forbidden, arguments);
    }
    (0, _inherits2.default)(Forbidden, _BadRequest3);
    return (0, _createClass2.default)(Forbidden);
  }(BadRequest);
  /**
   * NotFound
   */
  var NotFound = /*#__PURE__*/function (_BadRequest4) {
    function NotFound() {
      (0, _classCallCheck2.default)(this, NotFound);
      return _callSuper(this, NotFound, arguments);
    }
    (0, _inherits2.default)(NotFound, _BadRequest4);
    return (0, _createClass2.default)(NotFound);
  }(BadRequest);
  /**
   * MethodNotAllowed
   */
  var MethodNotAllowed = /*#__PURE__*/function (_BadRequest5) {
    function MethodNotAllowed() {
      (0, _classCallCheck2.default)(this, MethodNotAllowed);
      return _callSuper(this, MethodNotAllowed, arguments);
    }
    (0, _inherits2.default)(MethodNotAllowed, _BadRequest5);
    return (0, _createClass2.default)(MethodNotAllowed);
  }(BadRequest);
  /**
   * NotAcceptable
   */
  var NotAcceptable = /*#__PURE__*/function (_BadRequest6) {
    function NotAcceptable() {
      (0, _classCallCheck2.default)(this, NotAcceptable);
      return _callSuper(this, NotAcceptable, arguments);
    }
    (0, _inherits2.default)(NotAcceptable, _BadRequest6);
    return (0, _createClass2.default)(NotAcceptable);
  }(BadRequest);
  /**
   * ProxyAuthenticationRequired
   */
  var ProxyAuthenticationRequired = /*#__PURE__*/function (_BadRequest7) {
    function ProxyAuthenticationRequired() {
      (0, _classCallCheck2.default)(this, ProxyAuthenticationRequired);
      return _callSuper(this, ProxyAuthenticationRequired, arguments);
    }
    (0, _inherits2.default)(ProxyAuthenticationRequired, _BadRequest7);
    return (0, _createClass2.default)(ProxyAuthenticationRequired);
  }(BadRequest);
  /**
   * RequestTimeout
   */
  var RequestTimeout = /*#__PURE__*/function (_BadRequest8) {
    function RequestTimeout() {
      (0, _classCallCheck2.default)(this, RequestTimeout);
      return _callSuper(this, RequestTimeout, arguments);
    }
    (0, _inherits2.default)(RequestTimeout, _BadRequest8);
    return (0, _createClass2.default)(RequestTimeout);
  }(BadRequest);
  /**
   * Conflict
   */
  var Conflict = /*#__PURE__*/function (_BadRequest9) {
    function Conflict() {
      (0, _classCallCheck2.default)(this, Conflict);
      return _callSuper(this, Conflict, arguments);
    }
    (0, _inherits2.default)(Conflict, _BadRequest9);
    return (0, _createClass2.default)(Conflict);
  }(BadRequest);
  /**
   * Gone
   */
  var Gone = /*#__PURE__*/function (_BadRequest0) {
    function Gone() {
      (0, _classCallCheck2.default)(this, Gone);
      return _callSuper(this, Gone, arguments);
    }
    (0, _inherits2.default)(Gone, _BadRequest0);
    return (0, _createClass2.default)(Gone);
  }(BadRequest);
  /**
   * LengthRequired
   */
  var LengthRequired = /*#__PURE__*/function (_BadRequest1) {
    function LengthRequired() {
      (0, _classCallCheck2.default)(this, LengthRequired);
      return _callSuper(this, LengthRequired, arguments);
    }
    (0, _inherits2.default)(LengthRequired, _BadRequest1);
    return (0, _createClass2.default)(LengthRequired);
  }(BadRequest);
  /**
   * PreconditionFailed
   */
  var PreconditionFailed = /*#__PURE__*/function (_BadRequest10) {
    function PreconditionFailed() {
      (0, _classCallCheck2.default)(this, PreconditionFailed);
      return _callSuper(this, PreconditionFailed, arguments);
    }
    (0, _inherits2.default)(PreconditionFailed, _BadRequest10);
    return (0, _createClass2.default)(PreconditionFailed);
  }(BadRequest);
  /**
   * RequestEntityTooLarge
   */
  var RequestEntityTooLarge = /*#__PURE__*/function (_BadRequest11) {
    function RequestEntityTooLarge() {
      (0, _classCallCheck2.default)(this, RequestEntityTooLarge);
      return _callSuper(this, RequestEntityTooLarge, arguments);
    }
    (0, _inherits2.default)(RequestEntityTooLarge, _BadRequest11);
    return (0, _createClass2.default)(RequestEntityTooLarge);
  }(BadRequest);
  /**
   * RequestUriTooLong
   */
  var RequestUriTooLong = /*#__PURE__*/function (_BadRequest12) {
    function RequestUriTooLong() {
      (0, _classCallCheck2.default)(this, RequestUriTooLong);
      return _callSuper(this, RequestUriTooLong, arguments);
    }
    (0, _inherits2.default)(RequestUriTooLong, _BadRequest12);
    return (0, _createClass2.default)(RequestUriTooLong);
  }(BadRequest);
  /**
   * UnsupportedMediaType
   */
  var UnsupportedMediaType = /*#__PURE__*/function (_BadRequest13) {
    function UnsupportedMediaType() {
      (0, _classCallCheck2.default)(this, UnsupportedMediaType);
      return _callSuper(this, UnsupportedMediaType, arguments);
    }
    (0, _inherits2.default)(UnsupportedMediaType, _BadRequest13);
    return (0, _createClass2.default)(UnsupportedMediaType);
  }(BadRequest);
  /**
   * RequestRangeNotSatisfiable
   */
  var RequestRangeNotSatisfiable = /*#__PURE__*/function (_BadRequest14) {
    function RequestRangeNotSatisfiable() {
      (0, _classCallCheck2.default)(this, RequestRangeNotSatisfiable);
      return _callSuper(this, RequestRangeNotSatisfiable, arguments);
    }
    (0, _inherits2.default)(RequestRangeNotSatisfiable, _BadRequest14);
    return (0, _createClass2.default)(RequestRangeNotSatisfiable);
  }(BadRequest);
  /**
   * ExpectationFailed
   */
  var ExpectationFailed = /*#__PURE__*/function (_BadRequest15) {
    function ExpectationFailed() {
      (0, _classCallCheck2.default)(this, ExpectationFailed);
      return _callSuper(this, ExpectationFailed, arguments);
    }
    (0, _inherits2.default)(ExpectationFailed, _BadRequest15);
    return (0, _createClass2.default)(ExpectationFailed);
  }(BadRequest);
  /**
   * TooManyRequests
   */
  var TooManyRequests = /*#__PURE__*/function (_BadRequest16) {
    function TooManyRequests() {
      (0, _classCallCheck2.default)(this, TooManyRequests);
      return _callSuper(this, TooManyRequests, arguments);
    }
    (0, _inherits2.default)(TooManyRequests, _BadRequest16);
    return (0, _createClass2.default)(TooManyRequests);
  }(BadRequest);
  /**
   * InternalServerError
   */
  var InternalServerError = /*#__PURE__*/function (_Base3) {
    function InternalServerError() {
      (0, _classCallCheck2.default)(this, InternalServerError);
      return _callSuper(this, InternalServerError, arguments);
    }
    (0, _inherits2.default)(InternalServerError, _Base3);
    return (0, _createClass2.default)(InternalServerError);
  }(Base);
  /**
   * NotImplemented
   */
  var NotImplemented = /*#__PURE__*/function (_InternalServerError) {
    function NotImplemented() {
      (0, _classCallCheck2.default)(this, NotImplemented);
      return _callSuper(this, NotImplemented, arguments);
    }
    (0, _inherits2.default)(NotImplemented, _InternalServerError);
    return (0, _createClass2.default)(NotImplemented);
  }(InternalServerError);
  /**
   * BadGateway
   */
  var BadGateway = /*#__PURE__*/function (_InternalServerError2) {
    function BadGateway() {
      (0, _classCallCheck2.default)(this, BadGateway);
      return _callSuper(this, BadGateway, arguments);
    }
    (0, _inherits2.default)(BadGateway, _InternalServerError2);
    return (0, _createClass2.default)(BadGateway);
  }(InternalServerError);
  /**
   * ServiceUnavailable
   */
  var ServiceUnavailable = /*#__PURE__*/function (_InternalServerError3) {
    function ServiceUnavailable() {
      (0, _classCallCheck2.default)(this, ServiceUnavailable);
      return _callSuper(this, ServiceUnavailable, arguments);
    }
    (0, _inherits2.default)(ServiceUnavailable, _InternalServerError3);
    return (0, _createClass2.default)(ServiceUnavailable);
  }(InternalServerError);
  /**
   * GatewayTimeout
   */
  var GatewayTimeout = /*#__PURE__*/function (_InternalServerError4) {
    function GatewayTimeout() {
      (0, _classCallCheck2.default)(this, GatewayTimeout);
      return _callSuper(this, GatewayTimeout, arguments);
    }
    (0, _inherits2.default)(GatewayTimeout, _InternalServerError4);
    return (0, _createClass2.default)(GatewayTimeout);
  }(InternalServerError);
  /**
   * HttpVersionNotSupported
   */
  var HttpVersionNotSupported = /*#__PURE__*/function (_InternalServerError5) {
    function HttpVersionNotSupported() {
      (0, _classCallCheck2.default)(this, HttpVersionNotSupported);
      return _callSuper(this, HttpVersionNotSupported, arguments);
    }
    (0, _inherits2.default)(HttpVersionNotSupported, _InternalServerError5);
    return (0, _createClass2.default)(HttpVersionNotSupported);
  }(InternalServerError);
  (0, _assign.default)(Base, {
    0: NetworkOrCORSError,
    NetworkOrCORSError: NetworkOrCORSError,
    400: BadRequest,
    BadRequest: BadRequest,
    401: Unauthorized,
    Unauthorized: Unauthorized,
    402: PaymentRequired,
    PaymentRequired: PaymentRequired,
    403: Forbidden,
    Forbidden: Forbidden,
    404: NotFound,
    NotFound: NotFound,
    405: MethodNotAllowed,
    MethodNotAllowed: MethodNotAllowed,
    406: NotAcceptable,
    NotAcceptable: NotAcceptable,
    407: ProxyAuthenticationRequired,
    ProxyAuthenticationRequired: ProxyAuthenticationRequired,
    408: RequestTimeout,
    RequestTimeout: RequestTimeout,
    409: Conflict,
    Conflict: Conflict,
    410: Gone,
    Gone: Gone,
    411: LengthRequired,
    LengthRequired: LengthRequired,
    412: PreconditionFailed,
    PreconditionFailed: PreconditionFailed,
    413: RequestEntityTooLarge,
    RequestEntityTooLarge: RequestEntityTooLarge,
    414: RequestUriTooLong,
    RequestUriTooLong: RequestUriTooLong,
    415: UnsupportedMediaType,
    UnsupportedMediaType: UnsupportedMediaType,
    416: RequestRangeNotSatisfiable,
    RequestRangeNotSatisfiable: RequestRangeNotSatisfiable,
    417: ExpectationFailed,
    ExpectationFailed: ExpectationFailed,
    429: TooManyRequests,
    TooManyRequests: TooManyRequests,
    500: InternalServerError,
    InternalServerError: InternalServerError,
    501: NotImplemented,
    NotImplemented: NotImplemented,
    502: BadGateway,
    BadGateway: BadGateway,
    503: ServiceUnavailable,
    ServiceUnavailable: ServiceUnavailable,
    504: GatewayTimeout,
    GatewayTimeout: GatewayTimeout,
    505: HttpVersionNotSupported,
    HttpVersionNotSupported: HttpVersionNotSupported,
    select: function select(statusCode) {
      if (typeof statusCode === 'undefined' || statusCode === null) {
        return Base;
      }
      statusCode = statusCode.statusCode || statusCode;
      var E = Base[statusCode];
      if (E) {
        return E;
      }

      // Fallback to the default for the category (e.g. BadRequest for 429)
      statusCode = "".concat(statusCode.toString().split('').shift(), "00");
      statusCode = (0, _parseInt2.default)(statusCode, 10);
      return Base[statusCode] || Base;
    }
  });
}
